---
title: "01-Importing-data-RKC"
author: "Laura H Spencer"
date: "01/09/2022"
output: html_document
---

In this notebook I will import gene counts file that were generated by Bowtie --> FeatureCounts 

### Check working directory 

```{r}
getwd()
```

### Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
source("../references/biostats.R")

list.of.packages <- c("DESeq2", "RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "corrplot", "PerformanceAnalytics", "cowplot", "here") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})

# Need these packages for SNP assessment 
# if (!requireNamespace("BiocManager", quietly=TRUE))
#     install.packages("BiocManager")
# BiocManager::install("gdsfmt")
# BiocManager::install("SNPRelate")
```

### Import sample info, library/file names, and then join 

```{r}
sample.info <- read.csv("../data/RCK-exp-design.csv", header=T, na.strings = NA) %>%
          mutate_at(vars(Tank, Treatment, Treatment_Tank), as.factor)
save(sample.info, file="../data/sample.info")
```

### Import counts matrix file as dataframe, extract sample # from file/column names to simplify them 

```{r}
counts <- data.frame(read.table("../results/featureCounts/featurecounts_redking_gene-v2", header = T, stringsAsFactors = F, fill = FALSE)) %>%
          column_to_rownames(var="Geneid") %>% 
  rename_all(~as.character(str_sub(colnames(counts)) %>% 
                             gsub("X.scratch.lspencer.2022.redking.OA.aligned.bowtie.sorted.deduped.", "", .) %>% 
                             gsub(".sorted.bam", "", .)))

save(counts, file = "../results/gene-counts-v2") #save R object with all gene counts 
```

### Summarize counts and visualize (remove last column - that's undetermined counts)

```{r}
print(paste("Number of samples:", ncol(counts %>% select(contains("Tank"))), sep=" "))
print(paste("Total number of genes in dataframe:", prettyNum(nrow(counts), big.mark = ","), sep=" "))
print(paste("Average number of genes per sample:", prettyNum(mean(colSums(counts %>% select(contains("Tank")) != 0)), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(counts %>% select(contains("Tank")))), big.mark = ","), sep=" "))
#print(paste("Counts for", colnames(counts %>% select(contains("Tank"))), ":",  prettyNum(colSums(counts %>% select(contains("Tank"))), big.mark = ","), sep=" "))


#inspect total counts by sample
 ggplotly(
   ggplot(data.frame(colSums(counts %>% select(contains("Tank")))) %>% 
            dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample")) + 
     geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

# Filter sample sequencing stats 

## Inspect total # reads, set minimum for retaining samples 

```{r}
multiqc.stats <- read.delim(file="../results/fastqc/multiqc_data_trimmed-v2/multiqc_general_stats.txt", sep = "\t") 
colnames(multiqc.stats) <- gsub("FastQC_mqc.generalstats.fastqc.", "", colnames(multiqc.stats)) 
multiqc.stats <- multiqc.stats %>% 
  mutate(total_unique = total_sequences*(100-percent_duplicates)/100) %>%
  left_join(sample.info)
```
## Investigate possible correlations among # reads and prep stuff 

# TO DO: Get library prep stats from facility, and any other possible factors from Chris 

```{r}
cor(multiqc.stats %>% mutate(Tank=as.numeric(Tank)) %>% dplyr::select_if(is.numeric), use="complete.obs") %>% corrplot::corrplot(tl.cex=.45)
```

## OPTIONAL: FILTER WHOLE SAMPLES. RemovE whole samples from the data set here if needed  

```{r}
# remove.list <- c("Undetermined", (multiqc.stats %>% filter(pcr.cycles>17))$Sample) 
# counts.filtered <- counts[ , -which(names(counts) %in% remove.list)] %>% mutate_at(vars(starts_with("feature")), as.character)
# ncol(counts.filtered)
# key.filtered <- key[ -which(key$sample %in% remove.list), ]
# nrow(key.filtered)
# nrow(key.filtered) == ncol(counts.filtered) #should = TRUE.
# save(key.filtered, file="../raw-data/key.filtered") #save key object. IMPORTANT: this needs to be done here AFTER filtering out whole samples 
#save(counts.filtered, file = "../results/gene-counts-filtered")
```


### Remove extraneous gene info, then transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
```{r}
str(counts) #columns #1-#5 contain extraneous gene info (chr, start, end, strand, length). 
counts.t <- t(counts[,-1:-5]) #remove extraneous columns, transform data to have each sample a row, each column a gene 

# # do the same for filtered datasets 
# counts.t <- t(counts.filtered[,-1:-5]) #transform data to have each sample a row, each column a gene 
```

### Save counts file, and transformed counts file 
```{r}
save(counts, file = "../results/gene-counts")
save(counts.t, file = "../results/gene-counts-trans")
```

### Import gene feature file to get gene ID info for GO enrichment analysis, etc. 
# NEED TO FIX THE STRING EXTRACT, ISSUE WITH VARIOUS ";" AFTER EACH ITEM IN "notes" 

```{r}
# # Read in O. lurida gene file that connects OLUR gene ID to uniprot accession number 
# P.plat.gff <- read_delim(file = "../references/EVM.out_new.gff3", delim = "\t", col_names = c("scaffold", "source", "feature", "start", "end", "unknown1", "strand", "unknown2", "notes")) %>%
# # Split giant gene "Notes" column into separate columns 
#   mutate(ID=str_extract(notes, "ID=(.*?);"),
#        Parent=str_extract(notes, "Parent=(.*?)"), 
#   Name=str_extract(notes, "Name=(.*?)")) %>% 
#   
#   #remove extraneous info from gene ID
#   mutate(ID=str_remove(ID, "ID=")) %>% mutate(ID=str_remove(ID, ";")) %>% 
#   mutate(Parent=str_remove(Parent, "Parent=")) %>% mutate(Parent=str_remove(Parent, ";"))
# 
# save(P.plat.gff, file = "../references/P.plat.genes")
```
