---
title: "01-Importing-data-STAR"
author: "Laura H Spencer"
date: "2/9/2022"
output: html_document
---

### Load libraries

```{r message=FALSE, warning=FALSE, results=FALSE}
list.of.packages <- c("tidyverse", "reshape2", "here", "plotly", "purrr") #add new libraries here 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
02
# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```


```{r}
filenames <- read_csv(file="../results/star/star.v2/countsfilenames.csv", col_names = c("filename", "Sample"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

sample.info <- read.csv("../data/RCK-exp-design.csv", header=T, na.strings = NA) %>%
          mutate_at(vars(Tank, Treatment, Treatment_Tank), as.factor) %>%
  left_join(filenames)

```

## Read in count files, selecting only the first 2 columns (1=gene ID, 2=count); rename columns 

```{r}
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$Sample)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read.delim(file=files[i]))[-1:-4,1:2]
    names(file_list[[i]]) <- c("gene", filenames$Sample[i])
    print(paste("Total COUNTS,", names(file_list[[i]][2]), ":", prettyNum(sum(file_list[[i]][2]), big.mark=","), sep=" "))
    print(paste("Total GENES,", names(file_list[[i]][2]), ":", prettyNum(nrow(file_list[[i]] %>% filter(.[[2]] != 0)), big.mark=","), sep=" "))
}
```

```{r}
counts <- file_list %>% purrr::reduce(full_join, by = "gene") %>% column_to_rownames(var="gene") 
```

### Summarize counts and visualize (remove last column - that's undetermined counts)

```{r}
print(paste("Number of samples:", ncol(counts %>% select(contains("Tank"))), sep=" "))
print(paste("Total number of genes in dataframe:", prettyNum(nrow(counts), big.mark = ","), sep=" "))
print(paste("Average number of genes per sample:", prettyNum(mean(colSums(counts %>% select(contains("Tank")) != 0)), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(counts %>% select(contains("Tank")))), big.mark = ","), sep=" "))
#print(paste("Counts for", colnames(counts %>% select(contains("Tank"))), ":",  prettyNum(colSums(counts %>% select(contains("Tank"))), big.mark = ","), sep=" "))


#inspect total counts by sample
 ggplotly(
   ggplot(data.frame(colSums(counts %>% select(contains("Tank")))) %>% 
            dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample")) + 
     geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

## OPTIONAL: FILTER WHOLE SAMPLES. 

## DECISION FEBRUARY 9TH, 2022: Removing sample "Tank_7_Crab_4", which is an outlier in the PCAs and heat maps 
Remove whole samples from the data set and sample info here if needed

```{r}
remove.list <- c("Tank_7_Crab_4")
counts <- counts[ , -which(names(counts) %in% remove.list)]
sample.info <- sample.info[ -which(sample.info$Sample %in% remove.list), ]

# resave sample info object
save(sample.info, file="../data/sample.info")

nrow(sample.info) == ncol(counts) #should = TRUE
```

## OPTIONAL: FILTER WHOLE SCAFFOLDS 

This code provides the option to remove data that mapped to scaffolds 105+. This is possibly beneficial because I don't know whether or not those scaffolds are also included within scaffolds 1-104 (i.e. the chromosomes). 

NOTE:  For exploration purposes I will NOT remove them, and will see if expressed data mapping to scaffolds 105+ have the exact same genes/counts as on chromosomes. 

After preliminary examination of blast results (by looking at genes, lengths, and counts) it's difficult to tell whether the scaffolds 105+ are subsets of the chromosomes 1-104. For now I will retain scaffolds 105+. 


```{r}
# counts <- counts %>% separate(Chr, into = c("X", "Y", "scaffold"), sep = "_", remove = F) %>%
#   mutate(scaffold=as.integer(scaffold)) %>% filter(scaffold<=104) %>% dplyr::select(-X, -Y, -scaffold)
```

## Optional: FILTER GENES THAT ARE OUTLIERS IN OUTLIER SAMPLE (TANK_7_CRAB_4)  

## DECISION FEBRUARY 9TH, 2022: DO NOT REMOVE INDIVIDUAL GENES, INSTEAD REMOVE OUTLIER SAMPLE (ABOVE)

The sample Tank 7 Crab 4 is weird, and has outlier counts in many genes. I identified them in my Outlier notebook, so here I will remove them from the count matrix. 

NOTE: you cannot simply replace counts for some samples with NA. DESeq2 doesn't accept those. 

```{r}
# counts <- counts[-which(rownames(counts) %in% outliers),]
```

### Transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
```{r}
counts.t <- as.data.frame(t(counts)) #remove extraneous columns, transform data to have each sample a row, each column a gene 
```

## Optional 

### Pre-filtering - remove rows (genes) with less than a total of 10 reads (across all samples) 

NOTE: should update to remove low-frequency genes. This is performed in https://doi.org/10.1186/s12864-017-4392-0
E.g. "Genes with mean count less than ten across all samples were removed."

```{r}
keep <- colSums(counts.t, na.rm=TRUE) >= 10
counts.ts <- counts.t[,keep]

print(paste("# genes remaining after pre-filtering:", ncol(counts.ts)))
print(paste("# of genes dropped:", ncol(counts.t) - ncol(counts.ts), sep=" "))
```

### Save counts file, and transformed counts file 
```{r}
save(counts, file = "../results/gene-counts")
save(counts.t, file = "../results/gene-counts-trans")
```

```{bash}
for file in ../results/star/star.v2/*.Log.final.out
do
echo $file
cat $file
done >> ../results/star/star.v2/AllSamples.log
```

```{bash}
grep "% of reads unmapped: other" ../results/star/star.v2/AllSamples.log
```

## Save DEG lists identified using STAR aligned data 

```{r}
res.pco2.Star <- res.pco2.AM # save list of all genes in DESeq2 analysis regardless of p-value
res.pco2.AM.Star <- subset(res.pco2.AM, padj < 0.05)
res.pco2.AL.Star <- subset(res.pco2.AL, padj < 0.05)
res.pco2.ML.Star <- subset(res.pco2.ML, padj < 0.05)
```
