---
title: "01-Importing-data-RKC"
author: "Laura H Spencer"
date: "01/09/2022"
output: html_document
---

In this notebook I will import gene counts file that were generated by Bowtie --> FeatureCounts 

### Check working directory 

```{r}
getwd()
```

### Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
source("../references/biostats.R")

list.of.packages <- c("DESeq2", "RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "corrplot", "PerformanceAnalytics", "cowplot", "here", "janitor") #add new libraries here 
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})

# Need these packages for SNP assessment 
# if (!requireNamespace("BiocManager", quietly=TRUE))
#     install.packages("BiocManager")
# BiocManager::install("gdsfmt")
# BiocManager::install("SNPRelate")
```

### Import sample info, library/file names, and then join 

```{r}
library.stats <- read.csv("../data/library-prep-stats.csv", header=T) %>% clean_names %>%
  mutate_at(vars(tank, p_h, batch, date_extracted), as.factor)

sample.info <- read.csv("../data/RCK-exp-design.csv", header=T, na.strings = NA) %>%
          mutate_at(vars(Tank, Treatment, Treatment_Tank), as.factor) %>%
  left_join( ., library.stats, by=c("Sample"="tank_crab"))

save(sample.info, file="../data/sample.info")
```

## Import library stats from MultiQC and from library prep

```{r}
multiqc.stats <- read.delim(file="../results/fastqc/multiqc_data_trimmed-v2/multiqc_general_stats.txt", sep = "\t") 
colnames(multiqc.stats) <- gsub("FastQC_mqc.generalstats.fastqc.", "", colnames(multiqc.stats)) 
multiqc.stats <- multiqc.stats %>% 
  mutate(total_unique = total_sequences*(100-percent_duplicates)/100) %>%
  left_join(sample.info)
```

## Investigate possible correlations among # reads and prep stuff 

```{r}
cor(multiqc.stats %>% mutate(Tank=as.numeric(Tank)) %>% dplyr::select_if(is.numeric) %>% 
      dplyr::select(-Crab, -crab, -initial_volume, -volume_used_at_start_of_library_prep, -water_to_add_to_bring_to_50u_l, 
                    -diluted_concentration, -pcr_cycles, -volume_used_in_qc), use="complete.obs") %>% corrplot::corrplot(tl.cex=.75)
```

### Import counts matrix file as dataframe, extract sample # from file/column names to simplify them 

```{r}
counts <- data.frame(read.table("../results/featureCounts/featurecounts_redking_gene-v1", header = T, stringsAsFactors = F, fill = FALSE))
counts <- counts %>%  column_to_rownames(var="Geneid")
counts <- counts %>% rename_all(~as.character(str_sub(colnames(counts)) %>% 
                             gsub("X.scratch.lspencer.2022.redking.OA.aligned.bowtie.sorted.deduped.", "", .) %>% 
                             gsub(".sorted.bam", "", .)))

save(counts, file = "../results/gene-counts-v2") #save R object with all gene counts 
```

### Summarize counts and visualize (remove last column - that's undetermined counts)

```{r}
print(paste("Number of samples:", ncol(counts %>% select(contains("Tank"))), sep=" "))
print(paste("Total number of genes in dataframe:", prettyNum(nrow(counts), big.mark = ","), sep=" "))
print(paste("Average number of genes per sample:", prettyNum(mean(colSums(counts %>% select(contains("Tank")) != 0)), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(counts %>% select(contains("Tank")))), big.mark = ","), sep=" "))
#print(paste("Counts for", colnames(counts %>% select(contains("Tank"))), ":",  prettyNum(colSums(counts %>% select(contains("Tank"))), big.mark = ","), sep=" "))


#inspect total counts by sample
 ggplotly(
   ggplot(data.frame(colSums(counts %>% select(contains("Tank")))) %>% 
            dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample")) + 
     geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```


## OPTIONAL: FILTER WHOLE SAMPLES. 
Remove whole samples from the data set and sample info here if needed

January 19th, 2022:  Removing sample "Tank_7_Crab_4", which is an outlier in the PCAs and heat maps 

```{r}
# remove.list <- c("Undetermined", (multiqc.stats %>% filter(pcr.cycles>17))$Sample) 
remove.list <- c("Tank_7_Crab_4") 
counts <- counts[ , -which(names(counts) %in% remove.list)]
sample.info <- sample.info[ -which(sample.info$Sample %in% remove.list), ]

# resave sample info object
save(sample.info, file="../data/sample.info")

nrow(sample.info) == ncol(counts)-5 #should = TRUE.
```


## OPTIONAL: FILTER WHOLE SCAFFOLDS 

This code provides the option to remove data that mapped to scaffolds 105+. This is possibly beneficial because I don't know whether or not those scaffolds are also included within scaffolds 1-104 (i.e. the chromosomes). 

NOTE:  For exploration purposes I will NOT remove them, and will see if expressed data mapping to scaffolds 105+ have the exact same genes/counts as on chromosomes. 

```{r}
# counts <- counts %>% separate(Chr, into = c("X", "Y", "scaffold"), sep = "_", remove = F) %>%
#   mutate(scaffold=as.integer(scaffold)) %>% filter(scaffold<=104) %>% dplyr::select(-X, -Y, -scaffold)
```

### Remove extraneous gene info, then transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
```{r}
#str(counts) #columns #1-#5 contain extraneous gene info (chr, start, end, strand, length). 
counts.t <- t(counts[,-1:-5]) #remove extraneous columns, transform data to have each sample a row, each column a gene 
```

### Save counts file, and transformed counts file 
```{r}
save(counts, file = "../results/gene-counts")
save(counts.t, file = "../results/gene-counts-trans")
```

### Import gene feature file to get gene ID info for GO enrichment analysis, etc. 
# NEED TO FIX THE STRING EXTRACT, ISSUE WITH VARIOUS ";" AFTER EACH ITEM IN "notes" 

```{r}
# # Read in file that connects OLUR gene ID to uniprot accession number 
# P.plat.gff <- read_delim(file = "../references/EVM.out_new.gff3", delim = "\t", col_names = c("scaffold", "source", "feature", "start", "end", "unknown1", "strand", "unknown2", "notes")) %>%
# # Split giant gene "Notes" column into separate columns 
#   mutate(ID=str_extract(notes, "ID=(.*?);"),
#        Parent=str_extract(notes, "Parent=(.*?)"), 
#   Name=str_extract(notes, "Name=(.*?)")) %>% 
#   
#   #remove extraneous info from gene ID
#   mutate(ID=str_remove(ID, "ID=")) %>% mutate(ID=str_remove(ID, ";")) %>% 
#   mutate(Parent=str_remove(Parent, "Parent=")) %>% mutate(Parent=str_remove(Parent, ";"))
# 
# save(P.plat.gff, file = "../references/P.plat.genes")
```

## Import Uniprot/Swissprot annotated blue king crab genes 

blastx output format 6 is tab file with the following columns: 
1. qaccver = Query accesion.version
2. saccver = Subject accession.version
3. pident = Percentage of identical matches
4. length = Alignment length
5. mismatch = Number of mismatches
6. gapopen = Number of gap openings
7. qstart = Start of alignment in query
8. qend = End of alignment in query
9. sstart =  Start of alignment in subject
10. send = End of alignment in subject
11. evalue =  Expect value
12. bitscore = Bit score

```{r}
P.plat.blast <- read_delim(file = "../references/P.platypus.gene_blastx.tab", delim = "\t", col_names = c("qaccver", "saccver", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore")) %>% 
  separate(qaccver, sep = ":", into = c("geneID","na","Chr", "range")) %>%
  separate(range, sep="-", into=c("Start", "End")) %>% select(-na) %>%
  mutate_at(vars(Start, End), as.numeric) %>%
  separate(saccver, sep="\\|", into=c("na", "SPID", "gene.Uni"), remove = F) %>% 
  select(-na) %>% separate(gene.Uni, sep="_", into=c("gene.Uni", "species"), remove=T) %>% 
  group_by(geneID, Chr, Start, End) %>% slice(which.min(evalue))   # where multiple blast hits for same gene, select one with minimum e-value

write.table(P.plat.blast, file = "../references/P.platypus.gene_blastx_split.tab",quote = F, sep = "\t")
```

## NOTE:  I needed GO IDs and gene functions, which weren't included in the blast file.  So I opened the P.platypus.gene_blastx_split.tab file in Excel, copied the column containing all Uniprot IDs, then pasted those into the tool Uniprot batch retrieval tool (https://www.uniprot.org/uploadlists/), and selected the columns: Entry, Entry name, Reviewed(?), Protein names, Gene names, Organism, Gene ontology (biological process), Gene ontology (GO), Gene ontology Ids. I then downloaded all entries to a tab file, saved as: /references/P.platypus_genes_GO.tab. Now I'll read that into R and join with the P.plat.blast dataframe to link GO IDs with genes for exploration and enrichment analyses. 

```{r}
P.plat.blast.GO <- left_join(P.plat.blast, read_delim(file = "../references/P.platypus_genes_GO.tab", delim = "\t"),
                             by = c("SPID"="Entry")) %>% ungroup()
```


```{r}
counts.annot <- right_join(
  P.plat.blast %>% select(geneID, Chr, Start, End, saccver, SPID, gene.Uni, evalue), 
  counts %>% rownames_to_column(var = "geneID"), "geneID") %>% filter(!is.na(SPID))
write.csv(counts.annot, file = "../results/counts.annotated.csv", row.names = F, quote = F)
View(counts.annot)
```


