---
title: "01-Importing-data-STAR"
author: "Laura H Spencer"
date: "2/9/2022"
output: html_document
---

### Load libraries

```{r message=FALSE, warning=FALSE, results=FALSE}
list.of.packages <- c("tidyverse", "reshape2", "here", "plotly", "purrr", "janitor") #add new libraries here 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
02
# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})
```
Generate tab-separated file that lists file names 

```{bash}
# For STAR
for file in ../results/star/star.v2/*.ReadsPerGene.out.tab
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> ../results/star/star.v2/countsfilenames.txt
done

## For STAR+RSEM
#rm ../results/rsem/countsfilenames.txt
#for file in ../results/rsem/*.genes.results
#do
#filename="$(echo $file)"
#sample="$(basename -a $filename | cut -d "." -f 1)"
#printf "%s\t%s\n" "$filename" "$sample" >> ../results/rsem/countsfilenames.txt
#done
```

```{bash}
#head ../results/rsem/countsfilenames.txt
```



```{r}
# Using STAR aligned and STAR quantified counts 
filenames <- read_csv(file="../results/star/star.v2/countsfilenames.csv", col_names = c("filename", "Sample"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

# # Using STAR aligned and RSEM quantified counts 
# filenames <- read_delim(file="../results/rsem/countsfilenames.txt", col_names = c("filename", "Sample"))
# files <- file.path(filenames$filename) #extract vector of filenames 
# all(file.exists(files)) #easy code to check that all files exist! 

sample.info <- read.csv("../data/RCK-exp-design.csv", header=T, na.strings = NA) %>%
          mutate_at(vars(Tank, Treatment, Treatment_Tank), as.factor) %>%
  left_join(filenames)
```

## Read in count files

### STAR aligned and quantified 
selecting only the first 2 columns (1=gene ID, 2=count); rename columns 

```{r}
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$Sample)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read.delim(file=files[i]))[-1:-4,1:2]
    names(file_list[[i]]) <- c("gene", filenames$Sample[i])
    print(paste("Total COUNTS,", names(file_list[[i]][2]), ":", prettyNum(sum(file_list[[i]][2]), big.mark=","), sep=" "))
    print(paste("Total GENES,", names(file_list[[i]][2]), ":", prettyNum(nrow(file_list[[i]] %>% filter(.[[2]] != 0)), big.mark=","), sep=" "))
}
```

```{r}
file_list[[1]]
```


### STAR aligned and RSEM quantified 

Columns are:
1) gene_id	
2) transcript_id(s)
3) length	
4) effective_length
5) expected_count
6) TPM
7) FPKM

```{r}
# file_list <- vector(mode = "list", length = nrow(filenames))
# names(file_list) <- c(filenames$Sample)
# 
# for (i in 1:nrow(filenames)) {
#     file_list[[i]] <- data.frame(read.delim(file=files[i]))[,c(1,5)]
#     names(file_list[[i]]) <- c("gene", filenames$Sample[i])
#     print(paste("Total COUNTS,", names(file_list[[i]][2]), ":", prettyNum(sum(file_list[[i]][2]), big.mark=","), sep=" "))
#     print(paste("Total GENES,", names(file_list[[i]][2]), ":", prettyNum(nrow(file_list[[i]] %>% filter(.[[2]] != 0)), big.mark=","), sep=" "))
# }
```

```{r}
# STAR quantified
counts <- file_list %>% purrr::reduce(full_join, by = "gene") %>% column_to_rownames(var="gene")

# # RSEM quantified
# counts <- file_list %>% purrr::reduce(full_join, by = "gene") %>% column_to_rownames(var="gene") 
```

### Summarize counts and visualize (remove last column - that's undetermined counts)

```{r}
print(paste("Number of samples:", ncol(counts %>% select(contains("Tank"))), sep=" "))
print(paste("Total number of genes in dataframe:", prettyNum(nrow(counts), big.mark = ","), sep=" "))
print(paste("Average number of genes per sample:", prettyNum(mean(colSums(counts %>% select(contains("Tank")) != 0)), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(counts %>% select(contains("Tank")))), big.mark = ","), sep=" "))
#print(paste("Counts for", colnames(counts %>% select(contains("Tank"))), ":",  prettyNum(colSums(counts %>% select(contains("Tank"))), big.mark = ","), sep=" "))


#inspect total counts by sample
 ggplotly(
   ggplot(data.frame(colSums(counts %>% select(contains("Tank")))) %>% 
            dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample")) + 
     geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

## OPTIONAL: FILTER WHOLE SAMPLES. 

## DECISION FEBRUARY 9TH, 2022: Removing sample "Tank_7_Crab_4", which is an outlier in the PCAs and heat maps 
Remove whole samples from the data set and sample info here if needed

```{r}
remove.list <- c("Tank_7_Crab_4")
counts <- counts[ , -which(names(counts) %in% remove.list)]
sample.info <- sample.info[ -which(sample.info$Sample %in% remove.list), ]

# resave sample info object
save(sample.info, file="../data/sample.info")

nrow(sample.info) == ncol(counts) #should = TRUE
```

## OPTIONAL: FILTER WHOLE SCAFFOLDS 

This code provides the option to remove data that mapped to scaffolds 105+. This is possibly beneficial because I don't know whether or not those scaffolds are also included within scaffolds 1-104 (i.e. the chromosomes). 

NOTE:  For exploration purposes I will NOT remove them, and will see if expressed data mapping to scaffolds 105+ have the exact same genes/counts as on chromosomes. 

After preliminary examination of blast results (by looking at genes, lengths, and counts) it's difficult to tell whether the scaffolds 105+ are subsets of the chromosomes 1-104. For now I will retain scaffolds 105+. 


```{r}
# counts <- counts %>% separate(Chr, into = c("X", "Y", "scaffold"), sep = "_", remove = F) %>%
#   mutate(scaffold=as.integer(scaffold)) %>% filter(scaffold<=104) %>% dplyr::select(-X, -Y, -scaffold)
```

## Optional: FILTER GENES THAT ARE OUTLIERS IN OUTLIER SAMPLE (TANK_7_CRAB_4)  

## DECISION FEBRUARY 9TH, 2022: DO NOT REMOVE INDIVIDUAL GENES, INSTEAD REMOVE OUTLIER SAMPLE (ABOVE)

The sample Tank 7 Crab 4 is weird, and has outlier counts in many genes. I identified them in my Outlier notebook, so here I will remove them from the count matrix. 

NOTE: you cannot simply replace counts for some samples with NA. DESeq2 doesn't accept those. 

```{r}
# counts <- counts[-which(rownames(counts) %in% outliers),]
```

### Transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
```{r}
counts.t <- as.data.frame(t(counts)) #remove extraneous columns, transform data to have each sample a row, each column a gene 
```

## Optional 

### Pre-filtering - remove rows (genes) with less than a total of 10 reads (across all samples) 

NOTE: should update to remove low-frequency genes. This is performed in https://doi.org/10.1186/s12864-017-4392-0
E.g. "Genes with mean count less than ten across all samples were removed."

```{r}
keep1 <- colMeans(counts.t, na.rm=TRUE) >= 10 #identify genes with mean count >= 10 across all samples (excluding NAs = 10)
keep2 <- rowSums( counts >= 10 ) >= 0.1*43 #identify genes with counts>=10 across at minimum 10% of the samples
keep <- unique(c(names(which(keep1 == T)), names(which(keep2 == T)))) # list of genes meeting one of the two above criteria
counts.ts <- counts.t[,keep]

print(paste("# genes before filtering:", ncol(counts.t)))
print(paste("# genes remaining after pre-filtering:", ncol(counts.ts)))
print(paste("# of genes dropped:", ncol(counts.t) - ncol(counts.ts), sep=" "))
print(paste("% of fragments remaining after pre-filtering: ", signif(100*sum(counts.ts, na.rm = T)/sum(counts.t, na.rm = T), digits = 5), "%", sep=""))
print(paste("Number of fragments dropped: ", signif(sum(counts.t, na.rm = T)-sum(counts.ts, na.rm = T), digits = 5)))
print(paste("% of fragments dropped: ", signif(100*(sum(counts.t, na.rm = T)-sum(counts.ts, na.rm = T))/sum(counts.t, na.rm = T), digits = 5), "%", sep=""))
print(paste("Number of fragments remaining: ", signif(sum(counts.ts, na.rm = T), digits = 5)))
```

### Save counts file, and transformed counts file 
```{r}
save(counts, file = "../results/gene-counts")
save(counts.t, file = "../results/gene-counts-trans")
save(counts.ts, file = "../results/gene-counts-trans-filtered")
```


## Import Blue king crab Best Blast results 

Description of columns: 
```
 ID                 Sequence ID from the reference
 BlastType     BestBlast Hit, basically did the hit originate from the nr (X) or nt (N) database
 SeqLength   Length of the sequence being blasted  
 HitAcc          Accession number of the hit
 HitDesc        Description of the hit
 Score           Score of the alignment between the query and hit
 E.Value        E-Value of the alignment between the query and hit
 PercentIdent  The percent identity of the alignment (how well did they match)
 QueryStart    The base pair point were the alignment starts for the query
 QueryEnd      The base pair point were the alignment ends for the query
 HitStart       The base pair point were the alignment starts for the hit
 HitEnd        The base pair point were the alignment ends for the hit
```

I believe the accession numbers are the GenBank Accessions. 
Mutate "id" slightly - the gene IDs have different strings in the blast results (contains the string "model") and the counts file (contains the string "TU").
Mutate "pecent_ident" to create new column with numeric proportion

```{r}
P.plat.bestblast <- read_delim(file = "../references/bestblast.bkc_ref.tab", delim = "\t") %>% clean_names() %>% 
  mutate(id_gtf=gsub("model", "TU", id)) %>%
  separate(percent_ident, sep=" ", into=c("ratio", "percent_ident.num"), remove = F) %>% 
  mutate(percent_ident.num=gsub("\\(", "", percent_ident.num)) %>% mutate(percent_ident.num=gsub("\\%)", "", percent_ident.num)) %>% 
  mutate(percent_ident.num=as.numeric(percent_ident.num)/100) %>%
  mutate_at(vars(blast_type), as.factor) %>%
  select(-ratio)

# How many genes have hits with various e-values? 
paste("Total number of genes = ", P.plat.bestblast %>% nrow())
paste("Proportion of annotated genes with e-value < 1e-20 = ", (P.plat.bestblast %>% filter(e_value<1e-20) %>% nrow())/(P.plat.bestblast %>% nrow()))
paste("Proportion of annotated genes with e-value < 1e-10 = ", (P.plat.bestblast %>% filter(e_value<1e-10) %>% nrow())/(P.plat.bestblast %>% nrow()))
paste("Proportion of annotated genes with e-value < 1e-5 = ", (P.plat.bestblast %>% filter(e_value<1e-5) %>% nrow())/(P.plat.bestblast %>% nrow()))

save(P.plat.bestblast, file = "../references/P.plat.bestblast")
```
## Annotate gene counts dataframe 

```{r}
counts.annot <- right_join(
  P.plat.bestblast %>% select(id, id_gtf, blast_type, hit_acc, hit_desc, e_value, percent_ident.num),counts %>% 
    rownames_to_column(var = "id_gtf"), "id_gtf")
write.csv(counts.annot, file = "../results/counts.annotated.csv", row.names = F, quote = F)
```


## Annotate some genes with Uniprot IDs and Gene Ontology info 

The below gene annotations were generated by Laura Spencer using blast with the Uniprot/Swissprot database. It is no longer needed, as Giles provided annotations using his Best Blast protocol

## Import Uniprot/Swissprot annotated blue king crab genes 

blastx output format 6 is tab file with the following columns: 
1. qaccver = Query accesion.version
2. saccver = Subject accession.version
3. pident = Percentage of identical matches
4. length = Alignment length
5. mismatch = Number of mismatches
6. gapopen = Number of gap openings
7. qstart = Start of alignment in query
8. qend = End of alignment in query
9. sstart =  Start of alignment in subject
10. send = End of alignment in subject
11. evalue =  Expect value
12. bitscore = Bit score

## NOTE:  I needed GO IDs and gene functions, which weren't included in the blast file.  So I copied the column containing all Uniprot IDs from the P.plat.blast object, then pasted those into the tool Uniprot batch retrieval tool (https://www.uniprot.org/uploadlists/), and selected the columns: Entry, Entry name, Reviewed(?), Protein names, Gene names, Organism, Gene ontology (biological process), Gene ontology (GO), Gene ontology Ids. I then downloaded all entries to a tab file, saved as: /references/P.platypus_genes_GO.tab. Now I'll read that into R and join with the P.plat.blast dataframe to link GO IDs with genes for exploration and enrichment analyses. 

```{r}
P.plat.blast <- read_delim(file = "../references/P.platypus.gene_blastx_1e5.tab", delim = "\t", col_names = c("qaccver", "saccver", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore")) %>%
  separate(qaccver, sep = ":", into = c("geneID","na","Chr", "range")) %>%
  separate(range, sep="-", into=c("Start", "End")) %>% select(-na) %>%
  mutate_at(vars(Start, End), as.numeric) %>%
  separate(saccver, sep="\\|", into=c("na", "SPID", "gene.Uni"), remove = F) %>%
  select(-na) %>% separate(gene.Uni, sep="_", into=c("gene.Uni", "species"), remove=T) %>%
  group_by(geneID, Chr, Start, End) %>% dplyr::slice(which.min(evalue))   # where multiple blast hits for same gene, select one with minimum e-value

# P.plat.blast %>% ungroup() %>% dplyr::select(SPID) %>%  
#   na.omit() %>% unlist() %>% as.vector() %>% write_clip()
  
P.plat.blast.GO <- left_join(P.plat.blast, read_delim(file = "../references/P.platypus_genes_GO.tab", delim = "\t"),
                             by = c("SPID"="Entry")) %>% ungroup() %>% clean_names()

save(P.plat.blast, file = "../references/P.plat.blast")
save(P.plat.blast.GO, file = "../references/P.plat.blast.GO")
```



# BONEYARD 

```{bash}
for file in ../results/star/star.v2/*.Log.final.out
do
echo $file
cat $file
done >> ../results/star/star.v2/AllSamples.log
```

```{bash}
grep "% of reads unmapped: other" ../results/star/star.v2/AllSamples.log
```

## Save DEG lists identified using STAR aligned data 

```{r}
# res.pco2.Star <- res.pco2.AM # save list of all genes in DESeq2 analysis regardless of p-value
# res.pco2.AM.Star <- subset(res.pco2.AM, padj < 0.05)
# res.pco2.AL.Star <- subset(res.pco2.AL, padj < 0.05)
# res.pco2.ML.Star <- subset(res.pco2.ML, padj < 0.05)
```
