---
title: "02-DESeq2-etc-RKS"
author: "Laura H Spencer"
date: "01/09/2022"
output: html_document
---

```{r}
getwd()
```

Load libraries and source scripts 

```{r, message=FALSE, warning=FALSE, results=FALSE}
source("../references/biostats.R")

# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("RCurl", "tidyverse", "vegan", "pheatmap", "pastecs", "factoextra", "FactoMineR", "RColorBrewer", "tibble", "reshape2", "plotly", "cowplot", "clipr", "janitor", "ggpubr", "forcats", "apeglm", "car", "vsn", "devtools", "grid", "gridGraphics")
# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DESeq2", "WGCNA")

# Install BiocManager if needed
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})


# Github packages 
install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)

```

Load counts and sample info 

```{r}
load(file = "../results/gene-counts") #object = counts
load(file="../data/sample.info") #object = sample.info
sample.info
load(file = "../results/gene-counts-trans") #object = counts.t
load(file = "../results/gene-counts-trans-filtered") #object = counts.ts

load(file = "../references/P.camt.blast") #P.camt.blast
load(file = "../references/P.camt.blast.GO") #P.camt.blast.GO
```

Library summary stats

```{r}
# What's the average no. of genes per sample? 
data.frame(rowSums(counts.ts != 0)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample") %>% 
  summarise(mean=mean(count.total, na.rm=T), sd=sd(count.total, na.rm=T), se=sd/sqrt(length(sample)),
            min=min(count.total, na.rm=T), max=max(count.total, na.rm=T))

# How many genes in the filtered count matrix?
ncol(counts.ts)

# No. of fragments per sample? 
data.frame(rowSums(counts.ts)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>% #summary() 
summarise(mean=mean(read.total, na.rm=T), sd=sd(read.total, na.rm=T), se=sd/sqrt(length(sample)),
          min=min(read.total), max=max(read.total)) #use this to average across all samples 

# Did the no. of fragments per treatment differ? 
fragments <- data.frame(rowSums(counts.ts)) %>% 
                  dplyr::rename(read.total = 1) %>% 
                  rownames_to_column(var="sample") %>%
  left_join(sample.info %>% dplyr::select(Sample, Treatment, Tank), by=c("sample"="Sample"))

ggplot(fragments, aes(x=Treatment, y=read.total)) + geom_boxplot()

hist(fragments$read.total)
shapiro.test(fragments$read.total)
anova(lm(read.total ~ Treatment, fragments))

# How many samples per treatment?
fragments %>% group_by(Treatment) %>% tally()
fragments %>% group_by(Treatment, Tank) %>% tally()
```

How many genes were identified in each sample?

```{r}
ggplotly(
ggplot(data = data.frame(rowSums(counts.t != 0, na.rm=TRUE)) %>% 
                  dplyr::rename(count.total = 1) %>% 
                  rownames_to_column(var="sample")) +
           geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total # genes by sample") + 
             theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```


Use foa.plots to visualize data a bit: 

In how many samples does each gene occur? 
  - The **first four plots portray the gene’ frequency of occurrence among samples** in a number of different ways – either as an empirical cumulative distribution function (ECDF) of gene occurrence or as a histogram of gene occurrence. 

What is the mean abundance of each gene when it occurs (not averaging zeros for samples where it is absent)? 
  - The **fifth plot is an ECDF of gene mean abundance.** X-axis is samples, ranked from 1-n in terms of mean gene abundance. 
  
Is the mean abundance of genes correlated with the number of samples they occur in? 
  - The **sixth plot is a scatter plot of frequency of occurrence against mean abundance**. Is there any apparent relationship between the two? Are the widespread genes also generally more abundant? Are there many widespread genes that occur at low abundance? Conversely, are there genes much less widespread, but abundant where they occur?

Is the total abundance of gene in a sample correlated with the number of gene in a sample? To answer this question, first it is instructive to look at the number of gene per sample. 
  - The **eighth plot depicts the ECDF of sample richness.** Are there any interesting patterns? For example, do most samples support an average number of gene, while only a few samples supporting either very few or very many gene? Or is the pattern different?

Second, what is the pattern in the distribution of sample total abundance? 
  - The **ninth plot is the ECDF of total sample abundance.** How does it compare to the ECDF of sample richness?

Finally, to answer the question on the relation between total abundance and number of gene/sample ...
  - The **last plot is a scatter plot of the two variables.** Is there is relationship between the _number of genes per sample and the total abundance?_ Do gene-rich samples generally have a greater total abundance of those genes as well? 

```{r}
# note: you'll need to press return in the console for all plots 
#foa.plots(counts.ts)
```

Merge sample key info to count data, then sort, and generate heat map for initial inspection by treatment 

```{r}
# IMPORTANT NOTE: to use data for ALL genes, use object "counts.t". To remove low-frequency genes, use object "counts.ts" 

# merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 
counts.tk <- merge(x=sample.info[,c("Sample", "Tank", "Treatment", "Treatment_Tank")], by.x="Sample", y=counts.ts, by.y="row.names") %>% 
  arrange(Treatment, Tank)  %>% column_to_rownames(var="Sample") %>% droplevels()

head(counts.tk[1:24]) #check out results of merge/arrange
#counts.tk %>% dplyr::select(starts_with("evm")) #this is code to get only the gene columns 

```

Reformat for DESeq, ensure correct sample order for 

NOTE: It is absolutely critical that the **columns of the count matrix** and the **rows of the column data (information about samples)** are in the same order. DESeq2 will not make guesses as to which column of the count matrix belongs to which row of the column data, these must be provided to DESeq2 already in consistent order.

```{r}
all(rownames(counts.tk) == counts.tk %>% dplyr::select(starts_with("evm")) %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
```
If using counts which included multimapped counted fractionally, need to round to the nearest integer for DESeq2. 

```{r}
#counts.tk <- counts.tk %>% mutate_if(is.numeric, round)
```

Generate DESeq datasets with various treatment comparisons  

```{r}
dds.pH <- DESeqDataSetFromMatrix(countData = counts.tk[,grepl("g", colnames(counts.tk))] %>% t(),
                              colData = counts.tk[,"Treatment", drop=FALSE] ,
                              design = ~ Treatment)

dds.tank <- DESeqDataSetFromMatrix(countData = counts.tk[,grepl("g", colnames(counts.tk))] %>% t(),
                              colData = counts.tk[,"Tank", drop=FALSE] ,
                              design = ~ Tank)

dds.treat.tank <- DESeqDataSetFromMatrix(countData = counts.tk[,grepl("g", colnames(counts.tk))] %>% t(),
                              colData = counts.tk[,"Treatment_Tank", drop=FALSE] ,
                              design = ~ Treatment_Tank)
```

Transform data for visualization 

- Here we transform counts using a variance stabilizing transformation (VST), since we have many samples. 
- Here we use `blind=FALSE` b/c we are interested in differences explained by experimental design, and may wish to use this transformed data in downstream analyses. 

```{r}
vsd.pH <- varianceStabilizingTransformation(dds.pH, blind=FALSE)
vsd.tank <- varianceStabilizingTransformation(dds.tank, blind=FALSE)
vsd.treat.tank <- varianceStabilizingTransformation(dds.treat.tank, blind=FALSE)

# # Check out effect of VSD 
# meanSdPlot(assay(vsd.pH)) #check out normalized count SD by ranked mean - red line should be ~flat 
# meanSdPlot(assay(normTransform(dds.pH))) #this is variance after a simple log2(n+1) transformation 

# Save vsd.pH object for late ruse 
save(vsd.pH, file = "../results/deseq2/vsd.pH")
```

Generate heat maps before & after transformation  

NOTE: scale="column" b/c range of counts is so huge, so counts have been scaled 

```{r}
pheatmap(data.matrix(counts.tk %>% dplyr::select(starts_with("g"))), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, 
                     show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE, 
         annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")),
                     scale="column", #color=c("dodgerblue3", "goldenrod1"), 
                     main = "RKC gene counts - not transformed", annotation_row=counts.tk[,c("Treatment"), drop=FALSE])

# extract treatment info from VSD transformation 

# # generate heatmap from VSD counts 
# pheatmap(t(assay(vsd.pH)), Rowv=NA, Colv=NA, na.rm = TRUE, xlab = NA, scale="column", 
#                      show_colnames =FALSE, cluster_cols = FALSE, cluster_rows = TRUE, 
#          annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")),
#          annotation_row=as.data.frame(colData(vsd.pH)[,c("Treatment"), drop=FALSE]), 
#          main = "RKC gene counts - VSD-transformed")
```

Save variance-stabilization normalized gene counts x sample matrix, annotated for later use 

```{r}
counts.trans <- assay(vsd.pH) %>% t() %>% as.matrix() %>% as.data.frame() %>% 
  rownames_to_column(var = "Sample") %>% 
  left_join(sample.info[,c("Sample", "Treatment", "Tank", "Treatment_Tank")]) %>%
  column_to_rownames(var = "Sample") %>% 
  dplyr::select(Treatment, Tank, Treatment_Tank, starts_with("g")) %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
save(counts.trans, file = "../results/bowtie/counts.trans")
```

Visualize sample clustering via PCA (after transformation)

To view the the guts of the PlotPCA function in DESeq2, execute the next code chunk. It reveals that, by default, it only uses the top 500 most influential genes to plot the PCA! 

```{r}
#getMethod("plotPCA","DESeqTransform")
```

Generate PCA with points color coded by pH Treatment with various number of top genes to use for principal components, selected by highest row variance  
NOTE: Hover over points to see the sample numbers

```{r}
ggplotly(
  plotPCA(vsd.pH, intgroup="Treatment", ntop=500) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 500 genes") + 
    geom_point(size=3, aes(text=colnames(vsd.pH))) + 
  scale_color_manual(values=c("Ambient"="#2c7bb6", "Moderate"="#fdae61", "Low"="#d7191c")) +
    theme_minimal()+ stat_ellipse() , tooltip = "text")

ggplotly(
  plotPCA(vsd.pH, intgroup="Treatment", ntop=5000) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 5k genes") + 
    geom_point(size=3, aes(text=colnames(vsd.pH))) + 
  scale_color_manual(values=c("Ambient"="#2c7bb6", "Moderate"="#fdae61", "Low"="#d7191c")) +
    theme_minimal()+ stat_ellipse() , tooltip = "text")

ggplotly(
  plotPCA(vsd.pH, intgroup="Treatment", ntop=50000) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\ntop 50k genes") + 
    geom_point(size=3, aes(text=colnames(vsd.pH))) + 
  scale_color_manual(values=c("Ambient"="#2c7bb6", "Moderate"="#fdae61", "Low"="#d7191c")) +
    theme_minimal()+ stat_ellipse() , tooltip = "text")

ggplotly(
  plotPCA(vsd.pH, intgroup="Treatment", ntop=nrow(assay(vsd.pH))) + 
           ggtitle("PCA by Treatment (var-stabilizing transformed)\nall genes") + 
    geom_point(size=3, aes(text=colnames(vsd.pH))) + 
  scale_color_manual(values=c("Ambient"="#2c7bb6", "Moderate"="#fdae61", "Low"="#d7191c")) +
    theme_minimal()+ stat_ellipse() , tooltip = "text")


# PCA with points color coded by tank 
ggplotly(
  plotPCA(vsd.tank, intgroup="Tank") + 
           ggtitle("PCA by Tank (var-stabilizing transformed)") + 
    geom_point(size=3, aes(text=colnames(vsd.tank))) + 
   theme_minimal(), tooltip = "text")

# PCA with points color coded by treatment + tank 
#ggplotly(
  plotPCA(vsd.treat.tank, intgroup="Treatment_Tank") + 
           ggtitle("PCA by Treatment & Tank (var-stabilizing transformed)") + 
    geom_point(size=3, aes(text=colnames(vsd.treat.tank))) + 
   theme_minimal()#, tooltip = "text")

# Ambient pH (8.0) tanks:  1,2,7,10,11
# Moderate pH (7.8) tanks: 3,4,9,20
# Low pH (7.5) tanks: 5,13,15,16,18

PCA.data.pH <- plotPCA(vsd.pH, intgroup=c("Treatment"), returnData=TRUE)
save(PCA.data.pH, file="../results/PCA.data.pH")
```

Run PCA using prcomp, etc. to identify significant PC axes and examine potential tank effects 

```{r}
#pca.princomp <- prcomp(cov(assay(vsd.pH)), scale=F) #scale=F for variance-covariance matrix
pca.princomp <- prcomp(t(assay(vsd.pH))) # using the same code that's under the hood of PlotPCA but using all genes 
pca.eigenval(pca.princomp) #The Proporation of Variance = %variance 
pc.percent <- pca.eigenval(pca.princomp)[2,1:5]*100
pc.percent[1:2] %>% sum()
screeplot(pca.princomp, bstick=TRUE) #looks like PC 1 & 2 are significant 

# tab.expr <- data.frame(sample.id = colnames(assay(vsd.pH)),
#     EV1.all = pca.princomp$rotation[,1],    # the first eigenvector
#     EV2.all = pca.princomp$rotation[,2],    # the second eigenvector
#     EV3.all = pca.princomp$rotation[,3],    # the third eigenvector
#     EV4.all = pca.princomp$rotation[,4],    # the fourth eigenvector
#     EV5.all = pca.princomp$rotation[,5],    # the fifth eigenvector
#     stringsAsFactors = FALSE)
# shapiro.test(pca.princomp$rotation) #multivariate normality not met, hmmmm

tab.expr <- data.frame(sample.id = colnames(assay(vsd.pH)),
    EV1.all = pca.princomp$x[,1],    # the first eigenvector
    EV2.all = pca.princomp$x[,2],    # the second eigenvector
    EV3.all = pca.princomp$x[,3],    # the third eigenvector
    EV4.all = pca.princomp$x[,4],    # the fourth eigenvector
    EV5.all = pca.princomp$x[,5],    # the fifth eigenvector
    stringsAsFactors = FALSE)
shapiro.test(pca.princomp$x) #multivariate normality not met, hmmmm

tab.expr.annot <- left_join(tab.expr, sample.info, by=c("sample.id"="Sample")) %>% droplevels()

# GGSCATTER with ellipses and stars - PC1 PC2
(fig2 <- ggscatter(tab.expr.annot %>% 
            mutate(Treatment=str_replace(Treatment, "Ambient", "Ambient (pH 8.0)"),
                   Treatment=str_replace(Treatment, "Moderate", "Moderate (pH 7.8)"),
                   Treatment=str_replace(Treatment, "Low", "Severe (pH 7.6)")), 
          x="EV1.all", y="EV2.all", col="Treatment", size=2, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
           palette = c(`Ambient (pH 8.0)`="#2c7bb6", `Moderate (pH 7.8)`="#fdae61", `Severe (pH 7.6)`="#d7191c")) +
  theme_minimal() + ggtitle("Global gene expression, PC1xPC2") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC2 (", round(pc.percent[2], digits = 1), "%)", sep="")) + 
  labs(color="OA Treatment") + 
  labs(fill="OA Treatment") + 
  theme(legend.position = "right", plot.title = element_text(size=9), axis.title = element_text(size=8), 
        legend.title = element_text(size=8), legend.text = element_text(size=8)) + 
  # scale_color_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  # scale_fill_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  guides(colour = guide_legend(override.aes = list(size=3, linetype="blank")))) 

# GGSCATTER with ellipses and stars - PC1 PC3
ggscatter(tab.expr.annot, x="EV1.all", y="EV3.all",
          col="Treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
           palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
  theme_minimal() + ggtitle("Global gene expression, PC1xPC3") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC3 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  # scale_color_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  # scale_fill_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 

# GGSCATTER with ellipses and stars - PC1 PC4
ggscatter(tab.expr.annot, x="EV1.all", y="EV4.all",
          col="Treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
           palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
  theme_minimal() + ggtitle("Global gene expression, PC1xPC4") + 
  xlab(paste("PC1 (", round(pc.percent[1], digits = 1), "%)", sep="")) + 
  ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  # scale_color_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  # scale_fill_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 


# GGSCATTER with ellipses and stars - PC3 PC4
ggscatter(tab.expr.annot, x="EV3.all", y="EV4.all",
          col="Treatment", size=2.5, alpha=0.85, ellipse = TRUE, star.plot = TRUE, 
           palette = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
  theme_minimal() + ggtitle("Global gene expression, PC3xPC4") + 
  xlab(paste("PC1 (", round(pc.percent[3], digits = 1), "%)", sep="")) + 
  ylab(paste("PC4 (", round(pc.percent[4], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + 
  # scale_color_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  # scale_fill_manual(name="pCO2 Treatment", labels=c("8.1", "7.8", "7.6"), values=c("8.1"="#2c7bb6", "7.8"="#fdae61", "7.6"="#d7191c")) +
  guides(colour = guide_legend(override.aes = list(size=3.5, linetype="blank"))) 

```
Save Figure 2

```{r}
tiff(file = "../manuscript/PLOS-Climate-submission/Revision/Final/Figures-correct-dims/Fig2.tiff", width = 13.2, height = 8.36, units = "cm", res = 600)
print(fig2)
dev.off()
```

```{r}

# Examine potential tank effects for each treatment 
pc.axes <- names(tab.expr.annot[2:6])
x <- c(1,1,1,1,2,2,2,3,3,4)
y <- c(2,3,4,5,3,4,5,4,5,5)

pc.plots <- vector("list", length(x))
for (i in 1:length(x)) {
pc.plots[[i]] <- 
  #ggplotly(
    ggplot(
  tab.expr.annot %>% 
  mutate(Tank_n=as.numeric(as.character(Tank))) %>% 
  mutate_at("Tank_n", function(x) {
    case_when(
      x == 1 ~ "A",
      x == 5 ~ "A",
      x == 3 ~ "A",
      x == 2 ~ "B",
      x == 13 ~ "B",
      x == 4 ~ "B",
      x == 7 ~ "C",
      x == 15 ~ "C",
      x == 9 ~ "C",
      x == 10 ~ "D",
      x == 16 ~ "D",
      x == 20 ~ "D",
      x == 11 ~ "E",
      x == 18 ~ "E")}),
  aes(x=!!as.symbol(pc.axes[x[i]]),y=!!as.symbol(pc.axes[y[i]]))) + 
  geom_point(aes(col=Treatment, shape=Tank_n, fill=Treatment), size=3.5, alpha=0.85) + 
  theme_minimal() + ggtitle(paste("PC", x[i], " x ", "PC", y[i], sep = "")) + 
  xlab(paste("PC", x[i], " (", round(pc.percent[x[i]], digits = 1), "%)", sep="")) + 
  ylab(paste("PC", y[i], " (", round(pc.percent[y[i]], digits = 1), "%)", sep="")) + 
  theme(legend.position = "right") + scale_shape_manual(values=seq(0,5))#)
}
pc.plots
```

### Heatmap of the sample-to-sample distances
Another use of the transformed data is sample clustering. Here, we apply the dist function to the transpose of the transformed count matrix to get sample-to-sample distances.

A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.

```{r}
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

sampleDistss <- dist(t(assay(vsd.pH)))
sampleDistMatrixs <- as.matrix(sampleDistss)

# Here we only show pH treatment 
rownames(sampleDistMatrixs) <- vsd.pH$Treatment #set row names 
colnames(sampleDistMatrixs) <- NULL
pheatmap(sampleDistMatrixs,
         clustering_distance_rows=sampleDistss,
         clustering_distance_cols=sampleDistss,
         col=rev(colors), fontsize = 6)


# Here we also show tank number 
sampleDistss <- dist(t(assay(vsd.treat.tank)))
sampleDistMatrixs <- as.matrix(sampleDistss)

rownames(sampleDistMatrixs) <- vsd.treat.tank$Treatment_Tank #set row names 
colnames(sampleDistMatrixs) <- NULL
pheatmap(sampleDistMatrixs,
         clustering_distance_rows=sampleDistss,
         clustering_distance_cols=sampleDistss,
         col=rev(colors), fontsize = 6)

```

## perMANOVA - multivariate analysis of variance 

```{r}
# Effect of treatment 
vsd.t <- t(assay(vsd.pH)) %>% as.data.frame()
perm <- adonis(vsd.t ~ Treatment, data=sample.info %>% remove_rownames() %>% column_to_rownames("Sample"), permutations = 1000, method="bray")

# Pairwise comparisons 
vsd.t <- sample.info %>% dplyr::select(Sample, Treatment, Tank) %>% left_join(t(assay(vsd.pH)) %>% as.data.frame() %>% rownames_to_column("Sample"))
pairwise.adonis(vsd.t[,-1:-3], vsd.t$Treatment, perm = 1000)

```

## How many unique genes were measured per treatment. This takes the mean of each gene by group, and counts up those that have mean >10. 
# Warning: this takes forever 

```{r}
# # Counts before filtering 
# sample.info %>% select(Sample, Treatment) %>% left_join(counts.t %>% rownames_to_column("Sample")) %>% 
#   group_by(Treatment) %>% summarise_if(is.numeric, "mean") %>% mutate(genes=rowSums(select_if(., is.numeric)>10)) %>% select(Treatment, genes)

# # Counts after filtering
# sample.info %>% select(Sample, Treatment) %>% left_join(counts.tk %>% rownames_to_column("Sample")) %>% 
#   select(-Sample, -Tank, -Treatment_Tank) %>% group_by(Treatment) %>% summarise_all("mean") %>% mutate(genes=rowSums(is.numeric(.)>10)) %>% select(Treatment, genes)
```

## Differential Expression Analysis - multifactor design 

_This is the core DESeq2 analysis!_ 

Run the function `DESeq` to assess differential expression 

```{r}
dds.DESeq.pH <- DESeq(dds.pH) 
```

Any DEGs between pH treatment? 

```{r}
#colData(dds.DESeq.pH)
print("Comparison: Ambient pH vs. Moderate pH")
summary(res.pco2.AM <- results(dds.DESeq.pH, contrast=c("Treatment", "Ambient", "Moderate"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, Ambient vs. Moderate:",  sum(res.pco2.AM$padj < 0.05, na.rm=TRUE))

print("Comparison: Ambient pH vs. Low pH")
summary(res.pco2.AL <- results(dds.DESeq.pH, contrast=c("Treatment", "Ambient", "Low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, Ambient vs. Low:",  sum(res.pco2.AL$padj < 0.05, na.rm=TRUE))

print("Comparison: Moderate pH vs. Low pH")
summary(res.pco2.ML <- results(dds.DESeq.pH, contrast=c("Treatment", "Moderate", "Low"), alpha=0.05))
paste("No. of genes differentially expressed (padj<0.05) by pH, Moderate vs. Low:",  sum(res.pco2.ML$padj < 0.05, na.rm=TRUE))

# Number of genes analyzed 
nrow(res.pco2.AM)
save(res.pco2.AM, file = "../results/deseq2/res.pco2.AM")
save(res.pco2.AL, file = "../results/deseq2/res.pco2.AL")
save(res.pco2.ML, file = "../results/deseq2/res.pco2.ML")
```

How many DEGs in total?

```{r}
length(unique(c(diffex.AM %>% rownames(), diffex.AL %>% rownames(), diffex.ML %>% rownames())))
```

What % of all examined genes are DEGs 

```{r}
100*(length(unique(c(diffex.AM %>% rownames(), diffex.AL %>% rownames(), diffex.ML %>% rownames()))))/(res.pco2.AL %>% nrow())
```

Subset the DESeq results for only those statistically sign. ones 

```{r}
diffex.AM <- subset(res.pco2.AM, padj < 0.05)
diffex.AL <- subset(res.pco2.AL, padj < 0.05)
diffex.ML <- subset(res.pco2.ML, padj < 0.05)

save(diffex.AM, file = "../results/deseq2/diffex.AM")
save(diffex.AL, file = "../results/deseq2/diffex.AL")
save(diffex.ML, file = "../results/deseq2/diffex.ML")
```

Create a single heatmap with all DEGs among any treatment 

```{r}
# Generate counts matrix With DEGs among any pH treatment
diffex.pH.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% unique(c(rownames(diffex.AM),rownames(diffex.AL),rownames(diffex.ML))))

# Create a dataframe to annotate heatmap with treatments 
dds.pH.df <- sample.info[match(colnames(diffex.pH.counts), sample.info$Sample),c("Sample", "Treatment", "Tank")] %>%
  remove_rownames() %>% column_to_rownames(var = "Sample")

# Make sure treatment order is correct
all(colnames(diffex.pH.counts) == rownames(dds.pH.df)) #double check that samples are still in same order 

# All DEGs among any pH treatment 
pheatmap(diffex.pH.counts, cluster_rows=TRUE, cluster_cols=FALSE,
         show_rownames=FALSE, na.rm=TRUE, annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")),
         scale="row", main = "Differentially expressed genes (DEGs) among any pH treatment", 
         annotation_col=dds.pH.df[,"Treatment", drop=FALSE], fontsize = 8)

# # another way to do a heatmap 
# heatmap(diffex.pH.counts, Colv = NA, scale="row", ColSideColors = brewer.pal(3, "Set1")[dds.pH.df$Treatment])
```

Generate pairwise comparison heatmaps 

```{r}
# Create count matrices containing DEGs for each pairwise treatment comparison, rows (genes) ordered by pvalue, 

diffex.AM.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% rownames(diffex.AM)) %>% as.data.frame() %>%
  dplyr::select((sample.info %>% arrange(Treatment) %>% 
                   filter(Treatment=="Ambient" | Treatment== "Moderate"))$Sample) %>% 
  rownames_to_column(var = "gene") %>% arrange(match(gene, c(as.data.frame(diffex.AM) %>% arrange(padj) %>% rownames()))) %>%
  column_to_rownames(var = "gene") %>% as.matrix()
  
diffex.AL.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% rownames(diffex.AL)) %>% as.data.frame() %>%
  dplyr::select((sample.info %>% arrange(Treatment) %>% 
                   filter(Treatment=="Ambient" | Treatment== "Low"))$Sample) %>% 
  rownames_to_column(var = "gene") %>% arrange(match(gene, c(as.data.frame(diffex.AL) %>% arrange(padj) %>% rownames()))) %>% 
  column_to_rownames(var = "gene") %>% as.matrix()

diffex.ML.counts <- subset(assay(vsd.pH), rownames(dds.DESeq.pH) %in% rownames(diffex.ML)) %>% as.data.frame() %>%
  dplyr::select((sample.info %>% arrange(desc(Treatment)) %>% 
                   filter(Treatment=="Moderate" | Treatment== "Low"))$Sample) %>% 
  rownames_to_column(var = "gene") %>%
  arrange(match(gene, c(as.data.frame(diffex.ML) %>% arrange(padj) %>% rownames()))) %>% map_df(rev) %>% 
  column_to_rownames(var = "gene") %>% as.matrix()

# Define gradient range, breaks
my_colours <- rev(c("springgreen4", "gray95", "slateblue4"))
my_colours <- rev(c("springgreen4", "gray95", "black"))
my_colours <- rev(c("green4", "gray95", "black"))
my_colours <- rev(c("red3", "gray95", "mediumblue"))
my_gradient <- colorRampPalette(my_colours)(length(breaksList))

# NOTE, the row-wise scale produces Z-scores (x-mean/sd)

# Draw heatmaps 
#pdf(file = "../results/heatmap-AM.pdf", width = 6.15, height = 3.15)
tiff(file = "../results/heatmap-AM.tiff", width = 6.15, height = 3.15, units = "in", res = 600)
(heat.am <- pheatmap(diffex.AM.counts, cluster_cols = FALSE, cluster_rows=TRUE, 
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, scale="row", 
         #main = "DEGs among pH 8.0 (ambient) & pH 7.8 (moderate), sorted by p-value",
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 13, cutree_rows = 2,  border_color=F, 
         annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61")),
         annotation_col=dds.pH.df[colnames(diffex.AM.counts),"Treatment", drop=FALSE],
         annotation_legend = FALSE, annotation_names_col= FALSE,
         #color = colorRampPalette(rev(brewer.pal(n = 5, name = "BrBG")))(length(breaksList)),
         color = my_gradient,
         breaks = breaksList))
dev.off()

#pdf(file = "../results/heatmap-AL.pdf", width = 6.15, height = 3.15)
tiff(file = "../results/heatmap-AL.tiff", width = 6.15, height = 3.15, units = "in", res = 600)
(heat.al <- pheatmap(diffex.AL.counts, cluster_cols = FALSE, cluster_rows=TRUE, 
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, scale="row", 
         #main = "DEGs among pH 8.0 (ambient) & pH 7.5 (low), sorted by p-value",
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 13, cutree_rows = 2, border_color=F, 
         annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Low="#d7191c")),
         annotation_col=dds.pH.df[colnames(diffex.AL.counts),"Treatment", drop=FALSE],
         annotation_legend = FALSE, annotation_names_col= FALSE,
         #color = colorRampPalette(rev(brewer.pal(n = 5, name = "BrBG")))(length(breaksList)),
         color = my_gradient,
         breaks = breaksList))
dev.off()

#pdf(file = "../results/heatmap-ML.pdf", width = 6.15, height = 3.15)
tiff(file = "../results/heatmap-ML.tiff", width = 6.15, height = 3.15, units = "in", res = 600)
(heat.ml <- pheatmap(diffex.ML.counts, cluster_cols = FALSE, cluster_rows=TRUE, 
         show_rownames=FALSE, show_colnames = FALSE, na.rm=TRUE, scale="row", 
         #main = "DEGs among pH 7.8 (moderate) & pH 7.5 (low), sorted by p-value",
         treeheight_row = 0, treeheight_col = 0,
         fontsize = 13, cutree_rows = 2, border_color=F, 
         annotation_colors = list(Treatment=c(Moderate="#fdae61", Low="#d7191c")),
         annotation_col=dds.pH.df[colnames(diffex.ML.counts),"Treatment", drop=FALSE],
         annotation_legend = FALSE, annotation_names_col= FALSE,
         #color = colorRampPalette(rev(brewer.pal(n = 5, name = "BrBG")))(length(breaksList)),
         color = my_gradient,
         breaks = breaksList))
dev.off()

save(diffex.AM.counts, file = "../results/deseq2/diffex.AM.counts")
save(diffex.AL.counts, file = "../results/deseq2/diffex.AL.counts")
save(diffex.ML.counts, file = "../results/deseq2/diffex.ML.counts")
```

Volcano plots 

```{r}
# Set axis ticks for all volcano plots
xtick=seq(-6,4, by=2)
ytick=seq(0,15, by=5)
```

```{r}
# ----------- AMBIENT VS MODERATE 
topT <- as.data.frame(res.pco2.AM)

par(mar=c(2.5,2.5,1.5,1), cex.main=.9, cex.axis=.8, cex.lab=.8)

#Adjusted P values (FDR Q values)
with(topT, plot(-1*log2FoldChange, -log10(padj), pch=16, cex=0.45, main="", xlab="", ylab="", xlim=c(-7,4), ylim=c(0,15), xaxt="n", yaxt="n"))
#axis(side=1, at=xtick, labels=F)
#text(par("usr")[3], x=xtick, labels = xtick, pos = 1, xpd = TRUE, cex = 1)
axis(side=2, at=ytick, labels=F)
text(par("usr")[1], ytick, labels = ytick, pos = 2, xpd = TRUE, cex = 1)
title(main="(A) Ambient vs. Moderate OA", adj=0.01, line=0.4, font.main=1, cex.main=1.3)
title(xlab=NA, ylab=bquote(~-log[10]~P-value), line=1.3, cex.lab=1.15)

# plot genes downregulated in OA (i.e. upregulated in ambient)
# with LFC >1
with(subset(topT, padj<0.05 & -1*log2FoldChange< -0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#2c7bb6", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange< 0  & -1*log2FoldChange> -0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d1e5f0", cex=0.55))

# plot genes upregulated in moderate OA 
with(subset(topT, padj<0.05 & -1*log2FoldChange>0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fdae61", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange>0 & -1*log2FoldChange<0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fddbc7", cex=0.55))


#Add lines for absolute FC>0.5 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-0.5, col="black", lty=4, lwd=2.0)
abline(v=0.5, col="black", lty=4, lwd=2.0)
abline(h=-log10(max(topT$padj[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=2.0)
volcano.am <- recordPlot()

# pdf(file = "../results/volcano-AM.pdf", width = 6.25, height = 3.75)
# print(volcano.am)
# dev.off()

tiff(file = "../results/volcano-AM.tiff", width = 6.25, height = 3.75, units = "in", res = 600)
print(volcano.am)
dev.off()
```

```{r}
# ----------- AMBIENT VS LOW 
topT <- as.data.frame(res.pco2.AL)

par(mar=c(2.5,2.5,1.5,1), cex.main=.9, cex.axis=.8, cex.lab=.8)

#Adjusted P values (FDR Q values)
with(topT, plot(-1*log2FoldChange, -log10(padj), pch=16, cex=0.45, main="", xlab="", ylab="", xlim=c(-7,4), ylim=c(0,15), xaxt="n", yaxt="n"))
#axis(side=1, at=xtick, labels=F)
#text(par("usr")[3], x=xtick, labels = xtick, pos = 1, xpd = TRUE, cex = 1)
axis(side=2, at=ytick, labels=F)
text(par("usr")[1], ytick, labels = ytick, pos = 2, xpd = TRUE, cex = 1)
title(main="(B) Ambient vs. Severe OA", adj=0.01, line=0.4, font.main=1, cex.main=1.3)
title(xlab=NA, ylab=bquote(~-log[10]~P-value), line=1.3, cex.lab=1.15)

# plot genes upregulated in ambient 
# with LFC >1
with(subset(topT, padj<0.05 & -1*log2FoldChange< -0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#2c7bb6", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange< 0  & -1*log2FoldChange> -0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d1e5f0", cex=0.55))

# plot genes upregulated in severe OA 
with(subset(topT, padj<0.05 & -1*log2FoldChange>0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d7191c", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange>0 & -1*log2FoldChange<0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="pink2", cex=0.55))

#Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-0.5, col="black", lty=4, lwd=1.9)
abline(v=0.5, col="black", lty=4, lwd=1.9)
abline(h=-log10(max(topT$padj[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=1.9)
volcano.al <- recordPlot()

# pdf(file = "../results/volcano-AL.pdf", width = 6.25, height = 3.75)
# print(volcano.al)
# dev.off()

tiff(file = "../results/volcano-AL.tiff", width = 6.25, height = 3.75, units = "in", res = 600)
print(volcano.al)
dev.off()
```

```{r}
# ----------- MODERATE VS LOW 
topT <- as.data.frame(res.pco2.ML)

par(mar=c(2.5,2.5,1.5,1), cex.main=.9, cex.axis=.8, cex.lab=.8)
#Adjusted P values (FDR Q values)
with(topT, plot(-1*log2FoldChange, -log10(padj), pch=16, cex=0.45, main="", xlab="", ylab="", xlim=c(-7,4), ylim=c(0,15), xaxt="n", yaxt="n"))
axis(side=1, at=xtick, labels=F)
text(par("usr")[3], x=xtick, labels = xtick, pos = 1, xpd = TRUE, cex = 1)
axis(side=2, at=ytick, labels=F)
text(par("usr")[1], ytick, labels = ytick, pos = 2, xpd = TRUE, cex = 1)
title(main="(C) Moderate OA vs. Severe OA", adj=0.01, line=0.4, font.main=1, cex.main=1.3)
title(xlab=bquote(~Log[2]~fold~change), ylab=bquote(~-log[10]~P-value), line=1.3, cex.lab=1.15)
# plot genes upregulated in moderate OA 
# with LFC >1
with(subset(topT, padj<0.05 & -1*log2FoldChange< -0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fdae61", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange< 0  & -1*log2FoldChange> -0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#fddbc7", cex=0.55))

# plot genes upregulated in severe OA 
with(subset(topT, padj<0.05 & -1*log2FoldChange>0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="#d7191c", cex=0.55))
with(subset(topT, padj<0.05 & -1*log2FoldChange>0 & -1*log2FoldChange<0.5), points(-1*log2FoldChange, -log10(padj), pch=16, col="pink2", cex=0.55))


#Add lines for absolute FC>2 and P-value cut-off at FDR Q<0.05
abline(v=0, col="black", lty=3, lwd=1.0)
abline(v=-0.5, col="black", lty=4, lwd=2.0)
abline(v=0.5, col="black", lty=4, lwd=2.0)
abline(h=-log10(max(topT$padj[topT$padj<0.05], na.rm=TRUE)), col="black", lty=4, lwd=2.0)
volcano.ml <- recordPlot()

# pdf(file = "../results/volcano-ML.pdf", width = 6.25, height = 3.85)
# print(volcano.ml)
# dev.off()
# 
tiff(file = "../results/volcano-ML.tiff", width = 6.25, height = 3.85, units = "in", res = 600)
print(volcano.ml)
dev.off()
```

Question - are the moderate vs. low DEGs unique? or are they similar to ambient? 
-- answer - not super unique 

```{r}
diffex.ML %>% rownames() %>% length() #51 DEGs between moderate & low pH 
Reduce(setdiff, list(c(diffex.ML %>% rownames()), c(diffex.AL %>% rownames()))) %>% length() # 32 of which are also different among ambient & low pH 
Reduce(setdiff, list(c(diffex.ML %>% rownames()), c(diffex.AM %>% rownames()))) %>% length() # 65 of which are also different among ambient & moderate pH
```

Venn Diagram of DEGs by pairwise comparison 

```{r}
#install.packages("VennDiagram")
library(VennDiagram)

x <- list(
  AvsM = row.names(diffex.AM), 
  AvsL = row.names(diffex.AL), 
  MvsL = row.names(diffex.ML)
  )

lengths(x)

display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}


overrideTriple=T
display_venn(
        x,
        category.names = c("Ambient vs.\nModerate OA" , "Ambient vs.\nSevere OA" , "Moderate OA vs.\nSevere OA"),
        # Circles
       lwd = 0,
       lty = 'blank',
       fill = c("gray25", "gray50", "gray75"),
       alpha=c(0.5, 0.5, 0.5),
        #Numbers
        cex = 1,
        fontface = "italic",
        # Set names
        cat.cex = 0.83,
        cat.fontface = "bold",
        cat.default.pos = "outer" ,
        cat.dist = c(0.02, 0.02, 0.02)
)



# common among all pairwise comparisons 
length(Reduce(intersect, list(x[[1]], x[[2]], x[[3]])))

# Overlap between AvsM and AvsL, not including 1 common among all 
length(intersect(x[[1]], x[[2]]))-1

# Overlap between AvsM and MvsL, not including 1 common among all 
length(intersect(x[[1]], x[[3]]))-1

# Overlap between AvsS and MvsL, not including 1 common among all 
length(intersect(x[[2]], x[[3]]))-1

overrideTriple=T
venn.plot <- draw.triple.venn(area1 = as.vector(lengths(x))[1],
                 area2 = as.vector(lengths(x))[2],
                 area3 = as.vector(lengths(x))[3],
                 n12 = length(intersect(x[[1]], x[[2]])),
                 n23 = length(intersect(x[[1]], x[[3]])),
                 n13 = length(intersect(x[[2]], x[[3]])),
                 n123 = length(Reduce(intersect, list(x[[1]], x[[2]], x[[3]]))),
                 category = c("Ambient vs.\nModerate OA" , "Ambient vs.\nSevere OA" , "Moderate OA vs.\nSevere OA"),
                 fill=c("gray25", "gray50", "gray75"),
                 #lty="blank",
                 #cex=2,
                 #cat.cex=2,
                 cat.col="black")
grid.draw(venn.plot)
grid.newpage()

row.names(diffex.AM) %>% write_clip()
row.names(diffex.AL) %>% write_clip()
row.names(diffex.ML) %>% write_clip()
```

Save lists of annotated genes DEGs 
Here I merge the DEG lists with Uniprot/Swissprot annotations to get UniprotID and GO information. 

```{r}
diffex.AM.annot <- diffex.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  #left_join(P.plat.bestblast, by=c("gene"="id_gtf")) %>%
  left_join(P.camt.blast.GO %>% 
              dplyr::select(gene_id, start, end, saccver, spid, gene_uni,evalue, protein_names,
                            gene_ontology_biological_process, gene_ontology_i_ds), 
            by = c("gene"="gene_id")) %>% arrange(padj)
write.csv(diffex.AM.annot, file = "../results/DEGs.Amb-vs-Mod_annotated.csv", row.names = F, quote = T)

diffex.AL.annot <- diffex.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  #left_join(P.plat.bestblast, by=c("gene"="id_gtf")) %>%
  left_join(P.camt.blast.GO %>%  
              dplyr::select(gene_id, start, end, saccver, spid, gene_uni,evalue, protein_names,
                            gene_ontology_biological_process, gene_ontology_i_ds), 
            by = c("gene"="gene_id")) %>% arrange(padj)
write.csv(diffex.AL.annot, file = "../results/DEGs.Amb-vs-Low_annotated.csv", row.names = F, quote = T)

diffex.ML.annot <- diffex.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  #left_join(P.plat.bestblast, by=c("gene"="id_gtf")) %>%
  left_join(P.camt.blast.GO %>% 
              dplyr::select(gene_id, start, end, saccver, spid, gene_uni,evalue, protein_names,
                            gene_ontology_biological_process, gene_ontology_i_ds), 
            by = c("gene"="gene_id")) %>% arrange(padj)  
write.csv(diffex.ML.annot, file = "../results/DEGs.Mod-vs-Low_annotated.csv", row.names = F, quote = T)
  
# Octopamine receptor, SPID Q7JQF1 = g114647
# glutaminase, gene = g90416, spid=P49221
diffex.AL.annot %>% View()

# Plot genes of interest 
test <- diffex.AM.annot %>% as.data.frame() %>% filter(log2FoldChange<0) %>% rownames()
test <- (filter(diffex.AM.annot %>% as.data.frame(), grepl("GABA",protein_names)))$gene
test <- "g87351"

for (i in 1:length(test)) {
a <-  plotCounts(dds.DESeq.pH, gene=test[i], intgroup=c("Treatment"), returnData = TRUE) %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
print(ggplot(a %>% rownames_to_column("sample"),
       aes(x=Treatment, y=count, color=Treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=2) +
    theme_bw() +
    ggtitle(test[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
}
```

How many of the differentially expressed genes are transposable elements? 

```{r}
# After inspecting annotated DEGs, I determined that these strings will identify TEs in my dataset 
TE.genes <-  ("transposon|transposable|LINE|retrotransposable element|transposable element|mobile element jockey|Pol polyprotein")

print(paste("Percentage of Amb vs. Mod UPREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AM.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AM.annot %>% filter(log2FoldChange<0, !is.na(spid))), digits = 1)))
print(paste("Percentage of Amb vs. Mod DOWNREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AM.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AM.annot %>% filter(log2FoldChange>0, !is.na(spid))), digits = 1)))

print(paste("Percentage of Amb vs. Low UPREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AL.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AL.annot %>% filter(log2FoldChange<0, !is.na(spid))), digits = 1)))
print(paste("Percentage of Amb vs. Low DOWNREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.AL.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.AL.annot %>% filter(log2FoldChange>0, !is.na(spid))), digits = 1)))

print(paste("Percentage of Moderate vs. Low UPREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.ML.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.ML.annot %>% filter(log2FoldChange<0, !is.na(spid))), digits = 1)))
print(paste("Percentage of Moderate vs. Low DOWNREGULATED genes that are Transposable Elements:",
            round(100*nrow(diffex.ML.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))/nrow(diffex.ML.annot %>% filter(log2FoldChange>0, !is.na(spid))), digits = 1)))

```

Calculate average CV of TEs in upregulated / downregulated gene sets 

```{r}
TEs.AM.up <- (diffex.AM.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))$gene
TEs.AM.down <- (diffex.AM.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))$gene

TEs.AL.up <- (diffex.AL.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))$gene
TEs.AL.down <- (diffex.AL.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))$gene

TEs.ML.up <- (diffex.ML.annot %>% filter(log2FoldChange<0) %>% filter(str_detect(protein_names, TE.genes)))$gene
TEs.ML.down <- (diffex.ML.annot %>% filter(log2FoldChange>0) %>% filter(str_detect(protein_names, TE.genes)))$gene

genes.stats %>% 
  filter(gene %in% TEs.AM.up) %>%
  filter(Treatment %in% c("Ambient", "Moderate")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

genes.stats %>% 
  filter(gene %in% TEs.AL.up) %>%
  filter(Treatment %in% c("Ambient", "Low")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

genes.stats %>% 
  filter(gene %in% TEs.AM.down) %>%
  filter(Treatment %in% c("Ambient", "Moderate")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

genes.stats %>% 
  filter(gene %in% TEs.AL.down) %>%
  filter(Treatment %in% c("Ambient", "Low")) %>% 
  group_by(Treatment) %>% summarise(mean.cv = mean(cv))

```

## What % of all genes are TEs? 

```{r}
# What % are TEs across all genes? 
nrow(counts.annot.Pcamt %>% filter(str_detect(protein_names, TE.genes)))/nrow(counts.annot.Pcamt)

# What % are TEs across all annotated genes? 
nrow(counts.annot.Pcamt %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!is.na(SPID)))/nrow(counts.annot.Pcamt %>% filter(!is.na(SPID)))
```

## What % of genes are TEs by treatment? 

```{r}
# What % are TEs across all genes? 
nrow(counts.annot.Pcamt %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!grepl("E3 SUMO-protein ligase ZBED1",protein_names)))/nrow(counts.annot.Pcamt)

# What % are TEs across all annotated genes? 
nrow(counts.annot.Pcamt %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!is.na(SPID)))/nrow(counts.annot.Pcamt %>% filter(!is.na(SPID)))
```
```{r}
counts.annot.Pcamt %>% filter(str_detect(protein_names, TE.genes)) %>% filter(!is.na(SPID)) %>% filter(!grepl("E3 SUMO-protein ligase ZBED1",protein_names)) %>% arrange(protein_names) %>% select(protein_names) %>% distinct() %>% View()
```

##  DAVID enrichment analysis

It is a little clunky, but the DAVID functional analysis tool is one of the best ways to perform enrichment analysis (that I have found to date). To use, you copy a list of Uniprot IDs (or other gene identifier) to clipboard, then paste it into their tool. You then do the same for all genes in the analysis, providing an accurate background list of genes.  Here is code to copy all genes, then upregulated and downregulated DEGs identified by each contrast.

```{r}
# BACKGROUND - all genes submitted to DESeq2 analysis 
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()


# -------
# DEGs between Ambient & Moderate - DOWNREGULATED IN MODERATE OA VS. AMBIENT
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(padj<0.05) %>% filter(log2FoldChange>0.05) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between Ambient & Moderate - UPREGULATED IN MODERATE OA VS. AMBIENT
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(padj<0.05) %>% filter(log2FoldChange<(-0.05)) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# -------
# DEGs between Ambient & Low - DOWNREGULATED IN SEVERE OA VS. AMBIENT
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(padj<0.05) %>% filter(log2FoldChange>0.5) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between Ambient & Low - UPREGUALTED IN SEVERE OA VS. AMBIENT
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(padj<0.05) %>% filter(log2FoldChange<(-0.05)) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# -------
# DEGs between Moderate & Low - DOWNREGULATED IN SEVERE OA VS MODERATE OA 
res.pco2.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(padj<0.05) %>% filter(log2FoldChange>0.05) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between Moderate & Low - UPREGUALTED IN SEVERE OA VS MODERATE OA
res.pco2.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% filter(padj<0.05) %>% filter(log2FoldChange<(-0.05)) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Create a master spreadsheet of all enriched biological functions in DEGs 

```{r}
enriched.bp.degs <- vector(mode = "list", length = 6)
names(enriched.bp.degs) <- c("AL-UP", "AL-DOWN", "AM-UP", "AM-DOWN", "ML-UP", "ML-DOWN")

for (i in 1:6) {
  enriched.bp.degs[[i]] <- 
    read_delim(file = paste("../results/GO_MWU/DAVID_", names(enriched.bp.degs)[i], "_FC0.5.txt", sep=""), 
               delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT")
}

enriched.bp.degs <- bind_rows(enriched.bp.degs, .id = "deg_set") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%
  clean_names() #%>% 
  #select(deg_set, go, process, count, percent, p_value, fold_enrichment, fdr) 

enriched.bp.degs %>% write_clip() #copy to clipboard to paste into google spreadsheet 

# How many enriched BP at p<0.05?
enriched.bp.degs %>% filter(p_value<0.05) %>% group_by(deg_set) %>% tally()

# How many enriched BP at FDR<0.1?
enriched.bp.degs %>% filter(fdr<0.1) %>% group_by(deg_set) %>% tally()


```

```{r}
enriched.keywords.degs <- vector(mode = "list", length = 6)
names(enriched.keywords.degs) <- c("AL-UP", "AL-DOWN", "AM-UP", "AM-DOWN", "ML-UP", "ML-DOWN")

for (i in 1:6) {
  enriched.keywords.degs[[i]] <- 
    read_delim(file = paste("../results/GO_MWU/DAVID_", names(enriched.keywords.degs)[i], "_FC0.5.txt", sep=""), 
               delim = "\t") %>%
      filter(Category=="UP_KW_BIOLOGICAL_PROCESS")
}

enriched.keywords.degs <- bind_rows(enriched.keywords.degs, .id = "deg_set") %>%
  separate(Term, into = c("go", "process"), sep = "~") %>%
  clean_names() #%>% 
  #select(deg_set, go, process, count, percent, p_value, fold_enrichment, fdr) 

```

Save DEG enrichment results (BP, MF, CC) for supplemental tables 

```{r}
enriched.degs <- vector(mode = "list", length = 6)
names(enriched.degs) <- c("AL-UP", "AL-DOWN", "AM-UP", "AM-DOWN", "ML-UP", "ML-DOWN")

for (i in 1:6) {
  enriched.degs[[i]] <- 
    read_delim(file = paste("../results/GO_MWU/DAVID_", names(enriched.degs)[i], "_FC0.5.txt", sep=""), 
               delim = "\t") %>%
      filter(Category %in% c("GOTERM_BP_DIRECT", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT"))
}

# Create one master spreadsheet with all enrichment results, reformat, save to .csv
bind_rows(enriched.degs, .id = "deg_set") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%
  #clean_names() %>% 
  separate(deg_set, sep = "-", into = c("Contrast", "Change in OA")) %>%
  mutate(Contrast=case_when(
    Contrast=="AL" ~ "Severe OA vs. Ambient",
    Contrast=="AM" ~ "Moderate OA vs. Ambient",
    Contrast=="ML" ~ "Severe OA vs. Moderate OA")) %>%
  mutate(`Change in OA`=case_when(
    `Change in OA`=="UP" ~ "Upregulated", TRUE ~ "Downregulated")) %>%
  mutate(Category=case_when(
    Category=="GOTERM_BP_DIRECT" ~ "Biological Process",
    Category=="GOTERM_MF_DIRECT" ~ "Molecular Function",
    Category=="GOTERM_CC_DIRECT" ~ "Cellular Component")) %>%
  dplyr::rename(Term=process, `Number of Genes` = Count, 
                `P-Value`=PValue, `Genes (Uniprot IDs)`=Genes) %>%
  mutate(`Fold Enrichment`=round(`Fold Enrichment`, digits = 1),
    `P-Value`=as.numeric(formatC(`P-Value`, format = "e", digits = 2)),
    FDR=as.numeric(formatC(FDR, format = "e", digits = 2))) %>%
  select(Contrast, `Change in OA`, Category, GO, Term, `Number of Genes`, `Genes (Uniprot IDs)`, `Fold Enrichment`, `P-Value`, FDR) %>%
  arrange(Category, Contrast, `Change in OA`, `P-Value`) %>%
  filter(`P-Value`<0.05) %>%
  write_csv(file = "../results/Supplemental-Table-4_DEG-Enrichment-Results.csv", col_names = TRUE)
```


Bubble plot of enriched Biological Processes 

(Figure 5 in RKC manuscript)

```{r}
# Biological Processes

pdf(file="../results/Figure5.pdf", height = 7.5, width = 6.4)

enriched.bp.degs %>% filter(p_value<0.05) %>%
  group_by(deg_set, genes, count) %>%
  summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(enriched.bp.degs %>% select(deg_set, category, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("deg_set", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(deg_set, p_value) %>% group_by(deg_set) %>% dplyr::slice(1:20) %>%
  separate(deg_set, sep = "-", into = c("contrast", "OA.response"), remove = FALSE) %>%
  mutate(up.treatment=deg_set) %>%
  mutate_at("up.treatment", function(x) {
    case_when(
      x == "AL-DOWN" ~ "Ambient",
      x == "AL-UP" ~ "Severe",
      x == "AM-DOWN" ~ "Ambient",
      x == "AM-UP" ~ "Moderate",
      x == "ML-DOWN" ~ "Moderate",
      x == "ML-UP" ~ "Severe")}) %>%
  mutate_at(c("deg_set", "OA.response", "up.treatment"), factor) %>%
  mutate(contrast=factor(contrast, levels=c("AM", "AL", "ML"))) %>% 
  mutate(up.treatment=factor(up.treatment, levels=c("Severe", "Moderate", "Ambient"))) %>% 
  ungroup() %>%
  arrange(up.treatment, contrast, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=contrast, col=up.treatment)) + 
  geom_point(aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Enriched Biological\nProcesses"))) +
  scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"),
             guide = guide_legend(override.aes = list(size=4))) +
  # scale_alpha("FDR-adjusted\nP-value", range = c(.8,0.2), #breaks = c(0, .25, .5, .75, 1),
  #             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2.75,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_blank(), #element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.ticks.x = element_blank(), 
        legend.position = "right", 
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.45, label = "Ambient vs.\nModerate OA", size=2, col="gray15") +
  annotate("text", x = 2, y = -.45, label = "Ambient vs.\nSevere OA", size=2, col="gray15") +
  annotate("text", x = 3, y = -.45, label = "Moderate OA\nvs. Severe OA", size=2, col="gray15")
#coord_cartesian(clip = "off", ylim = c(-.7, 43.25))
#coord_cartesian(clip = "off", ylim = c(-0.6, 20))
dev.off()
```

## Messing around with the DEGs 
```{r}
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange, pvalue, padj) %>% 
  left_join(P.camt.blast.GO, by = c("gene"="gene_id")) %>% 
  #filter(!is.na(gene_ontology_i_ds)) %>% 
  filter(padj<0.05) %>% View()


# Here I check that the genes with LFC < 0 are UPREGULATED in the lower pH treatment in all pairwise comparisons
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>%
  filter(padj<0.05) %>% dplyr::select(gene,log2FoldChange, padj) %>% 
  left_join(P.camt.blast.GO, by = c("gene"="gene_id")) %>% 
  filter(!is.na(gene_ontology_i_ds)) %>% View()


test <- diffex.AM %>% as.data.frame() %>% filter(log2FoldChange<0) %>% rownames()


# Check out genes related to chitin degradation
test <- (diffex.AL %>% as.data.frame() %>% rownames_to_column(var = "geneID") %>%
  filter(geneID %in%
           (P.camt.blast %>% 
              filter(SPID %in% c("P49010", "Q9BZP6", "Q11174", "Q13231")))$geneID))$geneID

for (i in 1:length(test)) {
a <-  plotCounts(dds.DESeq.pH, gene=test[i], intgroup=c("Treatment"), returnData = TRUE) %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
print(ggplot(a %>% rownames_to_column("sample"),
       aes(x=Treatment, y=count, color=Treatment, label=sample)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=2) +
    theme_bw() +
    ggtitle(test[i]) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
}
```


## Prep for Ingenuity Pathway Analysis (IPA)
Need tab-delimited file with gene ID column, then for each comparison 2 columns: 1: LFC, 2) p-value 

```{r}
IPA.degs <- cbind(
res.pco2.AM %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% rename_with( ~ paste0("Amb_vs_Mod_", .x)),
res.pco2.AL %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% rename_with( ~ paste0("Amb_vs_Low_", .x)),
res.pco2.ML %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% rename_with( ~ paste0("Mod_vs_Low_", .x))) %>% 
  rownames_to_column("geneID")
save(IPA.degs, file = "../results/IPA/IPA.degs")

# write_delim(IPA.degs, file = "../results/deseq2/DESeq2-results-for-IPA.tab", delim = "\t", 
#               col_names = TRUE, quote = "none")
```


### Examine some of the most consistently differentially expressed genes in OA treatments 

```{r}
# Calculate per-treatment mean, sd, and cv for each gene (using variance-stabilizing normalized counts)

# first extract the normalized counts
gene.counts.norm <- assay(vsd.pH) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  pivot_longer(-gene, values_to = "counts.vsd", names_to = "Sample") %>%
  left_join(sample.info[,c("Sample", "Treatment")]) %>% 
  mutate_at(c("gene", "Sample"), factor) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))

# calculate mean, sd, cv and annotate 
# THIS CODE NEEDS TO BE UPDATED - need to re-write deleted code that assigns genes upreg / downreg in each treatment, NOT binning mod/low

genes.stats <- gene.counts.norm %>% group_by(Treatment, gene) %>% 
  summarize(mean=mean(counts.vsd), median=median(counts.vsd), sd=sd(counts.vsd), cv=sd(counts.vsd)/mean(counts.vsd)) %>% 
  mutate(DEG=if_else(gene %in% c(rownames(diffex.AM),rownames(diffex.AL), rownames(diffex.ML)), "DEG", "Non-DEG")) %>%
  mutate(change = case_when(
  gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) ~ "Up-regulated",
  gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) ~ "Down-regulated")) %>% 
  
  mutate(change2 = case_when(
    
    
  ))
  
    mutate(change=replace_na(change,"Non-DEG")) %>% 
  mutate(DEG=factor(DEG, levels=c("Non-DEG", "DEG"))) %>% 
  mutate(change=factor(change, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  ungroup() %>% 
  #left_join(P.plat.bestblast %>% select(id, id_gtf, hit_acc, hit_desc, e_value), by=c("gene"="id")) %>%
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, evalue, gene_ontology_biological_process, gene_ontology_i_ds), by=c("gene"="gene_id")) %>%
  
  # Add Log2FoldChange and padj for each pairwise treatment comparison for each gene
  left_join(cbind(res.pco2.AM %>% as.data.frame() %>% dplyr::rename("L2FC.AM"="log2FoldChange", "padj.AM"="padj") %>% 
  rownames_to_column("gene") %>% dplyr::select(gene, L2FC.AM, padj.AM),
res.pco2.AL %>% as.data.frame() %>% dplyr::rename("L2FC.AL"="log2FoldChange", "padj.AL"="padj") %>% 
  rownames_to_column("gene") %>% dplyr::select(L2FC.AL, padj.AL),
res.pco2.ML %>% as.data.frame() %>% dplyr::rename("L2FC.ML"="log2FoldChange", "padj.ML"="padj") %>% 
  rownames_to_column("gene") %>% dplyr::select(L2FC.ML, padj.ML)), by = "gene")
save(genes.stats, file = "../results/bowtie/per-gene-summary-stats")

hist(genes.stats$mean)
hist(genes.stats$cv, breaks = 50)

# What's the range of mean of the normalized counts, what is considered high? I'd say >10
ggplot(genes.stats) + geom_boxplot(aes(x=Treatment, y=mean, colour=Treatment)) + facet_wrap(~change)

# What's the range of CV of the normalized counts, what is considered low-level variation? Q1 extends to 0.02-0.04
ggplotly(ggplot(genes.stats) + geom_boxplot(aes(x=Treatment, y=cv, colour=Treatment)) + facet_wrap(~change))

# Which genes were highly UP-regulated (mean of transformed counts >10) consistently within an OA treatment? (CV<0.04)
genes.stats %>% filter(mean>10 & cv<.02 & change=="Up-regulated" & Treatment!="Ambient") %>% View()

# Which genes were highly DOWN-regulated (mean of transformed counts >10) consistently within an OA treatment? (CV<0.04)
genes.stats %>% filter(mean>10 & cv<.02 & change=="Down-regulated" & Treatment!="Ambient") %>% View()

# To inspect counts for individual genes of interest, use this code: < --- OUT OF DATE NEED TO USE RKC GENOME GENE NAMES (E.G. "g##")
# e.g. calcium-dependent secretion activator\nevm.TU.HiC_scaffold_92.184
# e.g. heat shock protein\nevm.TU.HiC_scaffold_43.188
# e.g. peroxiredoxin\nevm.TU.HiC_scaffold_16.208
# e.g. superoxide dismutase [Mn], mitochondrial-like\nevm.TU.HiC_scaffold_64.12
# e.g. Methylcytosine dioxygenase TET1\nevm.TU.HiC_scaffold_78.1

# Outlier SNPs
# e.g. nephrin-like \nevm.TU.HiC_scaffold_21.266
# e.g. histone deacetylase 4-like\nevm.TU.HiC_scaffold_91.210
# e.g. ATP-dependent RNA helicase-like, partial\nevm.TU.HiC_scaffold_61.127
# a <-  plotCounts(dds.DESeq.pH, gene="evm.TU.HiC_scaffold_92.184", intgroup=c("Treatment"), returnData = TRUE) %>% 
#   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
#   
# ggplot(a %>% rownames_to_column("sample"),
#        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
#   #geom_boxplot(outlier.shape = NA) +
#   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=2) +
#     theme_bw() +
#     ggtitle("calcium-dependent secretion activator\nevm.TU.HiC_scaffold_92.184") +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none") +
#   scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))
```

What is the mean Coeff. Variation (CV) for DEGs upregulated by treatment? 
theory - those upregulated in OA have lower CV 
```{r}
# Here the rows with change="Non-DEG" and DEG="Non-DEG" are genes that are not a DEG in any comparison
# But the rows with change="Non-DEG" and DEG="DEG" are genes that are a DEG between some comparison, but not for that specific treatment. 

genes.stats %>% 
  group_by(change, Treatment) %>%
  summarize(mean.cv=mean(cv), sd.cv=sd(cv))

genes.stats %>% 
  group_by(change, Treatment) %>%
  summarize(mean.cv=mean(cv), sd.cv=sd(cv)) %>%
  mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  ggplot(aes(x=pco2, y=mean.cv, group=change, color=change)) + 
  geom_line() +
  geom_point()+
  geom_errorbar(aes(ymin=mean.cv-sd.cv, ymax=mean.cv+sd.cv), width=.2,
                 position=position_dodge(0.05))

  


genes.stats %>% filter(DEG=="DEG" & change!="Non-DEG") %>% 
ggplot(aes(x=Treatment, y=log(cv), colour=Treatment)) + geom_boxplot() + facet_wrap(~change) + theme_minimal()

```

## Significant differences in CV among treatment by DEG group? 
- Can't use anova, too many data points. 

```{r}
genes.stats %>% 
  group_by(change, Treatment) %>%
  summarize(mean.cv=mean(cv), sd.cv=sd(cv))

x <- genes.stats %>% filter(change=="Up-regulated")
summary(aov(cv^2 ~ Treatment, data=x))
TukeyHSD(aov(cv^2 ~ Treatment, data=x))

x <- genes.stats %>% filter(change=="Down-regulated")
summary(aov(cv ~ Treatment, data=x))
TukeyHSD(aov(cv ~ Treatment, data=x))

x <- genes.stats %>% filter(DEG=="Non-DEG")
summary(aov(cv ~ Treatment, data=x))
TukeyHSD(aov(cv ~ Treatment, data=x))
```

# Exploring differences in within-group variances 

## Broad differences using multivariate approaches
- Identify centroid (mean of points on PC axes), calculate euclidean distances, compare those  
- Perform a multivariate homogeneity of variance test - betadisper from vegan package

### All genes in matrix 

```{r}
(a <- plotPCA(vsd.pH, intgroup="Treatment", ntop=nrow(assay(vsd.pH))))

# calculate centroids for each 
b <- a$data %>% group_by(Treatment) %>% summarize(PC1.mean=mean(PC1), 
                                             PC1.sd=sd(PC1),
                                             PC2.mean=mean(PC2),
                                             PC2.sd=sd(PC2))
# calculate euclidean distance from treatment centroid for each sample 
c <- a$data %>% left_join(b, by = "Treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2)) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
# examine euclidean distances from centroid 
ggplot(c, aes(x=Treatment, y=dist.cent, colour=Treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nAll genes") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))

ggplot(data=c, aes(x=PC1, y=PC2, group=Treatment, colour=Treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=Treatment), col="black", shape=22, size=4) +
   ggtitle("PCA by Treatment, All genes\n(var-stabilizing transformed)") + theme(legend.position = c(0.08, 0.85)) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
    scale_fill_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) 

# Examine homogeneity of multivariate dispersions using all data 

# By examining distance to centroid on PCA
hist(c$dist.cent) #normal distribution? 
shapiro.test(c$dist.cent) #not really 

hist(log(c$dist.cent)) #log transformed = yes  
shapiro.test(log(c$dist.cent)) #log transformed = yes  
summary(aov(log(dist.cent)~Treatment, data=c)) # no diff
TukeyHSD(aov(log(dist.cent)~Treatment, data=c)) # no diff 


# By using the vegan package's `betadisper` function 
# This uses the sample-sample distance matrix, "sampleDistss", generated previously 
all(rownames(dds.pH.df) == attributes(sampleDistss)$Labels) # should === TRUE
print(disper <- betadisper(sampleDistss, dds.pH.df$Treatment)) 
anova(disper) #yes a difference, but how?
TukeyHSD(disper) #Dispersion varies between Ambient & Low pH treatments only 
permutest(disper, pairwise = TRUE) #another way of testing, same result
boxplot(disper)
plot(disper)
```

### DEGs only 

```{r}
degs.vst <- vst(
  DESeqDataSetFromMatrix(
    countData = counts.tk[,unique(c(rownames(diffex.AM),rownames(diffex.AL),rownames(diffex.ML)))] %>% t(),
                              colData = counts.tk[,"Treatment", drop=FALSE] ,
                              design = ~ Treatment), blind=TRUE)

(a <- plotPCA(degs.vst, intgroup="Treatment", ntop=nrow(assay(vsd.pH))))

# calculate centroids for each 
b <- a$data %>% group_by(Treatment) %>% summarize(PC1.mean=mean(PC1), 
                                             PC1.sd=sd(PC1),
                                             PC2.mean=mean(PC2),
                                             PC2.sd=sd(PC2))
# calculate euclidean distance from treatment centroid for each sample 
c <- a$data %>% left_join(b, by = "Treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2)) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
# examine euclidean distances from centroid 
ggplot(c, aes(x=Treatment, y=dist.cent, colour=Treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nDEGs only") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))

ggplot(data=c, aes(x=PC1, y=PC2, group=Treatment, colour=Treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=Treatment), col="black", shape=22, size=4) +
   ggtitle("PCA by Treatment, DEGs only\n(var-stabilizing transformed)") + theme(legend.position = c(0.08, 0.85)) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
    scale_fill_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) 

# Examine homogeneity of multivariate dispersions using all data 

# By examining distance to centroid on PCA
hist(c$dist.cent) #normal distribution? 
shapiro.test(c$dist.cent) #not really 

hist(c$dist.cent^.5) #log transformed = yes  
shapiro.test(c$dist.cent^.5) #log transformed = yes  
summary(aov(dist.cent^.5~Treatment, data=c)) # very different 
TukeyHSD(aov(dist.cent^.5~Treatment, data=c)) # ambient diff from other treatments - more spread 


# By using the vegan package's `betadisper` function 
# Need to generate new distance matrix after selecting only DEGs 
sampleDistss.degs <- dist(t(assay(degs.vst)))

all(rownames(dds.pH.df) == attributes(sampleDistss.degs)$Labels) # should === TRUE
print(disper <- betadisper(sampleDistss.degs, dds.pH.df$Treatment)) 
anova(disper) #yes a difference, but how?
TukeyHSD(disper) #Dispersion varies between Ambient & other two factors  
permutest(disper, pairwise = TRUE) #another way of testing, same result
boxplot(disper)
plot(disper)
```

### Excluding DEGs 

```{r}
notdegs.vst <- vst(
  DESeqDataSetFromMatrix(countData = counts.tk[,-which(names(counts.tk) %in% c("Tank", "Treatment", "Treatment_Tank", rownames(diffex.AM),rownames(diffex.AL),rownames(diffex.ML)))] %>% t(),
                              colData = counts.tk[,"Treatment", drop=FALSE] ,
                              design = ~ Treatment), blind=TRUE)

(a <- plotPCA(notdegs.vst, intgroup="Treatment", ntop=nrow(assay(vsd.pH))))

# calculate centroids for each 
b <- a$data %>% group_by(Treatment) %>% summarize(PC1.mean=mean(PC1), 
                                             PC1.sd=sd(PC1),
                                             PC2.mean=mean(PC2),
                                             PC2.sd=sd(PC2))
# calculate euclidean distance from treatment centroid for each sample 
c <- a$data %>% left_join(b, by = "Treatment") %>% 
  mutate(dist.cent=sqrt((PC1-PC1.mean)^2+(PC2-PC2.mean)^2)) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
# examine euclidean distances from centroid 
ggplot(c, aes(x=Treatment, y=dist.cent, colour=Treatment)) + 
  geom_boxplot() + geom_jitter(width = .2) +
  ggtitle("Euclidean distances to centroid of PCA by treatment\nAll genes except DEGs") +
  theme_minimal() + theme(legend.position = "none", axis.title.x = element_blank()) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"))

ggplot(data=c, aes(x=PC1, y=PC2, group=Treatment, colour=Treatment)) + 
  geom_point() + 
    theme_minimal()+ stat_ellipse() +
  geom_segment(aes(x=PC1.mean, y=PC2.mean, xend=PC1, yend=PC2)) +
      geom_point(aes(x=PC1.mean, y=PC2.mean,fill=Treatment), col="black", shape=22, size=4) +
     ggtitle("PCA by Treatment, excluding DEGs\n(var-stabilizing transformed)") + theme(legend.position = c(0.08, 0.85)) +
  scale_color_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) +
    scale_fill_manual(values = c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")) 

# Examine homogeneity of multivariate dispersions using all data 

# By examining distance to centroid on PCA
hist(c$dist.cent) #normal distribution? 
shapiro.test(c$dist.cent) #not really 

hist(c$dist.cent^.5) #sqrt-transformed = yes  
shapiro.test(c$dist.cent^.5) #sqrt transformed = yes  
summary(aov(dist.cent^.5~Treatment, data=c)) # not different at all  
TukeyHSD(aov(dist.cent^.5~Treatment, data=c)) # no differences at all 

# By using the vegan package's `betadisper` function 
# Need to generate new distance matrix after selecting only DEGs 
sampleDistss.notdegs <- dist(t(assay(notdegs.vst)))

all(rownames(dds.pH.df) == attributes(sampleDistss.notdegs)$Labels) # should === TRUE
print(disper <- betadisper(sampleDistss.notdegs, dds.pH.df$Treatment)) 
anova(disper) #yes a difference, but how?
TukeyHSD(disper) #Dispersion varies between Ambient & other two factors  
permutest(disper, pairwise = TRUE) #another way of testing, same result
boxplot(disper)
plot(disper)
```


## ANOTHER Gene-wise variance 

Explore coefficient of variation by treatment, and gene groups 

```{r}
# assay(vsd.pH) #variance-stabilizing transformed count matrix
# mcols(mcols(dds.DESeq.pH))$description #Description of DESeq summary stats columns
# mcols(dds.DESeq.pH) #DESeq summary stats 

# Create dataframe in long format with gene, sample number, transformed counts, and treatment 
# use assay(vsd.pH) for variance-stabilizing transformed data 
# use assay(normTransform(dds.pH)) for log2(n+1) transformation 
d <- assay(vsd.pH) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  pivot_longer(-gene, values_to = "counts.vsd", names_to = "Sample") %>%
  left_join(sample.info[,c("Sample", "Treatment")]) %>% 
  mutate_at(c("gene", "Sample"), factor) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))

# Test for homogeneity of variances among treatments  
hist(d$counts.vsd) # first inspect transformed counts for normality - not super convincing 
shapiro.test(d$counts.vsd %>% sample(5000, replace=FALSE)) #not normal 
ks.test(x=d$counts.vsd,y='pnorm',alternative='two.sided')
leveneTest(counts.vsd ~ Treatment, data=d) #variances differ, but how? 
fligner.test(counts.vsd ~ Treatment, data=d) #variances differ, but how? 

ggplot(d %>% group_by(Treatment, gene) %>% summarise(cv=100*sd(counts.vsd)/mean(counts.vsd)), 
       aes(x=Treatment, y=log(cv), fill=Treatment)) + 
  geom_violin() + ggtitle("Normalized counts by treatment\nAll genes")# low looks lower than other treatments

cvs <- d %>% group_by(Treatment, gene) %>% summarise(cv=100*sd(counts.vsd)/mean(counts.vsd))
hist(log(test$cv))

cvs$cv %>% summary()
summary(aov(log(cv+0.001) ~ Treatment, data=cvs))
TukeyHSD(aov(log(cv+0.001) ~ Treatment, data=cvs))

cvs %>% group_by(Treatment) %>% summarise(mean=mean(cv), sd=sd(cv))
```


```{r}
# Calculate & inspect coefficient of variation for each gene within treatments 
e <- d %>% group_by(Treatment, gene) %>% 
  summarize(mean=mean(counts.vsd), sd=sd(counts.vsd), cv=sd(counts.vsd)/mean(counts.vsd))

ggplot(e, aes(x=Treatment, y=sd, colour=Treatment)) + geom_violin() +geom_boxplot() + ggtitle("Standard deviation by treatment\nAll genes")
ggplot(e, aes(x=Treatment, y=cv, colour=Treatment)) + geom_violin() +geom_boxplot() + ggtitle("CV by treatment\nAll genes")
ggplot(e, aes(x=Treatment, y=log(cv), colour=Treatment)) + geom_violin() +geom_boxplot() + ggtitle("log(CV) by treatment\nAll genes")
ggplot(e, aes(x=mean, y=cv, colour=Treatment)) + geom_point(size=1) + facet_wrap(~Treatment) + theme_minimal() + ggtitle("Coefficient of variation of counts (normalized) by treatment\nAll genes")

# inspect cv by treatment - Genes that are differentially expressed between Ambient and the two pH treatments only. 
# NOTE: this does not include genes that are diff. expressed between moderate and low pH (only a handful, though)

f <- e %>% mutate(DEG=if_else(gene %in% c(rownames(diffex.AM),rownames(diffex.AL)), "DEG", "Non-DEG")) %>%
  
  # # This column might have issues - it might erroneously assigns the term "upregulated" to both moderate and low pH treatments even if it was just upregulated in one of those (same for down regulated)
  # mutate(change = case_when(
  # gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) ~ "Up-regulated",
  # gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) ~ "Down-regulated")) %>% 
  # mutate(change=replace_na(change,"Non-DEG")) %>% 

  # Generate new column that indicates, RELATIVE TO THAT TREATMENT, whether the gene is upregulated, downregulated, or not a DEG 
    mutate(change_in_treatment = case_when(
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Moderate") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Low") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Moderate") ~ "Down-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Low") ~ "Down-regulated",
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Ambient") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) & Treatment=="Ambient") ~ "Up-regulated",
    (gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Ambient") ~ "Down-regulated",
    (gene %in% c(rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) & Treatment=="Ambient") ~ "Down-regulated",
    TRUE ~ "Non-DEG")) %>%
  mutate(DEG=factor(DEG, levels=c("Non-DEG", "DEG"))) %>% 
  #mutate(change=factor(change, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  mutate(change_in_treatment=factor(change_in_treatment, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  ungroup() %>%
  #left_join(P.plat.bestblast %>% select(id, id_gtf, hit_acc, hit_desc, e_value), by=c("gene"="id")) %>%
  left_join(P.camt.blast.GO %>% select(gene_id, spid, protein_names, evalue), by=c("gene"="gene_id")) 

# # create vector of gene names that are diff. expressed among moderate and low pH - not included in this variance comparison
# remove <- (f %>% ungroup() %>% filter(DEG=="DEG") %>% filter(change=="Non-DEG"))$gene %>% as.vector()
# removed this from the below ggplot: f %>% filter(!c(gene %in% remove))

# Plot CV by treatment for genes that are upregulated, downregulated, and not degs 
ggplot(f, aes(x=Treatment, y=log(cv), colour=Treatment)) + 
  geom_boxplot() + facet_wrap(~change_in_treatment) + theme_cleveland() + 
  theme(axis.title.x = element_blank()) + ylab("Log(Coeff. variation of transformed counts)") + 
  ggtitle("Gene-wise variance within treatments\nNon-DEGs, and DEGs that are down- and up-regulated in OA") +
  theme(axis.text.x = element_blank(), plot.title = element_text(size=10), legend.position = "bottom")

#ggplotly(
  ggplot(f, 
       aes(x=mean, y=sd, colour=Treatment)) + geom_point(size=1, aes(text=spid)) + facet_wrap(~Treatment+change_in_treatment) + 
  theme_minimal() + theme(plot.title = element_text(size=11), legend.position = "bottom") + 
  ylab("Standard deviation of transformed gene counts") + xlab("Mean of transformed counts") +
  ggtitle("Per-gene variation within treatments (SD) ~ mean of transformed")#, tooltip = "text")

# What are those highly variable genes that are upregulated in moderate and low pH? 
f %>% filter(Treatment!="Ambient") %>% filter(DEG=="DEG") %>% filter(change_in_treatment=="Up-regulated") %>% filter(sd>=1) %>% View()

# What are those genes that are upregulated in moderate and low pH at very consistent levels (very low variability)? These would be critical for function.  

# First define a "low variability" threshold 
ggplotly(ggplot(f, aes(x=Treatment, y=cv, colour=Treatment)) + 
  geom_boxplot() + facet_wrap(~change_in_treatment) + theme_cleveland() + 
  theme(axis.title.x = element_blank()) + ylab("Log(Coeff. variation of transformed counts)") + 
  ggtitle("Gene-wise variance within treatments\nNon-DEGs, and DEGs that are down- and up-regulated in OA") +
  theme(axis.text.x = element_blank(), plot.title = element_text(size=10), legend.position = "bottom"))

# Q1 = 0.02 - 0.03, use .03 
# Check out their functions
f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% filter(!is.na(spid)) %>% View()
f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% filter(!is.na(spid)) %>% View()
f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% filter(!is.na(spid)) %>% View()

# How many upregulated low CV genes in each treatment?
f %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  group_by(change_in_treatment, Treatment) %>% tally()


# How many genes were upregulated in severe OA treatment relative to ambient? 
rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0)) %>% unique() %>% length()

# How many genes were upregulated in moderate OA treatment relative to ambient? 
rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)) %>% unique() %>% length()

# How many genes were upregulated in ambient treatment relative to either OA treatment? 
c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) %>%
  unique() %>% length()


# Here I save a list of upregulated DEGs that are also expressed at highly consistent levels (very low variability, CV <0.02).
f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
    write.csv(file = "../results/DEGs.up-low-variability-Ambient_annotated.csv", row.names = F, quote = T)

f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
    write.csv(file = "../results/DEGs.up-low-variability-Moderate_annotated.csv", row.names = F, quote = T)

f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
    write.csv(file = "../results/DEGs.up-low-variability-Severe_annotated.csv", row.names = F, quote = T)

# Save the low CV upregulated genes (i.e. those that are super critical) for other analyses
critical.amb <- f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% dplyr::select(gene) %>% unlist() %>% as.vector()
critical.mod <- f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% dplyr::select(gene) %>% unlist() %>% as.vector()
critical.low <- f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% dplyr::select(gene) %>% unlist() %>% as.vector()

```



```{r}
# 
# # Upregulated DEGs in Moderate (compared to ambient) that are also expressed very consistently 
# res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.mod ~ 1,
#     TRUE ~ 0)) %>%
#   left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-mod.csv", quote = F,row.names = F)
# 
# # Upregulated DEGs in Low (compared to ambient) that are also expressed very consistently 
# res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.low ~ 1,
#     TRUE ~ 0)) %>%
#    left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-low.csv", quote = F,row.names = F)
# 
# # In comparison, what functions are upregulated in consistent levels in ambient (compared to low)?
# # Note: the filter change=="Down-regulated" refers to the fact that the gene is downregulated in OA treatment, and thus upregualted in ambient treatment 
# 
# res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.amb ~ 1,
#     TRUE ~ 0)) %>% 
#    left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb.csv", quote = F,row.names = F)
# 
# 
# critical.mod[! critical.mod %in% TE.genes] %>% length()
# critical.mod %>% length()
#     
# # What if we remove Transposable Elements from the list? 
# 
# res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.mod[! critical.mod %in% TE.genes] ~ 1,
#     TRUE ~ 0)) %>%
#   left_join(P.camt.blast.GO, by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
# 
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-mod.csv", quote = F,row.names = F)
# 
# # Upregulated DEGs in Low (compared to ambient) that are also expressed very consistently 
# res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
#   mutate(critical= case_when(
#     gene %in% critical.low ~ 1,
#     TRUE ~ 0)) %>%
#    left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
#   filter(!is.na(gene_ontology_i_ds)) %>%
#   dplyr::select(gene, critical) %>% 
#   write.csv("../results/GO_MWU/DEGs-upreg-lowCV_amb-low.csv", quote = F,row.names = F)

```

DAVID enrichment with critical genes 

```{r}
# Upregulated DEGs in Ambient (compared to either OA treatment) that are also expressed very consistently 
f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
   filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Upregulated DEGs in Moderate OA (compared to ambient) that are also expressed very consistently 
f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
   filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Upregulated DEGs in Severe OA (compared to ambient) that are also expressed very consistently 
f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
   filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Background - all genes in data set 
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names), by = c("gene"="gene_id")) %>% 
  filter(!is.na(spid)) %>% select(spid) %>% unlist() %>% as.vector() %>% write_clip()

# Read in results, create one table with all low CV genes 
enriched.bp.lowcv <- vector(mode = "list", length = 3)
names(enriched.bp.lowcv) <- c("ambient", "moderate", "low")

for (i in 1:3) {
  enriched.bp.lowcv[[i]] <- read_delim(file = paste("../results/GO_MWU/DAVID_lowCV-upreg-", names(enriched.bp.lowcv)[i], ".txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05)
}

enriched.bp.lowcv <- bind_rows(enriched.bp.lowcv, .id = "treatment") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%
  clean_names() %>% 
  select(treatment, go, process, count, percent, p_value, fold_enrichment, fdr) 

enriched.bp.lowcv %>% write_clip() #copy to clipboard to paste into google spreadsheet 

enriched.bp.lowcv %>% filter(fdr<0.1) %>% group_by(treatment) %>% tally()


# what % of low CV genes are TEs?
# Note, the vector "TE.genes" is defined elsewhere 

# In Ambient
(f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
  filter(str_detect(protein_names, TE.genes)) %>% nrow()) /
  (f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  filter(!is.na(spid)) %>% nrow())

# In Moderate OA
(f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
  filter(str_detect(protein_names, TE.genes)) %>% nrow()) /
  (f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  filter(!is.na(spid)) %>% nrow())

# In Severe OA
(f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>%
  filter(str_detect(protein_names, TE.genes)) %>% nrow()) /
  (f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% filter(cv<0.03) %>% 
  filter(!is.na(spid)) %>% nrow())

```

How many upregulated genes with CV < 0.03 in each treatment? What % of upregulated genes are low-CV? 

```{r}
critical.amb %>% length()
critical.mod %>% length()
critical.low %>% length()

critical.amb %>% length() / f %>% filter(Treatment=="Ambient") %>% filter(change_in_treatment=="Up-regulated") %>% nrow()
critical.mod %>% length() / f %>% filter(Treatment=="Moderate") %>% filter(change_in_treatment=="Up-regulated") %>% nrow()
critical.low %>% length() / f %>% filter(Treatment=="Low") %>% filter(change_in_treatment=="Up-regulated") %>% nrow()
```

What's the average +/- SD CV? 

```{r}
f %>% group_by(Treatment, change_in_treatment) %>%
  summarise(mean.cv=round(100*mean(cv), digits = 1), sd.cv=round(100*sd(cv), digits = 1))

f %>% filter(change_in_treatment=="Up-regulated")

```

```{r}
f %>% filter(!is.na(spid)) %>% filter(DEG == "DEG") %>% View()
```

Save Low-CV enrichment results (BP, MF, CC) for supplemental tables 

```{r}
# Read in results, create one table with all low CV genes 
enriched.lowcv <- vector(mode = "list", length = 3)
names(enriched.lowcv) <- c("ambient", "moderate", "low")

for (i in 1:3) {
  enriched.lowcv[[i]] <- read_delim(file = paste("../results/GO_MWU/DAVID_lowCV-upreg-", names(enriched.lowcv)[i], ".txt", sep=""), delim = "\t") %>%
      filter(Category %in% c("GOTERM_BP_DIRECT", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT"))
}

# Create one master spreadsheet with all enrichment results, reformat, save to .csv
bind_rows(enriched.lowcv, .id = "Treatment") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%

  mutate(Treatment=case_when(
    Treatment=="low" ~ "severe OA",
    Treatment=="moderate" ~ "moderate OA",
    TRUE ~ "ambient")) %>%

  mutate(Category=case_when(
    Category=="GOTERM_BP_DIRECT" ~ "Biological Process",
    Category=="GOTERM_MF_DIRECT" ~ "Molecular Function",
    Category=="GOTERM_CC_DIRECT" ~ "Cellular Component")) %>%
  
  dplyr::rename(Term=process, `Number of Genes` = Count, 
                `P-Value`=PValue, `Genes (Uniprot IDs)`=Genes) %>%
  mutate(`Fold Enrichment`=round(`Fold Enrichment`, digits = 1),
    `P-Value`=as.numeric(formatC(`P-Value`, format = "e", digits = 2)),
    FDR=as.numeric(formatC(FDR, format = "e", digits = 2))) %>%
  select(Treatment, Category, GO, Term, `Number of Genes`, `Genes (Uniprot IDs)`, `Fold Enrichment`, `P-Value`, FDR) %>%
  arrange(Category, Treatment, `P-Value`) %>%
  filter(`P-Value`<0.05) %>%
  write_csv(file = "../results/Supplemental-Table-5_Upregulated-Low-CV-Enrichment-Results.csv", col_names = TRUE)
```


<!-- How many don't have GO terms? -->
<!-- ```{r} -->
<!-- res.pco2.AL %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% -->
<!--   rownames_to_column("gene") %>%  -->
<!--   filter(gene %in% critical.low) %>% -->
<!--   left_join(P.camt.blast.GO %>%  -->
<!--               dplyr::select(gene_id, gene_ontology_i_ds, protein_names, spid, -->
<!--                             gene_names, organism, gene_ontology_biological_process),  -->
<!--             by = c("gene"="gene_id")) %>% filter(!is.na(protein_names)) %>% -->
<!--   mutate(my_category=case_when( -->
<!--     str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein") ~ "Transposable Elements", -->
<!--     str_detect(protein_names,  -->
<!--                "Exportin-5|methyltransferase|demethylase|Histone deacetylase") ~ "Epigenetic Factor", -->
<!--     str_detect(protein_names,  -->
<!--                     "Zinc finger protein|LIM domain-binding protein 2") ~ "Transcription Regulation",  -->
<!--     str_detect(protein_names,  -->
<!--                "Toll-like receptor Tollo|Serine incorporator 3") ~ "Immune Function", -->
<!--     str_detect(protein_names,  -->
<!--                     "Neurocalcin homolog|Puratrophin-1|Gamma-aminobutyric acid receptor|Orcokinin|Amyloid-beta|Teneurin|Multiple PDZ domain protein|Carboxypeptidase|Innexin|Cytoplasmic protein NCK2|cytokine signaling|Wnt|cAMP|Jerky|Adhesion G protein-coupled receptor") ~ "Signaling",  -->
<!--     str_detect(protein_names, "Cold shock|Bloom syndrome protein") ~ "Stress response", -->
<!--     TRUE ~ "Other")) %>%  -->
<!--   filter(my_category=="Transposable Elements") %>% View() -->
<!--   filter(is.na(gene_ontology_i_ds)) -->
<!-- ``` -->


<!-- # Plot genes from interesting categories  -->

<!-- ```{r} -->
<!-- x <- res.pco2.AL %>% as.data.frame() %>% dplyr::select(log2FoldChange, padj) %>% -->
<!--   rownames_to_column("gene") %>%  -->
<!--   filter(gene %in% critical.low) %>% -->
<!--   left_join(P.camt.blast.GO %>%  -->
<!--               dplyr::select(gene_id, gene_ontology_i_ds, protein_names, spid, -->
<!--                             gene_names, organism, gene_ontology_biological_process),  -->
<!--             by = c("gene"="gene_id")) %>% filter(!is.na(protein_names)) %>% -->
<!--   mutate(my_category=case_when( -->
<!--     str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein") ~ "Transposable Elements", -->
<!--     str_detect(protein_names,  -->
<!--                "Exportin-5|methyltransferase|demethylase|Histone deacetylase") ~ "Epigenetic Factor", -->
<!--     str_detect(protein_names,  -->
<!--                     "Zinc finger protein|LIM domain-binding protein 2") ~ "Transcription Regulation",  -->
<!--     str_detect(protein_names,  -->
<!--                "Toll-like receptor Tollo|Serine incorporator 3") ~ "Immune Function", -->
<!--     str_detect(protein_names,  -->
<!--                     "Neurocalcin homolog|Puratrophin-1|Gamma-aminobutyric acid receptor|Orcokinin|Amyloid-beta|Teneurin|Multiple PDZ domain protein|Carboxypeptidase|Innexin|Cytoplasmic protein NCK2|cytokine signaling|Wnt|cAMP|Jerky|Adhesion G protein-coupled receptor") ~ "Signaling",  -->
<!--     str_detect(protein_names, "Cold shock|Bloom syndrome protein") ~ "Stress response", -->
<!--     TRUE ~ "Other"))  -->

<!-- y <- x %>% filter(my_category=="Transposable Elements") %>% select(gene) %>% unlist() %>% as.vector() -->

<!-- diff_plots = list() -->
<!-- for (i in 1:length(y)) { -->
<!-- a <-  plotCounts(dds.DESeq.pH, gene=y[i], intgroup=c("Treatment"), returnData = TRUE) %>%  -->
<!--   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low"))) -->

<!-- b <- ggplot(a %>% rownames_to_column("sample"), -->
<!--        aes(x=Treatment, y=count, color=Treatment, label=sample)) + -->
<!--   #geom_boxplot(outlier.shape = NA) + -->
<!--   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) + -->
<!--     #geom_text(size=2) + -->
<!--     theme_bw() + -->
<!--     ggtitle(y[i]) + -->
<!--     theme(axis.title.x = element_blank(),  -->
<!--           plot.title = element_text(hjust = 0.5, size=10),  -->
<!--           legend.position = "none") -->
<!-- diff_plots[[i]] <- b -->
<!-- } -->
<!-- diff_plots -->
<!-- ``` -->


<!-- # Here is the average+/-SD for each gene  -->
<!-- # THIS ISN'T INTERESTING - I SHOULD PLOT TES ~ EPIGENETIC STUFF  -->

<!-- ```{r} -->
<!-- # plot mean counts for ambient, moderate, and low pH  -->
<!-- genes.stats %>% pivot_longer(cols = contains("L2FC"), names_to = "contrast", values_to = "L2FC") %>% -->
<!--   filter(gene %in% c(critical.low)) %>%  -->
<!--   #filter(abs(L2FC)>1) %>%  -->
<!--   dplyr::select(-contrast, -L2FC, -contains("padj")) %>% unique() %>%  -->
<!--     mutate(pco2=Treatment) %>% -->
<!--   mutate_at("pco2", function(x) { -->
<!--     case_when( -->
<!--       x == "Ambient" ~ 371, -->
<!--       x == "Moderate" ~ 704, -->
<!--       x == "Low" ~ 1424)}) %>% -->
<!--   filter(str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein")) %>% -->
<!--   #mutate(treat.num=as.numeric(Treatment)) %>%  #need to change treatment factor to numeric for plotting  -->
<!--   ggplot(aes(x=pco2, y=mean, group=gene)) + geom_point() + geom_line(size=.01) + -->
<!--   theme_minimal() + #facet_wrap(~change) + -->
<!--   #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) + -->
<!--   theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) + -->
<!--   xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of transposable elements by treatment & relative expression") -->
<!-- ``` -->
<!-- ```{r} -->
<!-- assay(vsd.pH) %>% as.data.frame() %>% rownames_to_column("gene") %>% -->
<!--   left_join(P.camt.blast.GO %>%  -->
<!--               dplyr::select(gene_id, gene_ontology_i_ds, protein_names, spid, -->
<!--                             gene_names, organism, gene_ontology_biological_process),  -->
<!--             by = c("gene"="gene_id")) %>% filter(!is.na(protein_names)) %>% -->
<!--   mutate(my_category=case_when( -->
<!--     str_detect(protein_names,  -->
<!--     "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein") ~ "Transposable Elements", -->
<!--     str_detect(protein_names,  -->
<!--                "Exportin-5|methyltransferase|demethylase|Histone deacetylase") ~ "Epigenetic Factor", TRUE ~ "Other")) %>% -->
<!--   #filter(gene %in% critical.low) %>% -->
<!--   filter(my_category!="Other") %>% -->
<!--   pivot_longer(cols = starts_with("Tank"), names_to = "Sample", values_to = "counts.vsd") %>% -->
<!--     left_join(sample.info %>% dplyr::select(Sample, Treatment), by="Sample") %>% -->
<!--   ggplot(aes(x=my_category, y=counts.vsd)) + geom_violin() + facet_wrap(~Treatment) -->
<!-- ``` -->

## For all DEGs - Reaction Norm Plot

```{r}
# plot mean counts for ambient, moderate, and low pH 

genes.stats %>% pivot_longer(cols = contains("L2FC"), names_to = "contrast", values_to = "L2FC") %>%
  filter(change!="Non-DEG") %>% 
  filter(abs(L2FC)>1) %>% 
  dplyr::select(-contrast, -L2FC, -contains("padj")) %>% unique() %>% 
    mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  #mutate(treat.num=as.numeric(Treatment)) %>%  #need to change treatment factor to numeric for plotting 
  ggplot(aes(x=pco2, y=mean, group=gene)) + geom_line(size=.01) +
  theme_minimal() + facet_wrap(~change) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")
```


```{r}
# # NOTE: When I do NOT filter out low-frequency genes prior to the DESeq2 analysis using the mean=10 threshold, some cv=0 for some genes/treatments
# # So I will remove the CV=0 genes before testing for differences 
# f %>% filter(cv!=0)
# hist(log((f %>% filter(cv!=0))$cv))
# summary(aov(log(cv) ~ DEG*Treatment, data=f %>% filter(cv!=0)))
# TukeyHSD(aov(log2(cv) ~ DEG+Treatment+DEG:Treatment, data=f %>% filter(cv!=0)))
# 
# # # Here you can plot the normalized gene counts for those genes. 
# # cv.zero <- (f %>% filter(cv==0))$gene %>% as.vector()
# # diff_plots = list()
# # for (i in 1:length(cv.zero)) {
# # a <-  plotCounts(dds.DESeq.pH, gene=cv.zero[i], intgroup=c("Treatment"), returnData = TRUE)
# # b <- ggplot(a %>% filter(Treatment==c("Ambient", "Moderate", "Low")) 
# #                          %>% rownames_to_column("sample"),
# #        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
# #   geom_point(position=position_jitter(w = 0.1,h = 0)) +
# #     geom_text() +
# #     theme_bw() +
# #     ggtitle(cv.zero[i]) +
# #     theme(plot.title = element_text(hjust = 0.5))
# # diff_plots[[i]] <- b
# # }
# # diff_plots
# 
# # Octopamine receptor, SPID Q7JQF1 = g114647
# 
# # evm.TU.HiC_scaffold_78.1 - Methylcytosine dioxygenase TET1 (EC 1.14.11.n2) (CXXC-type zinc finger protein 6) 
# a <-  plotCounts(dds.DESeq.pH, gene="g114647", intgroup=c("Treatment"), returnData = TRUE) %>% 
#   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
#   
# ggplot(a %>% rownames_to_column("sample"),
#        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
#   #geom_boxplot(outlier.shape = NA) +
#   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=2) +
#     theme_bw() +
#     #ggtitle("Octopamine") +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none")

```


```{r}
# plot mean counts for ambient, moderate, and low pH 
genes.stats %>% pivot_longer(cols = contains("L2FC"), names_to = "contrast", values_to = "L2FC") %>%
  #filter(gene %in% c(critical.low)) %>% 
  #filter(abs(L2FC)>1) %>% 
  filter(change == "Up-regulated") %>% 
  dplyr::select(-contrast, -L2FC, -contains("padj")) %>% unique() %>% 
    mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  filter(str_detect(protein_names, 
    "retrotransposable|LINE|transposable element|transposon|mobile element jockey|Pol polyprotein")) %>%
  #mutate(treat.num=as.numeric(Treatment)) %>%  #need to change treatment factor to numeric for plotting 
  ggplot(aes(x=Treatment, y=cv)) + geom_boxplot() + 
  theme_minimal() + #facet_wrap(~change) +
  #scale_x_continuous(breaks=c(1,2,3), labels=c("Ambient", "Moderate", "Low")) +
  theme(axis.title.x = element_blank(), plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per treatment mean of normalized counts") + ggtitle("Abundances of transposable elements by treatment & relative expression")

```


## Prep for GO_MWU analysis 

## IMPORTANT: For each DEG list I multiply the L2FC by -1 so that a positive LFC (>0) indicates more abundant in the OA treatment, and a negative LFC (<0) indicates less abundant in OA treatment. The DESeq2 results object reports everything relative to the treatment that comes first alphabetically, which is not what I want for the GO_MWU dendrogram. 

```{r files for GO_MWU}
# BACKGROUND - all genes submitted to DESeq2 analysis 
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>%
  filter(!is.na(gene_ontology_i_ds)) %>%
  write_delim("../results/GO_MWU/DESeq-genes_for-GOMWU.tab",delim = "\t")

# DEGs between Ambient & Moderate
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
  filter(!is.na(gene_ontology_i_ds)) %>% dplyr::select(gene, log2FoldChange) %>% 
  mutate(log2FoldChange=-1*log2FoldChange) %>%
  write.csv("../results/GO_MWU/DEGs_amb-mod.csv", quote = F,row.names = F)

res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  dplyr::select(gene,log2FoldChange, padj) %>%
  filter(padj<0.05) %>% ggplot() + geom_histogram(aes(x=log2FoldChange))

# DEGs between Ambient & Low
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
  filter(!is.na(gene_ontology_i_ds)) %>% dplyr::select(gene, log2FoldChange) %>%
  mutate(log2FoldChange=-1*log2FoldChange) %>%
  write.csv("../results/GO_MWU/DEGs_amb-low.csv", quote = F,row.names = F)

# DEGs between Moderate & Low
res.pco2.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% dplyr::select(gene,log2FoldChange) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>% 
  filter(!is.na(gene_ontology_i_ds)) %>% dplyr::select(gene, log2FoldChange) %>%
  mutate(log2FoldChange=-1*log2FoldChange) %>%
  write.csv("../results/GO_MWU/DEGs_mod-low.csv", quote = F,row.names = F)


```

Prep for GO_MWU using standard enrichment test (presence/absence, not using LFG values)
```{r}
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  mutate(significant= case_when(
    padj<0.05 ~ 1,
    TRUE ~ 0)) %>%
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>%
  filter(!is.na(gene_ontology_i_ds)) %>% 
  dplyr::select(gene, significant) %>% 
  write.csv("../results/GO_MWU/DEGs_amb-mod_STANDARD.csv", quote = F,row.names = F)


res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  mutate(significant= case_when(
    padj<0.05 ~ 1,
    TRUE ~ 0)) %>%
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>%
  filter(!is.na(gene_ontology_i_ds)) %>% 
  dplyr::select(gene, significant) %>% 
  write.csv("../results/GO_MWU/DEGs_amb-low_STANDARD.csv", quote = F,row.names = F)

res.pco2.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  mutate(significant= case_when(
    padj<0.05 ~ 1,
    TRUE ~ 0)) %>%
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds), by = c("gene"="gene_id")) %>%
  filter(!is.na(gene_ontology_i_ds)) %>% 
  dplyr::select(gene, significant) %>% 
  write.csv("../results/GO_MWU/DEGs_mod-low_STANDARD.csv", quote = F,row.names = F)
```


# Generating 2 bubble plots for Figure in manuscript - 1) Less Active, 2) More Active 
Figure 6 A and B 

```{r}
# Biological Processes DOWNREGULATED in OA compared to Ambient

pdf(file="../results/Figure6B-downregulated.pdf", height = 5.35, width = 6)

enriched.bp.degs %>% 
#enriched.keywords.degs %>%
  filter(p_value<0.05) %>%
  group_by(deg_set, genes, count) %>%
  summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(
   enriched.bp.degs %>% 
#    enriched.keywords.degs %>%
      select(deg_set, category, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("deg_set", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(deg_set, p_value) %>% group_by(deg_set) %>% dplyr::slice(1:20) %>%
  separate(deg_set, sep = "-", into = c("contrast", "OA.response"), remove = FALSE) %>%
  mutate(up.treatment=deg_set) %>%
  mutate_at("up.treatment", function(x) {
    case_when(
      x == "AL-DOWN" ~ "Ambient",
      x == "AL-UP" ~ "Severe",
      x == "AM-DOWN" ~ "Ambient",
      x == "AM-UP" ~ "Moderate",
      x == "ML-DOWN" ~ "Moderate",
      x == "ML-UP" ~ "Severe")}) %>%
  mutate_at(c("deg_set", "OA.response", "up.treatment"), factor) %>%
  mutate(contrast=factor(contrast, levels=c("AM", "AL", "ML"))) %>% 
  mutate(up.treatment=factor(up.treatment, levels=c("Severe", "Moderate", "Ambient"))) %>% 
  ungroup() %>%
  arrange(up.treatment, contrast, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  
  filter(up.treatment=="Ambient") %>%
  droplevels() %>%

ggplot(aes(y = process, x=contrast, color=contrast)) + 
  geom_point(aes(size=-log10(p_value)), alpha=0.75) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, 
            labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Downregulated\nBiological Processes"))) +
  scale_color_manual(name="Upregulated", values=c(AM="#fdae61", AL="#d7191c"),
             guide="none") +
  #guide = guide_legend(override.aes = list(size=4))) +
  # scale_alpha("FDR-adjusted\nP-value", range = c(.8,0.2), #breaks = c(0, .25, .5, .75, 1),
  #             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2.75,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_blank(), #element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.ticks.x = element_blank(), 
        legend.position = "right", 
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.35, label = "Moderate OA\n(pH 7.8)", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -.35, label = "Severe OA\n(pH 7.3)", size=2.5, col="gray15") +
#  annotate("text", x = 3, y = -.45, label = "Moderate OA\nvs. Severe OA", size=2, col="gray15") +
coord_cartesian(clip = "off", ylim = c(-.7, 32))
# oord_cartesian(clip = "off", ylim = c(-0.6, 20))
dev.off()
```


```{r}
# Biological Processes UPREGULATED in OA compared to Ambient

pdf(file="../results/Figure6A-upregulated.pdf", height = 4.25, width = 5.5)

enriched.bp.degs %>% 
#enriched.keywords.degs %>%
  filter(p_value<0.05) %>%
  group_by(deg_set, genes, count) %>%
  summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(
    enriched.bp.degs %>% 
#    enriched.keywords.degs %>%
      select(deg_set, category, go, process, count, percent, p_value, fold_enrichment, fdr, genes), by = c("deg_set", "genes", "p_value", "count")) %>%  #re-add data
  ungroup() %>% arrange(deg_set, p_value) %>% group_by(deg_set) %>% dplyr::slice(1:20) %>%
  separate(deg_set, sep = "-", into = c("contrast", "OA.response"), remove = FALSE) %>%
  mutate(up.treatment=deg_set) %>%
  mutate_at("up.treatment", function(x) {
    case_when(
      x == "AL-DOWN" ~ "Ambient",
      x == "AL-UP" ~ "Severe",
      x == "AM-DOWN" ~ "Ambient",
      x == "AM-UP" ~ "Moderate",
      x == "ML-DOWN" ~ "Moderate",
      x == "ML-UP" ~ "Severe")}) %>%
  mutate_at(c("deg_set", "OA.response", "up.treatment"), factor) %>%
  mutate(contrast=factor(contrast, levels=c("AM", "AL", "ML"))) %>% 
  mutate(up.treatment=factor(up.treatment, levels=c("Severe", "Moderate", "Ambient"))) %>% 
  ungroup() %>%
  arrange(up.treatment, contrast, p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  
  filter(up.treatment!="Ambient") %>%
  filter(contrast!="ML") %>%
  droplevels() %>%

ggplot(aes(y = process, x=contrast, color=up.treatment)) + 
  geom_point(aes(size=-log10(p_value)), alpha=0.75) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2, 
            labeller = as_labeller(c(`GOTERM_BP_DIRECT`="Upregulated\nBiological Processes"))) +
  scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Severe="#d7191c"),
             #guide = guide_legend(override.aes = list(size=4))) +
             guid="none") +
  # scale_alpha("FDR-adjusted\nP-value", range = c(.8,0.2), #breaks = c(0, .25, .5, .75, 1),
  #             guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2.75,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_blank(), #element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.ticks.x = element_blank(), 
        legend.position = "right", 
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.35, label = "Moderate OA\n(pH 7.8)", size=2.5, col="gray15") +
  annotate("text", x = 2, y = -.35, label = "Severe OA\n(pH 7.3)", size=2.5, col="gray15") +
coord_cartesian(clip = "off", ylim = c(-.7, 22.5))
dev.off()
```

```{r}
enriched.bp.degs %>% filter(deg_set=="AL-UP", p_value<0.05) %>% arrange(p_value) %>% View()
```

