---
title: "xx-WGCNA"
author: "Laura H Spencer"
date: "2/28/2022"
output: html_document
---

```{r}
 # Check memory limit
memory.limit()
 # Change memory limit
memory.limit(size = 50000)
```


```{r, message=FALSE, warning=FALSE, results=FALSE}
# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("tidyverse", "RColorBrewer", "ggpmisc", "ggpubr", "ape", "pheatmap", "plotly", "clipr")
# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("WGCNA", "DESeq2")

# Install BiocManager if needed
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})
```

Format of expression data should be: 
- rownames = sample 
- columns = genes 
- normalized 

# Load in filtered datasets also used for DESeq2 

```{r}
# load data
load(file = "../results/bowtie/counts.trans") #counts.trans
load(file = "../results/bowtie/per-gene-summary-stats") #genes.stats
load(file = "../results/deseq2/vsd.pH") #vsd.pH

# Read in sample metadata
load(file="../data/sample.info") #sample.info
```

```{r}
# Load annotation objects 
load(file = "../references/P.plat.blast")
load(file = "../references/P.plat.blast.GO")
load(file = "../references/P.plat.bestblast")

# load in DEG lists 
load(file = "../results/deseq2/diffex.AM")
load(file = "../results/deseq2/diffex.AL")
load(file = "../results/deseq2/diffex.ML")

# Load in DESeq2 results lists to pull background list of genes 
load(file = "../results/deseq2/res.pco2.AM") #res.pco2.AM
load(file = "../results/deseq2/res.pco2.AL") #res.pco2.AL
load(file = "../results/deseq2/res.pco2.ML") #res.pco2.ML

# Load in DESeq2 results prepped for IPA analysis 
load(file = "../results/IPA/IPA.degs")
```

```{r}
# Use WGCNA's `goodSamplesGenes` to remove any genes with very low variability or missing data  
datExpr0 <- counts.trans[,-1:-3]
gsg = goodSamplesGenes(datExpr0, verbose = 3);
sum(gsg$goodGenes==FALSE)  #were any removed?  Nope. 
ncol(gsg$allOK)

if (!gsg$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0) 
     printFlush(paste("Removing genes:", paste(names(datExpr0)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0) 
     printFlush(paste("Removing samples:", paste(rownames(datExpr0)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr0 = datExpr0[gsg$goodSamples, gsg$goodGenes]
}
```

```{r}
sampleTree = hclust(dist(datExpr0), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
#pdf(file = "Plots/sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2)
# Plot a line to show the cut
abline(h = 110, col = "red");

# Determine cluster under the line
clust = cutreeStatic(sampleTree, cutHeight = 110, minSize = 10)  #for bowtie2 aligned
#clust = cutreeStatic(sampleTree, cutHeight = 125, minSize = 10)  #for star aligned
table(clust)
# clust 1 contains the samples we want to keep.
keepSamples = (clust==1)
datExpr = datExpr0[keepSamples, ]
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)

# Decision March 16th - don't throw out Tank4_Crab1, it's not an outlier on the PCAs (in DESeq2 notebook) 
```

```{r}
# traitData = read.csv("wgcna-testing/ClinicalTraits.csv");
# dim(traitData)
# names(traitData)


# remove columns that hold information we do not need.
(allTraits = counts.trans[, 1:3] %>% rownames_to_column("Sample") %>%
  mutate(pH=Treatment) %>%
  mutate_at("pH", function(x) {
    case_when(
      x == "Ambient" ~ 8.0,
      x == "Moderate" ~ 7.8,
      x == "Low" ~ 7.6)}) %>%
  mutate(H_exp=(10^(-1*pH)) %>% formatC(., format = "e", digits = 2)) %>%
   mutate(pco2=Treatment) %>%
   mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}))

dim(allTraits)
names(allTraits)

# # Form a data frame analogous to expression data that will hold the traits.
# 
Samples = rownames(datExpr);
traitRows = match(Samples, allTraits$Sample);
datTraits = allTraits[traitRows, -1];
rownames(datTraits) = allTraits[traitRows, 1];

# 
collectGarbage();
```

```{r}
# Re-cluster samples
sampleTree2 = hclust(dist(datExpr), method = "average")
# Convert traits to a color representation: white means low, red means high, grey means missing entry
traitColors = numbers2colors(sapply(datTraits,as.numeric), signed = F, colors = colorRampPalette(brewer.pal(8, "Set2"))(14));

# Plot the sample dendrogram and the colors underneath.
plotDendroAndColors(sampleTree2, traitColors,
                    groupLabels = names(datTraits), 
                    main = "Sample dendrogram with treatment and tank ids")
```

```{r}
options(stringsAsFactors = FALSE);
```

"Constructing a weighted gene network entails the choice of the soft thresholding power Î² to which co-expression
similarity is raised to calculate adjacency. The authors of [1] have proposed to choose the soft thresholding power based on the criterion of approximate scale-free topology." 

I will choose the power 9, which is the lowest power for which the scale-free topology fit
index reaches 0.90.

```{r}
# Choose a set of soft-thresholding powers
powers = c(c(1:25))
# Call the network topology analysis function
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5, networkType = "signed")
# Plot the results:
sizeGrWindow(9, 5)
par(mfrow = c(1,2));
cex1 = 0.9;
# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red");
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col="red")
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
```

I now calculate the adjacencies, using the soft thresholding power 14:

```{r}
softPower = 14; # for bowtie2 aligned 
#softPower = 20; #for star aligned
#softPower = 10; #for star + RSEM
adjacency = adjacency(datExpr, power = softPower, type="signed", corFnc="bicor");

# for testing purposes, subset expression dataframe 
#adjacency = adjacency(datExpr[,1:2000], power = softPower, type="signed", corFnc="bicor");
```

To minimize effects of noise and spurious associations, we transform the adjacency into Topological Overlap Matrix, and calculate the corresponding dissimilarity:

```{r}
# TOM = TOMsimilarity(adjacency); #this takes a very long time
# dissTOM = 1-TOM

#save(dissTOM, file = "../results/wgcna/dissTom-bowtie")
load(file = "../results/wgcna/dissTom-bowtie") #load if needed


## dissTOM objects generated fro various STAR-aligned pipelines
#save(dissTOM, file = "../results/wgcna/dissTom-star-rsem")
#save(dissTOM, file = "../results/wgcna/dissTom-star-multimap")
#save(dissTOM, file = "../results/wgcna/dissTom-star")

#load(file = "../results/wgcna/dissTom-star-rsem")
#load(file = "../results/wgcna/dissTom-star-multimap") #load if needed 
#load(file = "../results/wgcna/dissTom-star") #load if needed 
```

We now use hierarchical clustering to produce a hierarchical clustering tree (dendrogram) of genes. Note that we use the function hclust that provides a much faster hierarchical clustering routine than the standard hclust function.

```{r}
# Call the hierarchical clustering function
geneTree = hclust(as.dist(dissTOM), method = "average");
# Plot the resulting clustering tree (dendrogram)
sizeGrWindow(12,9)
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
     labels = FALSE, hang = 0.04);
```

```{r}
# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 30;
# Module identification using dynamic tree cut:
dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM,
                deepSplit = 2, pamRespectsDendro = FALSE,
                minClusterSize = minModuleSize);
table(dynamicMods)
```

```{r}
# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene clustering on TOM-based dissimilarity\nDendrogram and module colors")
```

The Dynamic Tree Cut may identify modules whose expression profiles are very similar. It may be prudent to merge
such modules since their genes are highly co-expressed. To quantify co-expression similarity of entire modules, we
calculate their eigengenes and cluster them on their correlation.

We choose a height cut of 0.25, corresponding to correlation of 0.75, to merge. 


```{r}
# Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")

MEDissThres = 0.25
# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")
# Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;
```


```{r}
#sizeGrWindow(12, 9)
#pdf(file = "Plots/geneDendro-3.pdf", wi = 9, he = 6)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene clustering on TOM-based dissimilarity\nDendrogram and module colors")
# #dev.off()
sizeGrWindow(12, 9)
plotDendroAndColors(geneTree, mergedColors,
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene clustering on TOM-based dissimilarity\nFinalized modules")
```



```{r}
# Rename to moduleColors
moduleColors = mergedColors
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;
MEs = mergedMEs;
# Save module colors and labels for use in subsequent parts
#save(MEs, moduleLabels, moduleColors, geneTree, file = "../results/wgcna/networkConstruction-stepByStep.RData")
```


```{r}
# Define numbers of genes and samples
nGenes = ncol(datExpr);
nSamples = nrow(datExpr);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs = orderMEs(MEs0)

# Calculate correlation between eigengenes and treatment 
#moduleTraitCor = cor(MEs, datTraits[,c("pH"), drop=F], use = "p");
moduleTraitCor = cor(MEs, datTraits[,c("pco2"), drop=F], use = "p");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);

# Correlation among module eigengenes 
#cor(MEs0, use="complete.obs") %>% corrplot::corrplot(tl.cex=.75, diag = F)

# library(PerformanceAnalytics)
# chart.Correlation(MEs0[,], histogram=F, pch=19)
```

```{r}
sizeGrWindow(10,6)
# Will display correlations and their p-values
# textMatrix =  paste(signif(moduleTraitCor, 2), "\n(",
#                            signif(moduleTraitPvalue, 1), ")", sep = "");
textMatrix =  paste(signif(moduleTraitCor, 2), " (",
                           signif(moduleTraitPvalue, 2), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
#               xLabels = names(datTraits[,c("pH"), drop=F]),
               xLabels = names(datTraits[,c("pco2"), drop=F]),
               yLabels = names(MEs) %>% str_remove(pattern = "ME"),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.7,
               cex.main = 0.85,
               cex.lab.y = 0.75,
               cex.lab.x= 0.85,
               zlim = c(-1,1),
               main = paste("Module-trait relationships"))
```
```{r}
# Define variable containing the pH column
# pH = as.data.frame(datTraits$pH);
# names(pH) = "pH"
pco2 = as.data.frame(datTraits$pco2);
names(pco2) = "pco2"


# names (colors) of the modules
modNames = substring(names(MEs), 3)

geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));

names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");

#geneTraitSignificance = as.data.frame(cor(datExpr, pH, use = "p"));
geneTraitSignificance = as.data.frame(cor(datExpr, pco2, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));

# names(geneTraitSignificance) = paste("GS.", names(pH), sep="");
# names(GSPvalue) = paste("p.GS.", names(pH), sep="");
names(geneTraitSignificance) = paste("GS.", names(pco2), sep="");
names(GSPvalue) = paste("p.GS.", names(pco2), sep="");
```


```{r}
# Need to have every gene represented in the annotation file - here I merge the best blast (has all genes) & uniprot blast (some genes) results to have all genes, and to also be able to use either blast result. 
annot = left_join(P.plat.bestblast, P.plat.blast, by=c("id_gtf"="geneID")) %>% 
  filter(id_gtf %in% names(datExpr))
dim(annot)
names(annot)
probes = names(datExpr)
probes2annot = match(probes, annot$id_gtf)
# The following is the number or probes without annotation:
sum(is.na(probes2annot))
# Should return 0.
```

```{r}
# Create the starting data frame
geneInfo0 = data.frame(id = probes,
                      hit_acc = annot$hit_acc[probes2annot],
                      hit_desc = annot$hit_desc[probes2annot],
                      moduleColor = moduleColors,
                      geneTraitSignificance,
                      GSPvalue)
# Order modules by their significance for pH
#modOrder = order(-abs(cor(MEs, pH, use = "p")));
modOrder = order(-abs(cor(MEs, pco2, use = "p")));

# Add module membership information in the chosen order
for (mod in 1:ncol(geneModuleMembership))
{
  oldNames = names(geneInfo0)
  geneInfo0 = data.frame(geneInfo0, geneModuleMembership[, modOrder[mod]], 
                         MMPvalue[, modOrder[mod]]);
  names(geneInfo0) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
                       paste("p.MM.", modNames[modOrder[mod]], sep=""))
}
# Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance
#geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.pH));
geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.pco2));

geneInfo = geneInfo0[geneOrder, ]
write.csv(geneInfo, file = "../results/wgcna/geneInfo.csv")
save(geneInfo, file="../results/wgcna/geneInfo")
```

Visualizing the network of eigengenes

```{r}
# # Recalculate module eigengenes
# MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes
# # Add pH treatment to existing module eigengenes
# MET = orderMEs(cbind(MEs, pH))
# # Plot the relationships among the eigengenes and pH
# 
# # Plot the dendrogram
# sizeGrWindow(6,6);
# par(cex = 1.0)
# plotEigengeneNetworks(MET, "Eigengene dendrogram", marDendro = c(0,4,2,0),
# plotHeatmaps = FALSE)
# 
# # Plot the heatmap matrix (note: this plot will overwrite the dendrogram plot)
# par(cex = 1.0)
# plotEigengeneNetworks(MET, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
# plotDendrograms = FALSE, xLabelsAngle = 90)
# 
# 
# # Barplots of eigengenes by sample for each pH-associated module
# # !!!! BROKEN - BARS AREN'T COLOR-CODED CORRECTION
# sizeGrWindow(8,7);
# which.module="red"
# ME=MEs[, paste("ME",which.module, sep="")]
# 
# all(rownames(datTraits)==rownames(MEs)) # check order of samples treatment and eigengene dfs are same. Yes.
# 
# par(mar=c(5, 6, 0, 0.7))
# barplot(ME, col=numbers2colors(as.numeric(datTraits$Treatment)), main="", cex.main=2,
# xlab="eigengene expression",ylab=NULL, names=rownames(MEs), horiz=T, las=1, cex.names=0.65)
# legend("topright", legend=rev(c(levels(datTraits$Treatment))), fill=rev(c(unique(numbers2colors(as.numeric(datTraits$Treatment))))))
# 

```

# departure from WGCNA tutorials 

## Generate line plots of genes for each module of interest

```{r}
# Generate per-gene summary stats 
# first extract the normalized counts
gene.counts.norm <- assay(vsd.pH) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  pivot_longer(-gene, values_to = "counts.vsd", names_to = "Sample") %>%
  left_join(sample.info[,c("Sample", "Treatment")]) %>% 
  mutate_at(c("gene", "Sample"), factor) %>%
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low"))) %>% 
  filter(Sample %in% rownames(datExpr)) # <-- important, keep only samples that were retained in WGCNA analysis 

# calculate mean, sd, cv and annotate 
genes.stats <- gene.counts.norm %>% group_by(Treatment, gene) %>% 
  summarize(mean=mean(counts.vsd), median=median(counts.vsd), sd=sd(counts.vsd), cv=sd(counts.vsd)/mean(counts.vsd)) %>% 
  mutate(DEG=if_else(gene %in% c(rownames(diffex.AM),rownames(diffex.AL), rownames(diffex.ML)), "DEG", "Non-DEG")) %>%
  mutate(change = case_when(
  gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange < 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange < 0))) ~ "Up-regulated",
  gene %in% c(rownames(subset(res.pco2.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.pco2.AL, padj < 0.05 & log2FoldChange > 0))) ~ "Down-regulated")) %>% 
  mutate(change=replace_na(change,"Non-DEG")) %>% 
  mutate(DEG=factor(DEG, levels=c("Non-DEG", "DEG"))) %>% 
  mutate(change=factor(change, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  ungroup() %>% 
  left_join(P.plat.bestblast %>% dplyr::select(id, id_gtf, hit_acc, hit_desc, e_value), by=c("gene"="id_gtf")) %>%
  left_join(P.plat.blast.GO %>% dplyr::select(gene_id, spid, protein_names, evalue), by=c("gene"="gene_id"))

# Generate line plots of significant modules showing only genes significantly correlated to module 
# List of modules (their color IDs) that are significantly associated with pH 
#modules <- moduleTraitPvalue %>% as.data.frame() %>% filter(pH<0.05) %>% rownames() %>% str_remove(pattern = "ME")
modules <- cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>% filter(pvalue<0.05) %>% arrange(corr) %>%
  rownames() %>% str_remove(pattern = "ME")
modules.p <- paste("p.MM.", modules, sep = "")
modules.cor <- noquote(paste("MM.", modules, sep = ""))

save(modules, file="../results/wgcna/modules")
save(modules.p, file="../results/wgcna/modules.p")
save(modules.cor, file="../results/wgcna/modules.cor")

# create empty list to store summary statistics for each modules' sign. genes 
module.stats <- vector("list", length(modules))
names(module.stats) <- modules

# create empty list to store line plots for each modules' sign. genes 
module.lineplots <- vector("list", length(modules))
names(module.lineplots) <- modules

for (i in 1:length(modules)) {
a <- geneInfo %>%
  filter(moduleColor==modules[i]) %>% 
#  filter(p.GS.pH<0.05) %>% filter(!!as.symbol(modules.p[i])<0.05) %>% 
  filter(p.GS.pco2<0.05) %>% filter(!!as.symbol(modules.p[i])<0.05) %>% 
#  dplyr::select(id, moduleColor, GS.pH, p.GS.pH, !!as.symbol(modules.p[i]), !!as.symbol(modules.cor[i])) %>% 
  dplyr::select(id, moduleColor, GS.pco2, p.GS.pco2, !!as.symbol(modules.p[i]), !!as.symbol(modules.cor[i])) %>% 
  left_join(genes.stats, by = c("id"="gene")) %>% 
  #group_by(Treatment, moduleColor) %>% mutate(mean_z = scale(mean)) %>% ungroup() %>% #use this line of code to z-transform 
  mutate(pH=Treatment) %>%
  mutate_at("pH", function(x) {
    case_when(
      x == "Ambient" ~ 8.0,
      x == "Moderate" ~ 7.8,
      x == "Low" ~ 7.6)}) %>%
  mutate(H_exp=(10^(-1*pH)) %>% formatC(., format = "e", digits = 2)) %>%
   mutate(pco2=Treatment) %>%
   mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)})

#formula <- y ~ x
#colors <- c("brown","pink","cyan2", "red","cyan4","green","darkgreen")
module.lineplots[[i]] <-  
  #ggplot(a, aes(x=pH, y=median)) + 
  ggplot(a, aes(x=pco2, y=median)) + 
  geom_line(aes(group=id), size=.01, color=modules[i]) +
  theme_minimal(base_size = 8) + facet_wrap(~moduleColor + change) + 
  geom_smooth(method="loess",se = FALSE, color="blue", size=1) + #  method=lm, formula = formula,
  #scale_x_continuous(breaks=c(7.6, 7.8, 8.0)) + 
  theme(plot.title = element_text(size=8), axis.text.x = element_text(hjust = 0.2), 
        axis.title.x = element_text(size=6)) +
  #xlab("pH Treatment") #+
  xlab("pCO2 concentration") #+
  # stat_poly_eq(aes(label = ..eq.label..),
  #              label.x.npc = 0.5, label.y.npc = 0.99,
  #              formula = formula, parse = TRUE, size = 3)

module.stats[[i]] <- a %>% dplyr::select(-Treatment, -mean, -sd, -cv, -pH) %>% distinct()
}
print(module.lineplots)
save(module.lineplots, file = "../results/wgcna/module.lineplots")

#Save some figures to pdf
# fig <- ggarrange(plotlist=module.lineplots,
#                        ncol=2, nrow=2,
#                        common.legend = TRUE, legend="top")
# pdf("../results/wgcna/pH-associated-modules-genes.pdf", onefile = TRUE)
# fig
# dev.off()

```
Summary of modules significantly associated with pCO2 
```{r}
moduleTraitCor %>% nrow()
modules %>% length()

cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>% filter(pvalue<0.05) %>% arrange(corr)
```


# For pH-associated modules, look at correlation between module membership and gene significane for each gene 

```{r}
par(mfrow = c(4,2));

for (i in 1:length(modules)) {
module = modules[i]
column = match(module, modNames);
moduleGenes = moduleColors==module;

#sizeGrWindow(7, 7);
verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
                   abs(geneTraitSignificance[moduleGenes, 1]),
                   xlab = paste("Module Membership in", module, "module"),
                   ylab = "Gene significance for pH",
                   main = paste("Module membership vs. gene significance\n"),
                   cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
}

```

# Which metric to use for IPA? 

```{r}
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.brown, p.MM.brown) %>%
#   filter(moduleColor %in% modules) %>% 
#   filter(p.MM.brown<0.05) %>%
#   ggplot(aes(x=MM.brown, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Brown"))
# 
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.cyan, p.MM.cyan) %>%
#   filter(moduleColor %in% modules) %>%
#   filter(p.MM.cyan<0.05) %>%
#   ggplot(aes(x=MM.cyan, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Cyan"))
# 
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.darkgreen, p.MM.darkgreen) %>%
#   filter(moduleColor %in% modules) %>%
# #  filter(p.MM.darkgreen<0.05) %>%
#   ggplot(aes(x=MM.darkgreen, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Darkgreen"))
# 
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.green, p.MM.green) %>%
#   filter(moduleColor %in% modules) %>%
# #  filter(p.MM.green<0.05) %>%
#   ggplot(aes(x=MM.green, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Green"))
# 
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.lightcyan1, p.MM.lightcyan1) %>%
#   filter(moduleColor %in% modules) %>%
# #  filter(p.MM.lightcyan1<0.05) %>%
#   ggplot(aes(x=MM.lightcyan1, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Lightcyan1"))
# 
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.pink, p.MM.pink) %>%
#   filter(moduleColor %in% modules) %>%
# #  filter(p.MM.pink<0.05) %>%
#   ggplot(aes(x=MM.pink, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Pink"))
# 
# ggplotly(geneInfo %>% dplyr::select(id, hit_acc, hit_desc, moduleColor, GS.pH, p.GS.pH, MM.red, p.MM.red) %>%
#   filter(moduleColor %in% modules) %>%
#   filter(p.MM.red<0.05) %>%
#   ggplot(aes(x=MM.red, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   scale_fill_manual(values=sort(modules)) + 
#   ggtitle("Red"))
# 
# # How does module membership stat for each color vary for genes assigned to that color vs. those that are not 
# ggplotly(geneInfo %>% 
#            mutate(x=if_else(moduleColor == "brown", "yes", "no")) %>%
#   ggplot(aes(x=MM.brown, group=x, fill=x)) + geom_density(alpha=0.5) + 
#   ggtitle("Brown"))
# 
# ggplotly(geneInfo %>% 
#            mutate(x=if_else(moduleColor == "cyan", "yes", "no")) %>%
#   ggplot(aes(x=MM.cyan, group=x, fill=x)) + geom_density(alpha=0.5) + 
#   ggtitle("cyan"))
# 
# ggplotly(geneInfo %>% 
#            mutate(x=if_else(moduleColor == "red", "yes", "no")) %>%
#   ggplot(aes(x=MM.red, group=x, fill=x)) + geom_density(alpha=0.5) + 
#   ggtitle("Red"))
# 
# ggplotly(geneInfo %>% 
#            mutate(x=if_else(moduleColor == "red", "yes", "no")) %>%
#   ggplot(aes(x=MM.red, group=x, fill=x)) + geom_density(alpha=0.5) + 
#   ggtitle("Red"))
# 
# ggplotly(geneInfo %>% 
#            mutate(x=if_else(moduleColor == "green", "yes", "no")) %>%
#   ggplot(aes(x=MM.green, group=x, fill=x)) + geom_density(alpha=0.5) + 
#   ggtitle("green"))
```

# Which metric to use for IPA? 
Here is gene signficance for pH (GS.pH) by module color - certainly looks like the pH-associated modules have larger abs(GS.pH) and much lower p.GS.pH. Good candidate!. 

```{r}
# ggplotly(geneInfo %>% 
# #  filter(moduleColor %in% modules) %>% 
# #  filter(p.MM.brown<0.05) %>%
#   ggplot(aes(x=GS.pH, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   #scale_fill_manual(values=sort(modules)) + 
#   ggtitle("GS.pH by module"))
# 
# ggplotly(geneInfo %>% 
# #  filter(moduleColor %in% modules) %>% 
# #  filter(p.MM.brown<0.05) %>%
#   ggplot(aes(x=p.GS.pH, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) + 
#   #scale_fill_manual(values=sort(modules)) + 
#   ggtitle("p.GS.pH by module"))
```

# Look at overlap among WGCNA and DESeq2 results 

## What percentage of DEGs are captured by pH-associated WGCNA modules? Particularly interested in genes I will use for enrichment analysis, so I will filter genes for: 
- Only genes whose abundances are significantly correlated with pH  (i.e. Gene Significance (GS) p-value, p.GS, <0.05)
- NO ---> Only genes whose module membership (MM) is significant (p.MM < 0.05). 


```{r}
# First how many DEGs do I have total? 
n.DEGs <- genes.stats %>% filter(DEG=="DEG") %>% dplyr::select(gene) %>% distinct() %>% nrow()

# merge module association info with DEG info 
geneInfo %>%
  left_join(genes.stats %>% dplyr::select(gene, DEG) %>% distinct(), by = c("id"="gene")) %>% #join with table indicating DEG
  mutate(moduleColor = as.factor(moduleColor)) %>%
  #group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.pH<0.05), not.sign=sum(p.GS.pH>0.05)) %>% 
  group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.pco2<0.05), not.sign=sum(p.GS.pco2>0.05)) %>% 
  filter(moduleColor %in% modules) %>% 
  filter(DEG=="DEG") %>% 
  ungroup() %>% dplyr::select(n, sign) %>% colSums()/n.DEGs*100 

# replace "modules" filter with  c(modules, "turquose", "darkgrey", "tan") to include peak/valley modules 

# ANSWER: 
# 81.4% of the DEGs were assigned to a pH-associated module. 
# 73.9% of the DEGs were both assigned to a pH-associated module, AND those genes are sign. associated with pH (p.GS.pH<0.05)

```
## What percentage of DEGs are also identified as correlated with pH via WGCNA (and assigned to any module)

```{r}
geneInfo %>%
  left_join(genes.stats %>% dplyr::select(gene, DEG) %>% distinct(), by = c("id"="gene")) %>% #join with table indicating DEG
  mutate(moduleColor = as.factor(moduleColor)) %>%
#  group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.pH<0.05), not.sign=sum(p.GS.pH>0.05)) %>% 
  group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.pco2<0.05), not.sign=sum(p.GS.pco2>0.05)) %>% 
  filter(DEG=="DEG") %>% 
  ungroup() %>% dplyr::select(n, sign, not.sign) %>% colSums()/n.DEGs*100 

# 85% of DEGs are deemed to be significantly correlated with pH
```

## How many genes are correlated with pH (any module?)

```{r}
geneInfo %>%
  mutate(moduleColor = as.factor(moduleColor)) %>%
#  group_by(moduleColor) %>% summarise(n=n(), sign=sum(p.GS.pH<0.05), not.sign=sum(p.GS.pH>0.05)) %>% 
  group_by(moduleColor) %>% summarise(n=n(), sign=sum(p.GS.pco2<0.05), not.sign=sum(p.GS.pco2>0.05)) %>% 
  dplyr::select(moduleColor, n, sign, not.sign) %>% 
  mutate(perc.sig=sign/n) %>% 
  dplyr::select(sign, not.sign) %>% colSums()
```


# Venn diagram of DEGs and genes in WGCNA pH-associated modules 

```{r}
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

module.genes <- vector("list", length(modules)+1)
names(module.genes) <- c(modules, "all")
for (i in 1:length(modules)) {
module.genes[[i]] <- geneInfo %>%
  # filter for genes assigned to pH-associated modules 
  filter(moduleColor==modules[i]) %>% 
## further filter for genes that are significantly associated with pH and significantly assigned to their respective module
  #filter(p.GS.pH<0.05) %>% 
  #filter(!!as.symbol(modules.p[i])<0.05) %>%
## alternative approach to just select "hub genes"   
  #    filter(abs(GS.pH)>0.2) %>% filter(!!as.symbol(modules.cor[i])>0.8) %>% 
  rownames()
}
module.genes[[length(module.genes)]] <- Reduce(c, module.genes)

x <- list(
  AvsM = row.names(diffex.AM), 
  AvsL = row.names(diffex.AL), 
  MvsL = row.names(diffex.ML),
  WGCNA = module.genes[["all"]]
  )

display_venn(
        x,
        category.names = c("DEGs\nAmbient vs.\nModerate" , "DEGs\nAmbient vs.\nLow" , "DEGs\nModerate vs.\nLow", "WGCNA\npH-associated\nmodules"),
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = c("#999999", "#E69F00", "#56B4E9", "#009E73"),
        # Numbers
        cex = .9,
        fontface = "italic",
        # Set names
        cat.cex = .9,
        cat.fontface = "bold",
        cat.default.pos = "outer")#,
        #cat.dist = c(0.06, 0.06, 0.06))

# EFFECT OF VARIOUS FILTERING SETTINGS: 
# When I include all genes assigned to pH-associated modules, the % of DEGs included in WGCNA modules is 80.1%
# When I filter for only genes that are significantly associated with pH (p.GS.pH<0.05), that drops to 73.3%
# When I filter for genEs that are sign. associated with pH (p.GS.pH<0.05) AND module membership is sign. (p.MM<0.05) that drops to 72.2%
# Because I am using all genes assigned to a pH-associated module in the GO Enrichment analysis and IPA, I shall use the least filtered set. 

```

Now look for overlap among DEGs and each module 

```{r}
par(mfrow = c(4,2));
for (i in 1:length(modules)) {
  y <- list(
  AvsM = row.names(diffex.AM), 
  AvsL = row.names(diffex.AL), 
  MvsL = row.names(diffex.ML),
  WGCNA = module.genes[[i]]
  )
  
  display_venn(
        y,
        category.names = c("DEGs\nAmbient vs.\nModerate" , "DEGs\nAmbient vs.\nLow" , "DEGs\nModerate vs.\nLow", 
                           paste("WGCNA module: ", modules[i], "\npH-associated\nmodules", sep="")),
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = c("#999999", "#E69F00", "#56B4E9", modules[i]),
        # Numbers
        cex = .9,
        fontface = "italic",
        # Set names
        cat.cex = .9,
        cat.fontface = "bold",
        cat.default.pos = "outer")
}
```

# How many genes assigned to each pH-associated module?

```{r}
geneInfo %>%  
  filter(moduleColor %in% modules) %>%
  dplyr::count(moduleColor)

# How many total genes assigned to any pH-associated module? 
geneInfo %>%  
  filter(moduleColor %in% modules) %>%
  dplyr::count(moduleColor) #%>% dplyr::select(n) %>% colSums()

#9298/18150

# How many total genes assigned to any pH-associated module AND have high Gene Significance (significantly associated with pH)? 
geneInfo %>%  
  filter(moduleColor %in% modules) %>%
#  filter(p.GS.pH<0.05) %>%
  filter(p.GS.pco2<0.05) %>%
  dplyr::count(moduleColor) # %>% dplyr::select(n) %>% colSums()

```


# Generate heatmaps of pH-associated modules

```{r}
# create empty list to store line plots for each modules' sign. genes 
counts.modules <- vector("list", length(modules))
names(counts.modules) <- modules

#filter(abs(GS.pH)>0.2) %>% filter(!!as.symbol(modules.cor[i])>0.8)

for (i in 1:length(modules)) {
  counts.modules[[i]] <- 
    (assay(vsd.pH))[(geneInfo %>% 
                       filter(moduleColor==modules[i]) %>% 
                       filter(p.GS.pco2<0.05))$id,] %>% 
#                       filter(p.GS.pH<0.05))$id,] %>% 
    as.data.frame() %>% add_column(module=modules[i])
}

s <- sample.info %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low"))) %>% 
           arrange(factor(Treatment, levels=c("Ambient", "Moderate", "Low"))) %>% 
  remove_rownames() %>% column_to_rownames("Sample")

# define order of modules from most sign. positive correlation <--> most sign. negative correlation 
modules.sorted <- moduleTraitCor %>% as.data.frame() %>% 
  filter(rownames(.) %in% 
#           rownames(moduleTraitPvalue %>% as.data.frame() %>% filter(pH<0.05))) %>% 
           rownames(moduleTraitPvalue %>% as.data.frame() %>% filter(pco2<0.05))) %>% 
  #  arrange(pH) %>% rownames() %>% str_remove(pattern = "ME")
  arrange(pco2) %>% rownames() %>% str_remove(pattern = "ME")

counts.modules.matrix<- bind_rows(counts.modules[modules.sorted]) %>% 
  dplyr::select(module, rownames(s)) %>%
  mutate(module=factor(module, levels=modules.sorted))

pheatmap(counts.modules.matrix[,colnames(counts.modules.matrix) !="module"], 
         cluster_cols = FALSE, cluster_rows=F, annotation_names_col = FALSE, 
         show_rownames=FALSE, na.rm=TRUE, scale="row", 
         main = "Expression by pH-associated module\n(significant genes only)", fontsize = 8, 
         annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"),
                                  module=c(cyan="cyan", pink="pink", brown="brown", 
                                           lightcyan1="lightcyan1", green="green", darkgreen="darkgreen")),
         # annotation_colors = list(Treatment=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"),
         #                           module=c(magenta="magenta", orangered4="orangered4", darkgreen="darkgreen",
         #                                    plum1="plum1", salmon="salmon",
         #                                    darkgrey="darkgrey")),
         annotation_col=s[,"Treatment", drop=F],
         annotation_row = counts.modules.matrix[,"module", drop=FALSE])
```

## Prep for GO_MWU analysis 

#### Extract module membership values for each gene for each pH-associated module 


```{r files for GO_MWU}
# Generate a gene annotation file 
geneInfo.mwu <- geneInfo %>% left_join(
  (P.plat.blast.GO)[,c("gene_id","gene_ontology_i_ds")], by = c("id"="gene_id")) %>%  # could also opt to  %>% filter(evalue < 1e-10)
  filter(!is.na(gene_ontology_i_ds)) #filter for genes that are annotated with GO information 
write_delim(geneInfo.mwu[,c("id", "gene_ontology_i_ds")], "../results/GO_MWU/WGCNA-genes_for-GOMWU.tab",delim = "\t")

# Write out separate tables for each module that contain two columns: 1) gene id, and 2) module membership, i.e. significance. For each module, that column contains MM for genes assigned to that module, and 0 for genes not assigned to that module. 
geneInfo.mwu.mods <- vector("list", length(modules))
names(geneInfo.mwu.mods) <- modules

for (i in 1:length(modules)) {
  geneInfo.mwu.mods[[i]] <- geneInfo.mwu %>%
    mutate(sign = ifelse(moduleColor == modules[[i]],!!as.symbol(modules.cor[i]), 0)) %>%
    dplyr::select(id,sign) %>% remove_rownames()
  write.csv(geneInfo.mwu.mods[[i]], paste("../results/GO_MWU/WGCNA-module_",modules[[i]],".csv", sep = ""), quote = F,row.names = F)
}

```

## Prep for IPA

```{r}
modules
modules.p
modules.cor
```
NOTE: For WGCNA results I export the MM or GS and associated p-values. The p.MM is directly derived from MM, therefore p.MM is a sufficient metric for all IPA decisions (sign. genes, AND how to handle multi-hit). EXCEPT when I want to visualize the directionality of the changes (up/down regulated / correlated). That's when the GS and MM (which can be -1<x<1) come in handy. 

```{r}
ggplotly(ggplot(geneInfo, aes(y=MM.cyan, x=p.MM.cyan)) + geom_point() + ggtitle("abs(MM) ~ p.MM relationship"))

# what about Gene Significance stat? 
ggplotly(ggplot(geneInfo, aes(y=GS.pco2, x=p.GS.pco2)) + geom_point() + ggtitle("abs(MM) ~ p.MM relationship"))
```
```{r}
IPA.wgcna <- geneInfo %>% 
  dplyr::select(id,
         # GS.pH, p.GS.pH,
         GS.pco2, p.GS.pco2,
         MM.brown, p.MM.brown, 
         MM.pink, p.MM.pink, 
         MM.lightcyan1, p.MM.lightcyan1, 
         MM.cyan, p.MM.cyan, 
         MM.green, p.MM.green, 
         MM.darkgreen, p.MM.darkgreen) %>% remove_rownames()

save(IPA.wgcna, file = "../results/wgcna/IPA.wgcna")

# Merge DEG stats and WGCNA stats, so I make sure to have the same gene list in 
IPA.input <- left_join(IPA.degs, IPA.wgcna, by=c("geneID"="id")) %>%
  left_join(P.plat.bestblast %>% dplyr::select(id, id_gtf), by=c("geneID"="id_gtf")) %>%
  dplyr::select(id, everything(), -geneID)

# Save to file for IPA 

IPA.input %>% dplyr::select(id, colnames(IPA.degs[-1])) %>%
write_delim(file = "../results/IPA/RKC-for-IPA_DEGs.tab", delim = "\t",
              col_names = TRUE, quote = "none")

IPA.input %>% dplyr::select(id, colnames(IPA.wgcna[-1])) %>%
write_delim(file = "../results/IPA/RKC-for-IPA_WGCNA.tab", delim = "\t",
              col_names = TRUE, quote = "none")

```

What about the modules - that aren't sign. associated with pH? any patterns there? E.g. do some modules contain genes that are higher at moderate pH? 

Let's just look at ALL modules: 

```{r}
# Generate line plots of ALL modules showing only genes significantly correlated to module 

modules.all <- moduleTraitPvalue %>% as.data.frame() %>% rownames() %>% str_remove(pattern = "ME")
modules.all.p <- paste("p.MM.", modules.all, sep = "")
modules.all.cor <- noquote(paste("MM.", modules.all, sep = ""))

# create empty list to store summary statistics for each modules' sign. genes 
modules.all.stats <- vector("list", length(modules.all))
names(modules.all.stats) <- modules.all

# create empty list to store line plots for each modules' sign. genes 
modules.all.lineplots <- vector("list", length(modules.all))
names(modules.all.lineplots) <- modules.all

for (i in 1:length(modules.all)) {

a <- geneInfo %>%
  filter(moduleColor==modules.all[i]) %>% 
  filter(!!as.symbol(modules.all.p[i])<0.05) %>% 
#  dplyr::select(id, moduleColor, GS.pH, p.GS.pH, !!as.symbol(modules.all.p[i]), !!as.symbol(modules.all.cor[i])) %>% 
  dplyr::select(id, moduleColor, GS.pco2, p.GS.pco2, !!as.symbol(modules.all.p[i]), !!as.symbol(modules.all.cor[i])) %>% 
  left_join(genes.stats, by = c("id"="gene")) %>% 
  #group_by(Treatment, moduleColor) %>% mutate(mean_z = scale(mean)) %>% ungroup() %>% #use this line of code to z-transform 
  #filter(moduleColor %in% modules) %>% 
  mutate(pH=Treatment) %>%
  mutate_at("pH", function(x) {
    case_when(
      x == "Ambient" ~ 8.0,
      x == "Moderate" ~ 7.8,
      x == "Low" ~ 7.6)}) %>%
  mutate(H_exp=(10^(-1*pH)) %>% formatC(., format = "e", digits = 2)) %>%
   mutate(pco2=Treatment) %>%
   mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)})

#formula <- y ~ x

modules.all.lineplots[[i]] <-  
#  ggplot(a, aes(x=pH, y=median)) + 
  ggplot(a, aes(x=pco2, y=median)) + 
  geom_line(aes(group=id), size=.01, color=modules.all[i]) +
  theme_minimal() + facet_wrap(~moduleColor + change) + 
  geom_smooth(method="loess", se = FALSE, color="blue", size=1) + #method=lm, formula = formula, 
  #scale_x_continuous(breaks=c(7.6, 7.8, 8.0)) + 
  theme(plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("pH Treatment") #+
  # stat_poly_eq(aes(label = ..eq.label..), 
  #              label.x.npc = 0.5, label.y.npc = 0.99,
  #              formula = formula, parse = TRUE, size = 3)

modules.all.stats[[i]] <- a %>% dplyr::select(-Treatment, -mean, -sd, -cv, -pH) %>% distinct()
}
print(modules.all.lineplots)

modules.all.lineplots[["violet"]]
save(modules.all.lineplots, file = "../results/wgcna/modules.all.lineplots")
```

Modules of interest, as they have higher or lower expression at moderate pH:
- mediumpurple3, saddlebrown

For Go MWU
```{r}
# Write out separate tables for each module that contain two columns: 1) gene id, and 2) module membership, i.e. significance. For each module, that column contains MM for genes assigned to that module, and 0 for genes not assigned to that module. 
modules.interest <- c("mediumpurple3", "saddlebrown")
modules.interest.cor <- noquote(paste("MM.", modules.interest, sep = ""))

geneInfo.mwu.mods.interest <- vector("list", length(modules.interest))
names(geneInfo.mwu.mods.interest) <- modules.interest
for (i in 1:length(modules.interest)) {
  geneInfo.mwu.mods.interest[[i]] <- geneInfo.mwu %>%
    mutate(sign = ifelse(moduleColor == modules.interest[[i]],!!as.symbol(modules.interest.cor[i]), 0)) %>%
    dplyr::select(id,sign) %>% remove_rownames()
  write.csv(geneInfo.mwu.mods.interest[[i]], paste("../results/GO_MWU/WGCNA-module_",modules.interest[[i]],".csv", sep = ""), quote = F,row.names = F)
}
```

# BONEYARD
-------------------------


```{r}
# Here I try to make lineplots with delta counts (from ambient) - very ugly and not informative, but this code might be useful
genes.stats %>% ungroup() %>%
  filter(gene %in% names(datExpr)[moduleColors==module]) %>% 
  dplyr::select(gene, Treatment, mean, hit_acc, hit_desc, SPID) %>%
  pivot_wider(names_from = "Treatment", values_from = "mean") %>% 
  mutate(amb=0, mod=Moderate-Ambient, low=Low-Ambient) %>%
  dplyr::select(-Ambient, -Moderate, -Low) %>%
  pivot_longer(cols = c(amb, mod, low), names_to = "pH", values_to = "delta") %>%
  mutate_at("pH", function(x) {
    case_when(
      x == "amb" ~ 8.0,
      x == "mod" ~ 7.8,
      x == "low" ~ 7.6)}) %>% 
  ggplot(aes(x=pH, y=delta, group=gene)) + geom_line(size=.01) +
  theme_minimal() + #facet_wrap(~change) +
  scale_x_continuous(breaks=c(7.6, 7.8, 8.0)) +
  theme(plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  ylab("Per-treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression") 
```


```{r}

genes.stats %>% 
  left_join(geneInfo %>% dplyr::select(id, moduleColor, GS.pH, p.GS.pH), by = c("gene"="id")) %>% 
  #group_by(Treatment, moduleColor) %>% mutate(mean_z = scale(mean)) %>% ungroup() %>% #use this line of code to z-transform 
  filter(moduleColor %in% modules) %>% 
  mutate(pH=Treatment) %>%
  mutate_at("pH", function(x) {
    case_when(
      x == "Ambient" ~ 8.0,
      x == "Moderate" ~ 7.8,
      x == "Low" ~ 7.6)}) %>% 
  ggplot(aes(x=pH, y=log10(mean))) + geom_line(aes(group=gene), size=.01) +
  theme_minimal() + facet_wrap(~moduleColor+change) + geom_smooth(method=lm, se = FALSE, color="red", size=1.5) +
  scale_x_continuous(breaks=c(7.6, 7.8, 8.0)) + 
  theme(plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Per-treatment mean of normalized counts") + ggtitle("Abundances of DEGs by treatment & relative expression")

test <- genes.stats %>% filter(gene %in% names(datExpr)[moduleColors==module]) %>% 
  mutate(pH=Treatment) %>%
  mutate_at("pH", function(x) {
    case_when(
      x == "Ambient" ~ 8.0,
      x == "Moderate" ~ 7.8,
      x == "Low" ~ 7.6)})
hist(log(test$mean))
shapiro.test(log(test$mean))
summary(lm(log(mean)~pH, data=test))

# Try plotting expression level of all genes for all samples, organized by pH 
modules
```


```{r}
hierTOM = hclust(as.dist(dissTOM),method="average")
colorDynamicHybridTOM = labels2colors(cutreeDynamic(hierTOM, distM= dissTOM , cutHeight = 0.998,
deepSplit=2, pamRespectsDendro = FALSE))
colorh1= colorDynamicHybridTOM

cmd1=cmdscale(as.dist(dissTOM),2)
sizeGrWindow(7, 6)
par(mfrow=c(1,1))
plot(cmd1, col=as.character(colorh1), main="MDS plot",
xlab="Scaling Dimension 1", ylab="Scaling Dimension 2")

rownames(cmd1) <- colorh1


```


# If I want to start with raw count data and do my own outlier filtering 

```{r}
# counts <- data.frame(read.table("../results/featureCounts/featurecounts_redking_gene-v3", header = T, stringsAsFactors = F, fill = FALSE))
# counts <- counts %>%  column_to_rownames(var="Geneid")
# counts <- counts %>% rename_all(~as.character(str_sub(colnames(counts)) %>%
#                              gsub("X.scratch.lspencer.2022.redking.OA.aligned.bowtie.sorted.deduped.", "", .) %>%
#                              gsub(".sorted.bam", "", .)))
# 
# # Remove extraneous gene info, then transpose dataframe so each row = a sample (aka "objects"), and each column = genes (aka "variables") 
# counts.t <- as.data.frame(t(counts[,-1:-5])) #remove extraneous columns, transform data to have each sample a row, each column a gene 
# 
# # Read in sample metadata 
# sample.info <- read.csv("../data/RCK-exp-design.csv", header=T, na.strings = NA) %>%
#           mutate_at(vars(Tank, Treatment, Treatment_Tank), as.factor)
# 
# # filter low-frequency genes 
# keep1 <- colMeans(counts.t, na.rm=TRUE) >= 10 #identify genes with mean count across all samples (excluding NAs = 10)
# keep2 <- rowSums( counts >= 10 ) >= 0.2*43 #identify genes with counts>=10 across at minimum 25% of the samples
# keep <- unique(c(names(which(keep1 == T)), names(which(keep2 == T)))) # list of genes meeting one of the two above criteria
# counts.ts <- counts.t[,keep]
# 
# print(paste("# genes remaining after pre-filtering:", ncol(counts.ts)))
# print(paste("# of genes dropped:", ncol(counts.t) - ncol(counts.ts), sep=" "))
# 
# # merge count data with sample key, reset row names as sample names, and arrange by infection, then temperature, then day 
# counts.tk <- merge(x=sample.info[,c("Sample", "Tank", "Treatment", "Treatment_Tank")], by.x="Sample", y=counts.ts, by.y="row.names") %>% 
#   arrange(Treatment, Tank)  %>% column_to_rownames(var="Sample") %>% droplevels()
# 
# all(rownames(counts.tk) == counts.tk %>% dplyr::select(starts_with("evm")) %>% t() %>% colnames()) #check that rownames of untransformed matrix match column names of transformed matrix. Should print 'TRUE' 
# 
# # Use DESeq2 to perform a variance-stabilizing transformation 
# dds.pH <- DESeqDataSetFromMatrix(countData = counts.tk[,grepl("evm", colnames(counts.tk))] %>% t(),
#                               colData = counts.tk[,"Treatment", drop=FALSE] ,
#                               design = ~ Treatment)
# vsd.pH <- varianceStabilizingTransformation(dds.pH, blind=FALSE)
# 
# counts.trans <- assay(vsd.pH) %>% t() %>% as.matrix() %>% as.data.frame() %>%
#   rownames_to_column(var = "Sample") %>% 
#   left_join(sample.info[,c("Sample", "Treatment", "Tank", "Treatment_Tank")]) %>%
#   column_to_rownames(var = "Sample") %>% 
#   select(Treatment, Tank, Treatment_Tank, contains("evm")) %>% 
#   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
# save(counts.trans, file = "../results/wgcna/counts.filt")
```


```{r}
read_delim(file = "../../../Downloads/RKC-results-for-IPA.tab", delim = "\t") %>% 
  summary(GS.pH)
```


