---
title: "04_GO-enrichment"
author: "Laura H Spencer"
date: "3/22/2022"
output: 
  html_document
    code_folding: hide
    toc: yes
    toc_depth: 6
    toc_float: yes
  pdf_document:
    keep_tex: yes
---

```{r}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("GSEABase")

require(stringr)
require(ggpubr)
require(tidyverse)
require(GSEABase)
require(ape)
library(splitstackshape) 
install.packages("splitstackshape")

```

In this notebook I use GO_MWU to perform a gene ontology enrichment analysis for gene sets of interest, which include 1) Differentially expressed genes (DEGs) derived from DESeq2, and 2) Modules of genes associated with pH, derived from WGCNA. The input files for each of the gene sets were generated in the DESeq2 and WGCNA notebooks. 

This notebook contains code from the GO_MWU program. 

```{r}
setwd("~/red-king_RNASeq-2022/results/GO_MWU")
```

```{r}
getwd()
```

```{r}
# goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
# source("gomwu.functions.R")
```

```{r}
# Term("GO:0042302")
```

Create a list of GO devisions, Biological Processes (BP), Molecular Functions (MF), Cellular Components (CC), to loop over 

```{r}
# goDivisions <- c("BP", "MF", "CC")
```

# NOTE:  I write separate GO_MWU functions for the DEG lists and the WGCNA module lists. This is because GO_MWU has different settings for those two types of analyses. 

# Identify enriched functions in genes differentially expressed among pH treatments (from DESeq2)

## Create a GO_MWU function for enrichment analysis of DEGs 

```{r}
# log(0.05, base = 10)
```
Create an empty list in which to save the best GO terms from each enrichment analysis. This list consitutes GO terms that best represent *independent* groups of significant GO terms

```{r}
# best_GO <- list()
```

```{r}
# go.mwu.degs <- function(contrast) {
#   
#   input = paste("DEGs_", contrast, ".csv", sep="")
#   goAnnotations = "DESeq-genes_for-GOMWU.tab"
# 
# for (j in 1:length(goDivisions)) {
#   tryCatch({
#   
# # ------------- Calculating stats
# # It might take a few minutes for MF and BP. Do not rerun it if you just want to replot the data with different cutoffs, go straight to gomwuPlot. If you change any of the numeric values below, delete the files that were generated in previos runs first.
# 
# gomwuStats(input, goDatabase, goAnnotations, goDivisions[j],
# 	perlPath="C:/Strawberry/perl/bin/perl.exe", # replace with full path to perl executable if it is not in your system's PATH already
# 	largest=0.75,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
# 	smallest=5,   # a GO category should contain at least this many genes to be considered
# 	clusterCutHeight=0.25, # threshold for merging similar (gene-sharing) terms. See README for details.
# #	Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead. 
# 	#Module=TRUE,Alternative="g" # un-remark this if you are analyzing a SIGNED WGCNA module (values: 0 for not in module genes, kME for in-module genes). In the call to gomwuPlot below, specify absValue=0.001 (count number of "good genes" that fall into the module)
# #	Module=TRUE # un-remark this if you are analyzing an UNSIGNED WGCNA module 
# )
# # do not continue if the printout shows that no GO terms pass 10% FDR.
# 
# 
# # ----------- Plotting results
# 
# #quartz()
# pdf(file = paste("GO-tree_", contrast, "_", goDivisions[j], ".pdf", sep=""), height = length(results[[2]]$labels)/3)
# results=gomwuPlot(input,goAnnotations,goDivisions[j],
#  	#absValue=-log(0.05,10),   # genes with the measure value exceeding this will be counted as "good genes". This setting is for signed log-pvalues. Specify absValue=0.001 if you are doing Fisher's exact test for standard GO enrichment or analyzing a WGCNA module (all non-zero genes = "good genes").
#  absValue=.5, # un-remark this if you are using log2-fold changes
#  level1=0.05,level2=0.01,level3=0.001,	#stricter settings for many GO terms 
#  #level1=0.1, # FDR threshold for plotting. Specify level1=1 to plot all GO categories containing genes exceeding the absValue.
#  	#level2=0.05, # FDR cutoff to print in regular (not italic) font.
#  	#level3=0.01, # FDR cutoff to print in large bold font.
#  	txtsize=1,    # decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
#  	treeHeight=0.5, # height of the hierarchical clustering tree
#  	#colors=c(color, color, color, color, color)
#  colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") # these are default colors, un-remar and change if needed
#  )
# mtext(paste(contrast, " - ", goDivisions[j], sep=""), side = 3, line = 3, adj = -.05, col="black")
# dev.off()
#  # manually rescale the plot so the tree matches the text 
#  # if there are too many categories displayed, try make it more stringent with level1=0.05,level2=0.01,level3=0.001.  
#  
#  # text representation of results, with actual adjusted p-values
#  print(results[[1]])
# 
# 
# # ------- extracting representative GOs
# 
# # this module chooses GO terms that best represent *independent* groups of significant GO terms
# 
# pcut=1e-2 # adjusted pvalue cutoff for representative GO
# hcut=0.9 # height at which cut the GO terms tree to get "independent groups". 
# 
# # plotting the GO tree with the cut level (un-remark the next two lines to plot)
# # plot(results[[2]],cex=0.6)
# # abline(h=hcut,col="red")
# 
# # cutting
# ct=cutree(results[[2]],h=hcut)
# annots=c();ci=1
# for (ci in unique(ct)) {
#   message(ci)
# 	rn=names(ct)[ct==ci]
# 	obs=grep("obsolete",rn)
# 	if(length(obs)>0) { rn=rn[-obs] }
# 	if (length(rn)==0) {next}
# 	rr=results[[1]][rn,]
# 	bestrr=rr[which(rr$pval==min(rr$pval)),]
# 	best=1
# 	if(nrow(bestrr)>1) {
# 		nns=sub(" .+","",row.names(bestrr))
# 		fr=c()
# 		for (i in 1:length(nns)) { fr=c(fr,eval(parse(text=nns[i]))) }
# 		best=which(fr==max(fr))
# 	}
# 	if (bestrr$pval[best]<=pcut) { annots=c(annots,sub("\\d+\\/\\d+ ","",row.names(bestrr)[best]))}
# }
# 
# mwus=read.table(paste("MWU",goDivisions[j],input,sep="_"),header=T)
# bestGOs=mwus[mwus$name %in% annots,]
# print(bestGOs)
# 
# # Save best go terms to list for use outside of function
# best_GO[[paste(contrast, goDivisions[j], sep="_")]] <<- bestGOs 
# },
# error=function(e){cat("ERROR:", conditionMessage(e), "\n")})
# }
# }
```


# NOTE: in dendrograms, RED=upregulated, BLUE=downregulated 

## DEGs between ambient conditions (pH ~8.0) and moderate OA (pH ~7.8)

```{r}
#go.mwu.degs(contrast="amb-mod")
```

```{r}
#best_GO$`amb-mod_BP`
```


## DEGs between ambient conditions (pH ~8.0) and severe OA (pH ~7.5)

```{r}
#go.mwu.degs(contrast="amb-low")
```

## DEGs between moderate conditions (pH ~8.0) and severe OA (pH ~7.5)

```{r}
#go.mwu.degs(contrast="mod-low")
```

# Identify enriched functions in genes within modules that are signficantly associated with pH (from WGCNA) 

The "color" argument refers to the color for each GO tree (so for some modules, it's best to make that a slightly different color)

```{r}
# go.mwu.wgcna <- function(file, module, color) {
#   
#   input=file # two columns of comma-separated values: gene id, continuous measure of significance. To perform standard GO enrichment analysis based on Fisher's exact test, use binary measure (0 or 1, i.e., either sgnificant or not).
#   
#   goAnnotations.wgcna="WGCNA-genes_for-GOMWU.tab" # two-column, tab-delimited, one line per gene, multiple GO terms separated by semicolon. If you have multiple lines per gene, use nrify_GOtable.pl prior to running this script.
# 
# for (j in 1:length(goDivisions)) {
#   tryCatch({
#   
# # ------------- Calculating stats
# # It might take a few minutes for MF and BP. Do not rerun it if you just want to replot the data with different cutoffs, go straight to gomwuPlot. If you change any of the numeric values below, delete the files that were generated in previos runs first.
# 
#     gomwuStats(input, goDatabase, goAnnotations.wgcna, goDivisions[j],
# 	perlPath="perl", # replace with full path to perl executable if it is not in your system's PATH already
# 	largest=0.90,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
# 	smallest=3,   # a GO category should contain at least this many genes to be considered
# 	clusterCutHeight=0.25, # threshold for merging similar (gene-sharing) terms. See README for details.
# #	Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead. 
# 	Module=TRUE,Alternative="g" # un-remark this if you are analyzing a SIGNED WGCNA module (values: 0 for not in module genes, kME for in-module genes). In the call to gomwuPlot below, specify absValue=0.001 (count number of "good genes" that fall into the module)
# #	Module=TRUE # un-remark this if you are analyzing an UNSIGNED WGCNA module 
# )
# # do not continue if the printout shows that no GO terms pass 10% FDR.
# 
# # ----------- Plotting results
# 
# #quartz()
# results=gomwuPlot(input,goAnnotations.wgcna,goDivisions[j],
#  	absValue=0.001,  #-log(0.05,10) # genes with the measure value exceeding this will be counted as "good genes". This setting is for signed log-pvalues. Specify absValue=0.001 if you are doing Fisher's exact test for standard GO enrichment or analyzing a WGCNA module (all non-zero genes = "good genes").
#  #	absValue=1, # un-remark this if you are using log2-fold changes
#  	level1=0.1, # FDR threshold for plotting. Specify level1=1 to plot all GO categories containing genes exceeding the absValue.
#  	level2=0.05, # FDR cutoff to print in regular (not italic) font.
#  	level3=0.01, # FDR cutoff to print in large bold font.
#  	txtsize=1.2,    # decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
#  	treeHeight=0.5, # height of the hierarchical clustering tree
#  	#colors=c(color, color, color, color, color)
#  colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") # these are default colors, un-remar and change if needed
#  )
# mtext(paste(module, "-", goDivisions[j], sep=""), side = 3, line = .25, adj = 0, col=color) #module
#  # manually rescale the plot so the tree matches the text 
#  # if there are too many categories displayed, try make it more stringent with level1=0.05,level2=0.01,level3=0.001.  
#  
#  # text representation of results, with actual adjusted p-values
#  print(results[[1]])
# 
# 
# # ------- extracting representative GOs
# # this module chooses GO terms that best represent *independent* groups of significant GO terms
# 
# pcut=1e-2 # adjusted pvalue cutoff for representative GO
# hcut=0.9 # height at which cut the GO terms tree to get "independent groups". 
# 
# # plotting the GO tree with the cut level (un-remark the next two lines to plot)
# # plot(results[[2]],cex=0.6)
# # abline(h=hcut,col="red")
# 
# # cutting
# ct=cutree(results[[2]],h=hcut)
# annots=c();ci=1
# for (ci in unique(ct)) {
#   message(ci)
# 	rn=names(ct)[ct==ci]
# 	obs=grep("obsolete",rn)
# 	if(length(obs)>0) { rn=rn[-obs] }
# 	if (length(rn)==0) {next}
# 	rr=results[[1]][rn,]
# 	bestrr=rr[which(rr$pval==min(rr$pval)),]
# 	best=1
# 	if(nrow(bestrr)>1) {
# 		nns=sub(" .+","",row.names(bestrr))
# 		fr=c()
# 		for (i in 1:length(nns)) { fr=c(fr,eval(parse(text=nns[i]))) }
# 		best=which(fr==max(fr))
# 	}
# 	if (bestrr$pval[best]<=pcut) { annots=c(annots,sub("\\d+\\/\\d+ ","",row.names(bestrr)[best]))}
# }
# 
# mwus=read.table(paste("MWU",goDivisions[j],input,sep="_"),header=T)
# bestGOs=mwus[mwus$name %in% annots,]
# print(bestGOs)
# 
# # Save best go terms to list for use outside of function
# best_GO[[paste(contrast, goDivisions[j], sep="_")]] <<- bestGOs 
# },
# error=function(e){cat("ERROR:", conditionMessage(e), "\n")})
# }
# }
```

# pH-associated modules 
```{r}
load(file="../wgcna/modules")
modules
```


This is the list of modules that are associated with pH 
```{r}
load(file = "../wgcna/module.lineplots") #load lineplots to plot alongside enrichment to aid in interpretation 
load(file = "../wgcna/modules.all.lineplots")
```

```{r}
cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>% filter(pvalue<0.05) %>% arrange(corr)
```

## Module == `modules[[1]]`
DOWNREGULATED

```{r}
# # Line plot 
module.lineplots[[1]] # <--- NEED TO REMAKE WITH JUST DEGS
# 
# #go.mwu.wgcna(file=paste("WGCNA-module_", modules[[1]], ".csv", sep=""), module = modules[[1]], color = modules[[1]])
# #go.mwu.wgcna(file=paste("WGCNA-module_", modules[[1]], "_DEGs.csv", sep=""), module = modules[[1]], color = modules[[1]]) # genes in module that are DEGs 
# go.mwu.wgcna(file=paste("WGCNA-module_", modules[[1]], "_GS05.csv", sep=""), module = modules[[1]], color = modules[[1]]) # genes in module with p.GS<0.05 
```

## Background for DAVID 

```{r}
# Background for DAVID! 
geneInfo.mwu %>% 
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
## Using DAVID 

# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[1]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[1]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[2]]`
DOWNREGULATED

```{r}
module.lineplots[[2]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[2]], ".csv", sep=""), module = modules[[2]], color = modules[[2]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[2]], "_DEGs.csv", sep=""), module = modules[[2]], color = modules[[2]]) # genes in module that are DEGs
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[2]], "_GS05.csv", sep=""), module = modules[[2]], color = modules[[2]]) # genes in module with p.GS<0.05
```

# Test using DAVID - same results? 

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[2]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[2]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```


## Module == `modules[[3]]`
DOWNREGULATED

```{r}
module.lineplots[[3]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[3]], ".csv", sep=""), module = modules[[3]], color = modules[[3]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[3]], "_DEGs.csv", sep=""), module = modules[[3]], color = modules[[3]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[3]], "_GS05.csv", sep=""), module = modules[[3]], color = modules[[3]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[3]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[3]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[4]]`
DOWNREGULATED

```{r}
module.lineplots[[4]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[4]], ".csv", sep=""), module = modules[[4]], color = modules[[4]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[4]], "_DEGs.csv", sep=""), module = modules[[4]], color = modules[[4]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[4]], "_GS05.csv", sep=""), module = modules[[4]], color = modules[[4]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[4]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
#geneInfo.mwu %>% filter(moduleColor == modules[[4]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[5]]`

DOWNREGULATED

```{r}
module.lineplots[[5]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[5]], ".csv", sep=""), module = modules[[5]], color = modules[[5]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[5]], "_DEGs.csv", sep=""), module = modules[[5]], color = modules[[5]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[5]], "_GS05.csv", sep=""), module = modules[[5]], color = modules[[5]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[5]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[5]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[6]]`

UPREGULATED in moderate OA, DOWNREGULATED in severe OA

```{r}
module.lineplots[[6]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[6]], ".csv", sep=""), module = modules[[6]], color = modules[[6]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[6]], "_DEGs.csv", sep=""), module = modules[[6]], color = modules[[6]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[6]], "_GS05.csv", sep=""), module = modules[[6]], color = modules[[6]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[6]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
#geneInfo.mwu %>% filter(moduleColor == modules[[6]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[7]]`
UPREGULATED in moderate OA, DOWNREGULATED in severe OA

```{r}
module.lineplots[[7]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[7]], ".csv", sep=""), module = modules[[7]], color = modules[[7]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[7]], "_DEGs.csv", sep=""), module = modules[[7]], color = modules[[7]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[7]], "_GS05.csv", sep=""), module = modules[[7]], color = modules[[7]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[7]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[7]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[8]]`
DOWNREGULATED

```{r}
module.lineplots[[8]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[8]], ".csv", sep=""), module = modules[[8]], color = modules[[8]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[8]], "_DEGs.csv", sep=""), module = modules[[8]], color = modules[[8]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[8]], "_GS05.csv", sep=""), module = modules[[8]], color = modules[[8]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[8]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[8]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[9]]`

UPREGULATED 

```{r}
module.lineplots[[9]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[9]], ".csv", sep=""), module = modules[[9]], color = modules[[9]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[9]], "_DEGs.csv", sep=""), module = modules[[9]], color = modules[[9]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[9]], "_GS05.csv", sep=""), module = modules[[9]], color = modules[[9]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[9]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[9]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[10]]`

UPREGULATED 

```{r}
module.lineplots[[10]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[10]], ".csv", sep=""), module = modules[[10]], color = modules[[10]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[10]], "_DEGs.csv", sep=""), module = modules[[10]], color = modules[[10]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[10]], "_GS05.csv", sep=""), module = modules[[10]], color = modules[[10]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[10]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[10]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[11]]`

UPREGULATED

```{r}
module.lineplots[[11]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[11]], ".csv", sep=""), module = modules[[11]], color = modules[[11]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[11]], "_DEGs.csv", sep=""), module = modules[[11]], color = modules[[11]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[11]], "_GS05.csv", sep=""), module = modules[[11]], color = modules[[11]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[11]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[11]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[12]]`

UPREGULATED

```{r}
module.lineplots[[12]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[12]], ".csv", sep=""), module = modules[[12]], color = modules[[12]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[12]], "_DEGs.csv", sep=""), module = modules[[12]], color = modules[[12]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[12]], "_GS05.csv", sep=""), module = modules[[12]], color = modules[[12]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[12]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[12]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[13]]`

UPREGULATED

```{r}
module.lineplots[[13]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[13]], ".csv", sep=""), module = modules[[13]], color = modules[[13]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[13]], "_DEGs.csv", sep=""), module = modules[[13]], color = modules[[13]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[13]], "_GS05.csv", sep=""), module = modules[[13]], color = modules[[13]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[13]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[13]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## Module == `modules[[14]]`

UPREGULATED

```{r}
module.lineplots[[14]]

#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[14]], ".csv", sep=""), module = modules[[14]], color = modules[[14]]) # all genes in module
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[14]], "_DEGs.csv", sep=""), module = modules[[14]], color = modules[[14]]) # genes in module that are DEGs 
#go.mwu.wgcna(file=paste("WGCNA-module_", modules[[14]], "_GS05.csv", sep=""), module = modules[[14]], color = modules[[14]]) # genes in module with p.GS<0.05 
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == modules[[14]]) %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == modules[[14]]) %>% 
#  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

# Modules of interest - those that aren't sign. correlated woth pCO2 but are DOWN UP DOWN, i.e. higher only in moderate OA 

Modules of interest - those that contain a substantial portion of DEG 
c("darkseagreen4", "lightcyan1","black", "lightsteelblue", "tan", "sienna", "navajowhite2", "skyblue3")

```{r}
modules.all.lineplots[["darkseagreen4"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "darkseagreen4") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "darkseagreen4") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
modules.all.lineplots[["lightcyan1"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "lightcyan1") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "lightcyan1") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

c("darkseagreen4", "lightcyan1","black", "lightsteelblue", "tan", "sienna", "navajowhite2", "skyblue3")

```{r}
modules.all.lineplots[["black"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "black") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "black") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
modules.all.lineplots[["lightsteelblue"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "lightsteelblue") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "lightsteelblue") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```


```{r}
modules.all.lineplots[["tan"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "tan") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "tan") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
modules.all.lineplots[["sienna3"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "sienna3") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "sienna3") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
modules.all.lineplots[["navajowhite2"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "navajowhite2") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "navajowhite2") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

```{r}
modules.all.lineplots[["skyblue3"]]
```

```{r}
# Create vector of genes that is either a DEG or is associated with pCO2 in this module
z <- c(degs,
       (geneInfo.mwu %>% filter(moduleColor == "skyblue3") %>% filter(p.GS.pco2<0.05))$id)

# Copy Uniprot IDs to clipboard for DAVID
geneInfo.mwu %>% filter(moduleColor == "skyblue3") %>% 
  filter(id %in% z) %>%
  dplyr::select(SPID) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

## REALLY HELPFUL/CRITICAL STEP - for understanding which genes are super influential in determining the GO enrichment results

Here I identify genes that comprise the enriched GO Terms in each module. For example, the BP term "cell cycle progression" was enriched in the "purple" module, so I import the GO_MWU biological process results, which connects each gene ID to its GO Terms (along with the "value", which is the WGCNA module membership, where 0=not a member, and >0=membership coorelation). I then filter for only those genes with the "cell cycle process" term, and further filter for genes assigned to the "purple" module. This gives me a list of genes to examine for their function, to see what's driving the enrichment. 

Upregulated biological processes: 

Purple module: 
"cell cycle process"
"spindle organization"

Green module:
"nervous system development"
"system development"


```{r}
read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%  separate(Term, into = c("GO", "name"), sep = "~") %>% View()
filter(., grepl("intracellular protein transport",name)) %>% View()

# DAVID 
level <- "BP" #select BP, MF, or CC
MOD <- "pink"

# magenta
c("response to virus", "interaction with symbiont")

# pink
"positive regulation of microglial cell activation|proteolysis|response to xenobiotic stimulus"
#"viral entry into host cell|virion assembly|viral genome integration into host DNA"
#"proteolysis"
#"positive regulation of microglial cell activation|response to xenobiotic stimulus"

# coral1
"activation of innate immune response"

# plum1
"chaperone mediated protein folding requiring cofactor"


# get list of degs or genes with p.GS < 0.05
z <- unique(c(degs,
       (geneInfo.mwu %>% filter(moduleColor %in% c("magenta", "pink", "plum1")) %>%
          filter(p.GS.pco2<0.05))$id))

# get list of uniprot ids contributing to enriched GO term 

# MAGENTA
h <- read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%  separate(Term, into = c("GO", "name"), sep = "~") %>% 
filter(., grepl("response to virus",name)) %>%
select(Genes) %>% noquote() %>% 
cSplit(., 'Genes', ',') %>%
pivot_longer(cols = everything()) %>% select(value) %>% unlist() %>% as.vector()

# PINK
j <- read_delim(file = paste("DAVID_pink_DEGs-or-GS.txt", sep=""), delim = "\t") %>%  separate(Term, into = c("GO", "name"), sep = "~") %>% 
filter(., grepl("positive regulation of microglial cell activation",name)) %>%
select(Genes) %>% noquote() %>% 
cSplit(., 'Genes', ',') %>%
pivot_longer(cols = everything()) %>% select(value) %>% unlist() %>% as.vector()

# CORAL1
k <- read_delim(file = paste("DAVID_coral1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%  separate(Term, into = c("GO", "name"), sep = "~") %>% 
filter(., grepl("activation of innate immune response",name)) %>%
select(Genes) %>% noquote() %>% 
cSplit(., 'Genes', ',') %>%
pivot_longer(cols = everything()) %>% select(value) %>% unlist() %>% as.vector()

# Find gene ids 
q <- geneInfo.mwu %>% filter(moduleColor %in% c("magenta", "pink", "coral1")) %>% filter(id %in% z) %>% 
  filter(SPID %in% c(h,j,k)) #%>%
#  mutate(process="immune-response")


# CORAL1
m <- read_delim(file = paste("DAVID_plum1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%  separate(Term, into = c("GO", "name"), sep = "~") %>% 
filter(., grepl("chaperone mediated protein folding requiring cofactor",name)) %>%
select(Genes) %>% noquote() %>% 
cSplit(., 'Genes', ',') %>%
pivot_longer(cols = everything()) %>% select(value) %>% unlist() %>% as.vector()

n <- read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%  separate(Term, into = c("GO", "name"), sep = "~") %>% 
filter(., grepl("signal transduction",name)) %>%
select(Genes) %>% noquote() %>% 
cSplit(., 'Genes', ',') %>%
pivot_longer(cols = everything()) %>% select(value) %>% unlist() %>% as.vector()

r <- rbind(geneInfo.mwu %>% filter(moduleColor %in% c("ivory", "green", "lightgreen")) %>% filter(., grepl("Octopamine|tyramine receptor|mitogen-activated protein kinase|Orcokinin", protein_names)),
           geneInfo.mwu %>% filter(moduleColor %in% c("plum1", "green")) %>% filter(SPID %in% m)) # c(m,n)))
  
genes.stats %>% filter(gene %in% c(q$id, r$id)) %>% 
  left_join(rbind(q,r) %>% select(id, moduleColor), by=c("gene"="id")) %>%
  mutate(moduleColor = as.factor(moduleColor)) %>%
  mutate(process=case_when(
    moduleColor %in% unique(q$moduleColor) ~ "immune-response",
    moduleColor %in% unique(r$moduleColor) ~ "stress-response")) %>%

  #ggplot(aes(x=Treatment, y=median, colour=Treatment)) + geom_boxplot(outlier.shape = NA) + geom_jitter(size=2, width = 0.25) + 
  mutate(pco2=Treatment) %>%
  mutate_at("pco2", function(x) {
    case_when(
      x == "Ambient" ~ 371,
      x == "Moderate" ~ 704,
      x == "Low" ~ 1424)}) %>%
  # mutate(pH=Treatment) %>%
  # mutate_at("pH", function(x) {
  #   case_when(
  #     x == "Ambient" ~ 8.0,
  #     x == "Moderate" ~ 7.8,
  #     x == "Low" ~ 7.6)}) %>%
  ggplot(aes(x=pco2, y=mean, group=gene, color=moduleColor)) + geom_line(size=.01) +geom_point() + 
  scale_x_continuous(name ="pCO2 treatment", 
                    limits=c(350, 1450),
                   breaks=c(370, 700, 1425)) +
                   #  limits=c(7.56, 8.04),
                   # breaks=c(7.6, 7.8, 8.0)) +
  # geom_errorbar(aes(ymin=median-sd, ymax=median+sd), width=.2,
  #                position=position_dodge(0.05)) +
  theme_minimal() +
  theme(legend.position = "right") +
  facet_wrap(~process) + 
  scale_color_manual(values = c("magenta" = "magenta2",
                                "pink"="pink3",
                                "coral1"="coral1",
                                "ivory"="green",
                                "green"="green",
                                "lightgreen"="green",
                                "plum1"="blue"), 
                     name = "Enriched function", 
                     labels = c("magenta"="response to virus", 
                                "pink"="positive regulation of microglial cell activation", 
                                "coral1"="activation of innate immune response",
                                "plum1"="chaperone mediated protein folding",
                                #"green"="Signal transduction",
                                "lightgreen"="Octopamine & tyromine receptors")) + # "ivory"=NA, "lightgreen"=NA
  ggtitle("Expression of immune-response genes") + ylab("median count\n(variance-stabilized transformed)") + xlab("pCO2 concentration")

  
# Individual genes
for (i in 1:length(q$id)) {
a <-  plotCounts(dds.DESeq.pH, gene=q$id[i], intgroup=c("Treatment"), returnData = TRUE) %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
print(ggplot(a %>% rownames_to_column("sample"),
       aes(x=Treatment, y=count, color=Treatment, label=sample)) +
  #geom_boxplot(outlier.shape = NA) +
  geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=2) +
    theme_bw() +
    ggtitle(paste(gsub("\\(.*","",q$protein_names[i]), "\n", q$SPID[i], " (", q$id[i], ")")) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
}



c("orcokinin", )
c("g74523")
```


## Individual genes 

```{r}
a <-  plotCounts(dds.DESeq.pH, gene="g21591", intgroup=c("Treatment"), returnData = TRUE) %>% 
  mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
  
ggplotly(ggplot(a %>% rownames_to_column("sample"),
       aes(x=Treatment, y=count, color=Treatment, label=sample)) +
  #geom_boxplot(outlier.shape = NA) +
  geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
    #geom_text(size=2) +
    theme_bw() +
    #ggtitle(paste(gsub("\\(.*","",q$protein_names[i]), "\n", q$SPID[i], " (", q$id[i], ")")) +
    theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
```

# POTENTIAL PLOTS FOR PAPER 

## Use DAVID enrichment results to make functional line plots 

```{r}
# # Check out enriched biological processes as per DAVID 
# read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(PValue<0.05) %>% filter(Category=="GOTERM_BP_DIRECT") %>% View()
# 
# read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(PValue<0.05) %>% filter(Category=="GOTERM_BP_DIRECT") %>% View()
# 
# read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(PValue<0.05) %>% filter(Category=="GOTERM_BP_DIRECT") %>% View()
# 
# read_delim(file = paste("DAVID_AM-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(PValue<0.05) %>% filter(Category=="GOTERM_BP_DIRECT") %>% View()
# 
# read_delim(file = paste("DAVID_ML-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(PValue<0.05) %>% filter(Category=="GOTERM_BP_DIRECT") %>% View()
# 
# read_delim(file = paste("DAVID_ML-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(PValue<0.05) %>% filter(Category=="GOTERM_BP_DIRECT") %>% View()
# 
# # pco2-correlated modules
# 
# # negative correlated with pco2 
# 
# #Protein biosynthesis
# read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Tricarboxylic acid cycle
# read_delim(file = paste("DAVID_lightcyan_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Steroid metabolism
# read_delim(file = paste("DAVID_firebrick4_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Cell adhesion
# read_delim(file = paste("DAVID_darkviolet_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Antiviral defense
# read_delim(file = paste("DAVID_coral1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Stress response
# read_delim(file = paste("DAVID_plum1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #DNA integration
# read_delim(file = paste("DAVID_pink_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# # positive correlated with pco2
# read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Transcription regulation
# read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #DNA recombination
# read_delim(file = paste("DAVID_purple_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #Transcription regulation
# read_delim(file = paste("DAVID_lightgreen_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# #cGMP biosynthesis
# read_delim(file = paste("DAVID_ivory_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
```


```{r}
# # Set the max number of genes to show in each line plot (too many = too busy)
# ngenes <- 10
# 
# translation <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("~translation", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Translation", change="Downregulated")
# translation <- translation %>% filter(gene_id %in% 
#     (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
#   filter(gene_id %in% translation$gene_id) %>% arrange(pvalue) %>% 
#   head(n=ngenes))$gene_id)
#                                         
# 
# TCA <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("tricarboxylic acid cycle", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="TCA Cycle", change="Downregulated")
# TCA <- TCA %>% filter(gene_id %in% 
#     (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
#   filter(gene_id %in% TCA$gene_id) %>% arrange(pvalue) %>% 
#   head(n=ngenes))$gene_id)
# 
# 
# aerobic <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("aerobic respiration", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Aerobic Respiration", change="Downregulated")
# 
# carb.metabolism <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("carbohydrate metabolic process", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Carbohydrate Metabolism", change="Downregulated")
# carb.metabolism <- carb.metabolism %>% filter(gene_id %in% 
#     (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
#   filter(gene_id %in% carb.metabolism$gene_id) %>% arrange(pvalue) %>% 
#   head(n=ngenes))$gene_id)
# 
# 
# 
# immune <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("T cell mediated cytotoxicity|positive regulation of microglial cell activation", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Immune System", change="Downregulated")
# 
# 
# growth.reg <- left_join(read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("cell growth|cell proliferation", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Growth suppression", change="Upregulated")
#          
# 
# folding <- left_join(read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("protein folding", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Protein refolding\n(stress-response)", change="Upregulated")
# 
# transposition <- left_join(read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("transposition|DNA integration", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Transposition", change="Upregulated")
# transposition <- transposition %>% filter(gene_id %in% 
#     (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
#        filter(log2FoldChange < 0) %>%
#   filter(gene_id %in% transposition$gene_id) %>% arrange(pvalue) %>% 
#   head(n=ngenes))$gene_id)
# 
# ## Enrichment of axonogenesis is not signficant, pval~0.09. 
# # axonogenesis <- left_join(read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
# #       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
# #       filter(grepl("axonogenesis", Term)) %>% select(Genes) %>%
# #       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
# #     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
# #       filter(gene_id %in% rownames(diffex.AL)),
# #   by=c("Genes"="spid")) %>%
# #   mutate(funct="Axonogenesis", change="Upregulated")
# 
# regulation <- left_join(read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% 
#       filter(grepl("regulation of transcription", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Transcription\nregulation", change="Upregulated")
# regulation <- regulation %>% filter(gene_id %in% 
#     (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
#        filter(log2FoldChange < 0) %>%
#   filter(gene_id %in% regulation$gene_id) %>% arrange(pvalue) %>% 
#   head(n=ngenes))$gene_id)


```


```{r}
# genes.stats %>% select(-change) %>% filter(gene %in% 
#                          c(translation$gene_id,
#                            TCA$gene_id,
#                            aerobic$gene_id,
#                            carb.metabolism$gene_id,
#                            immune$gene_id,
#                            growth.reg$gene_id,
#                            folding$gene_id,
#                            transposition$gene_id,
#                            #axonogenesis$gene_id,
#                            regulation$gene_id)) %>% 
#   left_join(rbind(
#     translation,
#     TCA, 
#     aerobic, 
#     immune, 
#     growth.reg, 
#     folding,
#     transposition, 
#     #axonogenesis, 
#     regulation
#     ) %>% select(gene_id, funct, change), 
#             by=c("gene"="gene_id")) %>%
#   mutate(funct = as.factor(funct), change = as.factor(change)) %>%
#   #ggplot(aes(x=Treatment, y=median, colour=Treatment)) + geom_boxplot(outlier.shape = NA) + geom_jitter(size=2, width = 0.25) + 
#  mutate(pco2=Treatment) %>%
#  mutate_at("pco2", function(x) {
#   case_when(
#     x == "Ambient" ~ 371,
#     x == "Moderate" ~ 704,
#     x == "Low" ~ 1424)}) %>%
#   # mutate(pH=Treatment) %>%
#   # mutate_at("pH", function(x) {
#   #   case_when(
#   #     x == "Ambient" ~ 8.0,
#   #     x == "Moderate" ~ 7.8,
#   #     x == "Low" ~ 7.6)}) %>%
#   
#   filter(change=="Downregulated") %>%
#   #filter(change=="Upregulated") %>%
# 
#   ggplot(aes(x=pco2, y=mean, group=gene)) + geom_line(size=.01, color="forestgreen") +geom_point(color="forestgreen") +  #firebrick3
#   scale_x_continuous(name ="pCO2 treatment", 
#                     limits=c(350, 1450),
#                    breaks=c(370, 700, 1425)) +
#                    #  limits=c(7.56, 8.04),
#                    # breaks=c(7.6, 7.8, 8.0)) +
#   # geom_errorbar(aes(ymin=median-sd, ymax=median+sd), width=.2,
#   #                position=position_dodge(0.05)) +
#   theme_minimal() +
#   theme(legend.position = "right") +
#   facet_wrap(~funct, scales = "free", ncol = 2) + 
#   # scale_color_manual(values = c("magenta" = "magenta2",
#   #                               "pink"="pink3",
#   #                               "coral1"="coral1",
#   #                               "ivory"="green",
#   #                               "green"="green",
#   #                               "lightgreen"="green",
#   #                               "plum1"="blue"), 
#   #                    name = "Enriched function", 
#   #                    labels = c("magenta"="response to virus", 
#   #                               "pink"="positive regulation of microglial cell activation", 
#   #                               "coral1"="activation of innate immune response",
#   #                               "plum1"="chaperone mediated protein folding",
#   #                               #"green"="Signal transduction",
#   #                               "lightgreen"="Octopamine & tyromine receptors")) + # "ivory"=NA, "lightgreen"=NA
#   ggtitle("Expression of downregulated functions") + 
#   ylab("median count\n(variance-stabilized transformed)") + xlab("pCO2 treatment")
```

# WGCNA enrichment results 

MODULES FROM WGCNA WITH HIGH CORRELATION COEFFICIENT (>=0.5)

```{r}
#(geneInfo %>% filter(moduleColor=="magenta"))$id
```


```{r}
# Set the max number of genes to show in each line plot (too many = too busy)
ngenes <- 10

### GENES MODULES DOWNREGULATED IN OA 

# magenta
read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

magenta <- left_join(read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
      filter(grepl("~translation", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="magenta"))$id) %>%
  mutate(funct="Translation", change="Downregulated")
(magenta <- magenta %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% magenta$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))


# lightcyan
# TCA cycle
read_delim(file = paste("DAVID_lightcyan_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

lightcyan <- left_join(read_delim(file = paste("DAVID_lightcyan_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
      filter(grepl("~tricarboxylic acid cycle", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="lightcyan"))$id) %>%
  mutate(funct="Energy production", change="Downregulated")
(lightcyan <- lightcyan %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% lightcyan$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))


# firebrick4
# glycoprotein catabolic process (carbohydrate metabolic processes)
read_delim(file = paste("DAVID_firebrick4_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

(firebrick4 <- left_join(read_delim(file = paste("DAVID_firebrick4_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
      filter(grepl("~glycoprotein catabolic process|vesicle transport along actin filament", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="firebrick4"))$id) %>%
  mutate(funct="Carbohydrate metabolism,\nvesicle transport", change="Downregulated"))


# darkviolet
# tissue development
read_delim(file = paste("DAVID_darkviolet_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

(darkviolet <- left_join(read_delim(file = paste("DAVID_darkviolet_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("~tissue development", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="darkviolet"))$id) %>% #filter for only genes in that module 
  mutate(funct="Tissue development", change="Downregulated"))


# royalblue
# sarcomere organization (striated muscle cell development)
read_delim(file = paste("DAVID_royalblue_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

royalblue <- left_join(read_delim(file = paste("DAVID_royalblue_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("~sarcomere organization", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="royalblue"))$id) %>% #filter for only genes in that module 
  mutate(funct="Muscle development\nand activity", change="Downregulated")
(royalblue <- royalblue %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% royalblue$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))


# coral1
# chromatin silencing by small RNA
read_delim(file = paste("DAVID_coral1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

(coral1 <- left_join(read_delim(file = paste("DAVID_coral1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("~chromatin silencing by small RNA", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="coral1"))$id) %>% #filter for only genes in that module 
  mutate(funct="Immune response", change="Downregulated"))


# plum1
# chaperone mediated protein folding requiring cofactor
read_delim(file = paste("DAVID_plum1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

(plum1 <- left_join(read_delim(file = paste("DAVID_plum1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("protein", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="plum1"))$id) %>% #filter for only genes in that module 
  mutate(funct="Protein folding\n(stress response)", change="Downregulated"))



# pink
read_delim(file = paste("DAVID_pink_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()
pink <- left_join(read_delim(file = paste("DAVID_pink_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("metabolic process|fatty acid|positive regulation of microglial cell activation|T cell mediated cytotoxicity", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="pink"))$id) %>% #filter for only genes in that module 
  mutate(funct="Lipid & carbohydrate metabolic processes\n,immune function", change="Downregulated")
(pink <- pink %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% pink$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))


### GENES MODULES DOWNREGULATED IN OA 


# blue2 -  no enriched biological processes
read_delim(file = paste("DAVID_blue2_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()


# lightsteelblue1
read_delim(file = paste("DAVID_lightsteelblue1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>% View()

(lightsteelblue1 <- left_join(
  read_delim(file = paste("DAVID_lightsteelblue1_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
      filter(grepl("transcription ", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process), #%>% 
      #filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="lightsteelblue1"))$id) %>% #filter for only genes in that module 
  mutate(funct="Transcription regulation", change="Upregulated"))


# ivory
read_delim(file = paste("DAVID_ivory_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

(ivory <- left_join(read_delim(file = paste("DAVID_ivory_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
      filter(grepl("insulin secretion|neuron|signaling|cGMP", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="ivory"))$id) %>% #filter for only genes in that module 
  mutate(funct="Signaling", change="Upregulated"))


# lightgreen
read_delim(file = paste("DAVID_lightgreen_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

lightgreen <- left_join(read_delim(file = paste("DAVID_lightgreen_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("transcription|DNA methlation|imprinting", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="lightgreen"))$id) %>% #filter for only genes in that module 
  mutate(funct="Transcription regulation)", change="Upregulated")
(lightgreen <- lightgreen %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% lightgreen$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))



# purple
read_delim(file = paste("DAVID_purple_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

purple <- left_join(read_delim(file = paste("DAVID_purple_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("transcription", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="purple"))$id) %>% #filter for only genes in that module 
  mutate(funct="Transcription regulation", change="Upregulated")
(purple <- purple %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% purple$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))




# green
read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>% View()

green <- left_join(read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.01) %>%
      filter(grepl("Transcription regulation|signal transduction", Term)) %>% select(Genes) %>%
      mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
    P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
      filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM), rownames(diffex.ML))),
  by=c("Genes"="spid")) %>%
  filter(gene_id %in% (geneInfo %>% filter(moduleColor=="green"))$id) %>% #filter for only genes in that module 
  mutate(funct="Transcription regulation\nsignal transduction", change="Upregulated")
(green <- green %>% filter(gene_id %in% 
    (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
  filter(gene_id %in% green$gene_id) %>% arrange(pvalue) %>% 
  head(n=ngenes))$gene_id))

```

## Create a master spreadsheet of all enriched biological functions in modules 

```{r}
enriched.bp.modules <- vector(mode = "list", length = length(modules))
names(enriched.bp.modules) <- modules


for (i in 1:length(modules)) {
  enriched.bp.modules[[i]] <- read_delim(file = paste("DAVID_", modules[i], "_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05)
}

enriched.bp.modules <- bind_rows(enriched.bp.modules, .id = "module") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%
  clean_names() %>% 
  select(module, go, process, count, percent, p_value, fold_enrichment, fdr) 

enriched.bp.modules %>% write_clip() #copy to clipboard to paste into google spreadsheet 

enriched.bp.modules %>% filter(fdr<0.1) %>% group_by(module) %>% tally()

```

Save WGCNA enrichment results (BP, MF, CC) for supplemental tables 

```{r}
enriched.modules <- vector(mode = "list", length = length(modules))
names(enriched.modules) <- modules

for (i in 1:length(modules)) {
  enriched.modules[[i]] <- read_delim(file = paste("DAVID_", modules[i], "_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
      filter(Category %in% c("GOTERM_BP_DIRECT", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT"))
}

# Create one master spreadsheet with all enrichment results, reformat, save to .csv
bind_rows(enriched.modules, .id = "Module") %>% 
  separate(Term, into = c("GO", "process"), sep = "~") %>%
    mutate(`OA Effect`=Module) %>%
  mutate(`OA Effect`=case_when(
    `OA Effect`%in% c("magenta", "lightcyan", "firebrick4", "darkviolet", 
                      "royalblue", "coral1", "plum1", "pink") ~ "Downregulated", 
    TRUE ~ "Upregulated")) %>%

  mutate(Module=case_when(
    Module=="magenta" ~ "a (magenta)",
    Module=="lightcyan" ~ "b (lightcyan)",
    Module=="firebrick4" ~ "c (firebrick4)",
    Module=="darkviolet" ~ "d (darkviolet)",
    Module=="royalblue" ~ "e (royalblue)",
    Module=="coral1" ~ "f (coral1)",
    Module=="plum1" ~ "g (plum1)",
    Module=="pink" ~ "h (pink)",
    Module=="blue2" ~ "i (blue2)",
    Module=="lightsteelblue1" ~ "j (lightsteelblue1)",
    Module=="ivory" ~ "k (ivory)",
    Module=="lightgreen" ~ "l (lightgreen)",
    Module=="purple" ~ "m (purple)",
    Module=="green" ~ "n (green)")) %>%

  mutate(Category=case_when(
    Category=="GOTERM_BP_DIRECT" ~ "Biological Process",
    Category=="GOTERM_MF_DIRECT" ~ "Molecular Function",
    Category=="GOTERM_CC_DIRECT" ~ "Cellular Component")) %>%
  dplyr::rename(Term=process, `Number of Genes` = Count, 
                `P-Value`=PValue, `Genes (Uniprot IDs)`=Genes) %>%
  mutate(`Fold Enrichment`=round(`Fold Enrichment`, digits = 1),
    `P-Value`=as.numeric(formatC(`P-Value`, format = "e", digits = 2)),
    FDR=as.numeric(formatC(FDR, format = "e", digits = 2))) %>%
  select(Module, `OA Effect`, Category, GO, Term, `Number of Genes`, `Genes (Uniprot IDs)`, `Fold Enrichment`, `P-Value`, FDR) %>%
  arrange(Category, Module, `OA Effect`, `P-Value`) %>%
  filter(`P-Value`<0.05) %>%
  write_csv(file = "../Supplemental-Table-3_WGCNA-Enrichment-Results.csv", col_names = TRUE)
```

USING UNIPROT KEYWORDS to plot representative functions 

```{r}
# ## UNIPROT KEYWORDS
# read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# read_delim(file = paste("DAVID_AM-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# read_delim(file = paste("DAVID_ML-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# read_delim(file = paste("DAVID_ML-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# # GREEN & MAGENTA MODULES
# read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
```

```{r}
# ngenes <- 10
# 
# # Upregulated processes in OA compared to ambient 
# 
# transposition <- left_join(
#   read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("DNA", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Transposition", change="Upregulated")
# transposition <- transposition %>% filter(gene_id %in% 
#     (diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>%
#   filter(gene_id %in% transposition$gene_id) %>% arrange(pvalue) %>% 
#   filter(log2FoldChange < 0) %>%
#   head(n=ngenes))$gene_id)
#                           
# transc.reg <- left_join(
#   rbind(
#   read_delim(file = paste("DAVID_AL-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Transcription regulation", Term)) %>% select(Genes), 
#   read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Transcription regulation", Term)) %>% select(Genes))  %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes) %>% distinct(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% rownames(diffex.AL)),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Transcription\nregulation", change="Upregulated")
# transc.reg <- transc.reg %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% transc.reg$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% transc.reg$gene_id)) %>% arrange(pvalue) %>% head(n=ngenes))$gene_id)
# 
# differentiation <- left_join(read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Differentiation", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Differentiation", change="Upregulated")
# differentiation <- differentiation %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% differentiation$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% differentiation$gene_id)) %>% 
#        arrange(pvalue) %>% head(n=ngenes))$gene_id)
# 
# neurogenesis <- left_join(read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Neurogenesis", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Neurogenesis", change="Upregulated")
# 
# stress <- left_join(read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Stress response", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Stress-response\n(Heat shock proteins)", change="Upregulated")
# 
# growth <- left_join(read_delim(file = paste("DAVID_AM-UP_FC0.5.txt", sep=""), delim = "\t") %>% 
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>%
#       filter(grepl("cell growth|cell proliferation", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Growth control", change="Upregulated")
# 
# 
# # DOWNREGULATED IN OA 
# 
# TCA <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Tricarboxylic acid cycle", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="TCA Cycle", change="Downregulated")
# TCA <- TCA %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% TCA$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% TCA$gene_id)) %>% 
#        arrange(pvalue) %>% head(n=ngenes))$gene_id)
# 
# carb.metabolism <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Carbohydrate metabolism", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Carbohydrate\nmetabolism", change="Downregulated")
# carb.metabolism <- carb.metabolism %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% carb.metabolism$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% carb.metabolism$gene_id)) %>% 
#        arrange(pvalue) %>% head(n=ngenes))$gene_id)
# 
# DNA.replication <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("DNA replication", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="DNA replication", change="Downregulated")
# DNA.replication <- DNA.replication %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% DNA.replication$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% DNA.replication$gene_id)) %>% 
#        arrange(pvalue) %>% head(n=ngenes))$gene_id)
# 
# ATP <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("ATP synthesis", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="ATP synthesis", change="Downregulated")
# 
# lipid <- left_join(read_delim(file = paste("DAVID_AL-DOWN_FC0.5.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Lipid metabolism", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Lipid\nmetabolism", change="Downregulated")
# lipid <- lipid %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% lipid$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% lipid$gene_id)) %>% 
#        arrange(pvalue) %>% head(n=ngenes))$gene_id)
# 
# protein.biosyn <- left_join(read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05) %>%
#       filter(grepl("Protein biosynthesis", Term)) %>% select(Genes) %>%
#       mutate(Genes = strsplit(as.character(Genes), ", ")) %>% unnest(Genes),
#     P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, gene_ontology_biological_process) %>% 
#       filter(gene_id %in% c(rownames(diffex.AL), rownames(diffex.AM))),
#   by=c("Genes"="spid")) %>%
#   mutate(funct="Protein\nbiosynthesis", change="Downregulated")
# protein.biosyn <- protein.biosyn %>% filter(gene_id %in% 
#     (rbind(diffex.AL %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% protein.biosyn$gene_id),
#            diffex.AM %>% as.data.frame() %>% rownames_to_column("gene_id") %>% filter(gene_id %in% protein.biosyn$gene_id)) %>% 
#        arrange(pvalue) %>% head(n=ngenes))$gene_id)
```

```{r}
# # MAKE LINE PLOTS
# genes.stats %>% select(-change) %>% filter(gene %in% 
#                          c(transposition$gene_id,
#                            transc.reg$gene_id,
#                            stress$gene_id,
#                            differentiation$gene_id,
#                            neurogenesis$gene_id,
#                            growth$gene_id,
#                            TCA$gene_id,
#                            carb.metabolism$gene_id,
#                            DNA.replication$gene_id,
#                            ATP$gene_id,
#                            lipid$gene_id,
#                            protein.biosyn$gene_id)) %>% 
#   left_join(rbind(
#     transposition,
#     transc.reg,
#     stress,
#     differentiation,
#     neurogenesis,
#     growth,
#     TCA,
#     carb.metabolism,
#     DNA.replication,
#     ATP,
#     lipid,
#     protein.biosyn) %>% select(gene_id, funct, change), 
#             by=c("gene"="gene_id")) %>%
#   mutate(funct = as.factor(funct), change = as.factor(change)) %>%
#   #ggplot(aes(x=Treatment, y=median, colour=Treatment)) + geom_boxplot(outlier.shape = NA) + geom_jitter(size=2, width = 0.25) + 
#  # mutate(pco2=Treatment) %>%
#  # mutate_at("pco2", function(x) {
#  #  case_when(
#  #    x == "Ambient" ~ 371,
#  #    x == "Moderate" ~ 704,
#  #    x == "Low" ~ 1424)}) %>%
#   mutate(pH=Treatment) %>%
#   mutate_at("pH", function(x) {
#     case_when(
#       x == "Ambient" ~ 8.0,
#       x == "Moderate" ~ 7.8,
#       x == "Low" ~ 7.6)}) %>%
#   
#   filter(change=="Downregulated") %>%
#   #filter(change=="Upregulated") %>%
# 
#   ggplot(aes(x=pH, y=mean, group=gene)) + 
#   # geom_line(size=.01, color="forestgreen") +
#   # geom_point(color="forestgreen") +  #upregulated
#   geom_line(size=.01, color="firebrick3") +
#   geom_point(color="firebrick3") +  #downregulated
#   scale_x_continuous(name ="pH treatment", 
#                    #  limits=c(350, 1450),
#                    # breaks=c(370, 700, 1425)) +
#                     limits=c(7.56, 8.04),
#                    breaks=c(7.6, 7.8, 8.0)) +
#   # geom_errorbar(aes(ymin=median-sd, ymax=median+sd), width=.2,
#   #                position=position_dodge(0.05)) +
#   theme_minimal() +
#   theme(legend.position = "right") +
#   facet_wrap(~funct, scales = "free", ncol = 2) + 
#   # scale_color_manual(values = c("magenta" = "magenta2",
#   #                               "pink"="pink3",
#   #                               "coral1"="coral1",
#   #                               "ivory"="green",
#   #                               "green"="green",
#   #                               "lightgreen"="green",
#   #                               "plum1"="blue"), 
#   #                    name = "Enriched function", 
#   #                    labels = c("magenta"="response to virus", 
#   #                               "pink"="positive regulation of microglial cell activation", 
#   #                               "coral1"="activation of innate immune response",
#   #                               "plum1"="chaperone mediated protein folding",
#   #                               #"green"="Signal transduction",
#   #                               "lightgreen"="Octopamine & tyromine receptors")) + # "ivory"=NA, "lightgreen"=NA
#   #ggtitle("Upregulated Processes") + 
#   ggtitle("Downregulated Processes") + 
#   ylab("median count\n(variance-stabilized transformed)") #+ xlab(bquote(~pCO[2]~'Treatment'))
# ```
# 
# 
# MODULES FROM WGCNA WITH HIGH CORRELATION COEFFICIENT (>=0.5)
# 
# ```{r}
# # Check out enriched biological processes as per DAVID - MAGENTA, filtered for genes that are either DEGs or are correlated with pCO2 on an individual basis
# read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05) %>% View()
# 
# # Check out enriched biological processes as per DAVID - GREEN, filtered for genes that are either DEGs or are correlated with pCO2 on an individual basis
# read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% 
#   filter(PValue<0.05) %>% View()
# 
# # Check out enriched UNIPROT KEYWORDS as per DAVID - MAGENTA, filtered for genes that are either DEGs or are correlated with pCO2 on an individual basis
# read_delim(file = paste("DAVID_magenta_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
# 
# # Check out enriched biological processes as per DAVID - GREEN, filtered for genes that are either DEGs or are correlated with pCO2 on an individual basis
# read_delim(file = paste("DAVID_green_DEGs-or-GS.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05) %>% View()
```
CRITICAL GENES:  LOW-CV Upregulated Gene DAVID results! 

```{r}
# # CRITICAL IN SEVERE OA 
# read_delim(file = paste("DAVID_lowCV-upreg-low.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05, Count>2) %>% View()
# 
# read_delim(file = paste("DAVID_lowCV-upreg-low.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_CC_DIRECT") %>% filter(PValue<0.05, Count>2) %>% View()
# 
# # CRITICAL IN MODERATE OA
# read_delim(file = paste("DAVID_lowCV-upreg-moderate.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_BP_DIRECT") %>% filter(PValue<0.05, Count>2) %>% View()
# 
# # CRITICAL IN MODERATE OA
# read_delim(file = paste("DAVID_lowCV-upreg-moderate.txt", sep=""), delim = "\t") %>%
#       filter(Category=="GOTERM_CC_DIRECT") %>% filter(PValue<0.05, Count>2) %>% View()
# 
# # CRITICAL IN AMBIENT 
# read_delim(file = paste("DAVID_lowCV-upreg-ambient.txt", sep=""), delim = "\t") %>%
#   filter(Category=="GOTERM_BP_DIRECT") %>%  filter(PValue<0.05, Count>2) %>% View()
# 
# read_delim(file = paste("DAVID_lowCV-upreg-ambient.txt", sep=""), delim = "\t") %>%
#   filter(Category=="GOTERM_CC_DIRECT") %>%  filter(PValue<0.05, Count>2) %>% View()
# 
# 
# # CRITICAL IN SEVERE OA 
# read_delim(file = paste("DAVID_lowCV-upreg-low.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05, Count>2) 
# 
# # CRITICAL IN MODERATE OA
# read_delim(file = paste("DAVID_lowCV-upreg-moderate.txt", sep=""), delim = "\t") %>%
#       filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% filter(PValue<0.05, Count>2)
# 
# # CRITICAL IN AMBIENT 
# read_delim(file = paste("DAVID_lowCV-upreg-ambient.txt", sep=""), delim = "\t") %>%
#   filter(Category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
#   filter(PValue<0.05, Count>2)

```

```{r}
# level <- "BP" #select BP, MF, or CC
# MOD <- "pink"
# # GO <- "negative regulation of cell proliferation"
# # GO <- "immune|cytotoxicity|pattern recognition"
# # GO <- "viral"
# GO <- "immune"
# 
# 
# # GO MWU 
# q <- read_delim(file = paste(level, "_WGCNA-module_", MOD, ".csv", sep=""), delim = "\t") %>% #filter(name==GO) %>% 
# filter(., grepl(GO,name)) %>% 
#   left_join(geneInfo.mwu, by = c("seq"="id")) %>% filter(moduleColor == MOD)%>% 
#   select(seq, SPID, protein_names) %>% unique()
# 
#   # # If interested- further filter to only retain genes that were differentially expressed 
#   # filter(seq %in% unique(c(rownames(diffex.AM),rownames(diffex.AL), rownames(diffex.ML)))) %>%
# #  View()
# 
# # plot genes contributing to enriched GO term of interest 
# for (i in 1:length(q$seq)) {
# a <-  plotCounts(dds.DESeq.pH, gene=q$seq[i], intgroup=c("Treatment"), returnData = TRUE) %>% 
#   mutate(Treatment=factor(Treatment, levels=c("Ambient", "Moderate", "Low")))
#   
# print(ggplot(a %>% rownames_to_column("sample"),
#        aes(x=Treatment, y=count, color=Treatment, label=sample)) +
#   #geom_boxplot(outlier.shape = NA) +
#   geom_point(size=2, position=position_jitter(w = 0.15,h = 0)) +
#     #geom_text(size=2) +
#     theme_bw() +
#     ggtitle(paste(gsub("\\(.*","",q$protein_names[i]), "\n", q$SPID[i], " (", q$seq[i], ")")) +
#     theme(axis.title.x = element_blank(), plot.title = element_text(hjust = 0.5, size=10), legend.position = "none"))
# }
# 
# genes.stats %>% filter(gene %in% q$seq) %>% 
# ggplot(aes(x=Treatment, y=median, colour=Treatment)) + geom_boxplot(outlier.shape = NA) + geom_jitter(size=2, width = 0.25) + theme_minimal() +
#   ggtitle("Immune-response gene counts")
# 
# 
# # 
# genes.stats %>% filter(., grepl(GO, gene_ontology_biological_process)) %>% filter(gene %in% degs) %>% filter(change=="Down-regulated") %>%
#   mutate(pco2=Treatment) %>%
#   mutate_at("pco2", function(x) {
#     case_when(
#       x == "Ambient" ~ 371,
#       x == "Moderate" ~ 704,
#       x == "Low" ~ 1424)}) %>%
#   ggplot(aes(x=pco2, y=mean, group=gene)) + geom_point() + geom_line(size=.01) +
#   #ggplot(aes(x=Treatment, y=median, colour=Treatment)) + geom_violin(outlier.shape = NA) + geom_jitter(size=2, width = 0.25) + 
#   theme_minimal() +
#   ggtitle("Immune-response gene counts - degs")
#   
#   
#   
# 
# 
# 
# 
# 
# 
# geneInfo.mwu %>% filter(grepl("Cuticle", protein_names)) %>% #View()
#   filter(id %in% unique(c(rownames(diffex.AM),rownames(diffex.AL), rownames(diffex.ML)))) %>% View()
# 
# 
# read_delim(file = paste(level, "_WGCNA-module_", MOD, ".csv", sep=""), delim = "\t") %>% 
#   left_join(geneInfo.mwu, by = c("seq"="id")) %>% View()
# 
# geneInfo.mwu %>% filter(grepl("Mannosyltransferase", protein_names)) %>% #View()
#   filter(id %in% rownames(diffex.AL)) %>% dplyr::select(id, SPID, protein_names, moduleColor, GS.pco2, p.GS.pco2) %>%
#   left_join(diffex.AL %>% as.data.frame() %>% rownames_to_column("id"))
#            #unique(c(rownames(diffex.AM),rownames(diffex.AL), rownames(diffex.ML))))
# 
# 
# geneInfo.mwu %>% filter(moduleColor =="lightcyan") %>%
#   left_join(P.camt.blast.GO[,c("gene_id","gene_ontology_i_ds")], by = c("id"="gene_id")) %>%
#   filter(grepl("Cuticle", protein_names)) %>% select(id) %>% unlist() %>% as.vector()
# 
# Term("GO:0032197")
# 
# geneInfo.mwu.mods[["magenta"]] %>% filter(sign!=0) %>% left_join(P.camt.blast.GO[c("gene_id", "spid", "evalue", "protein_names", "gene_ontology_biological_process", "gene_ontology_i_ds")], by=c("id"="gene_id")) %>% 
#   filter(grepl("transposon|LINE-1|retrotransposable|transposable element|mobile element jockey", protein_names)) %>% View() 
```

# Calculate % of genes that are transposable elements in each module, and in the background, ALSO include % in DEGs, and low CV 

```{r}
getwd()
load(file="../wgcna/geneInfo")
TE.genes <-  ("transposon|transposable|LINE|retrotransposable element|transposable element|mobile element jockey|Pol polyprotein")

te.perc <- data.frame(module= character(0), te.perc= numeric(0), te.n= integer(0), n= integer(0), type=integer(0))

# add row with %TEs in all genes 
te.perc[1,1] <- "background"
te.perc[1,2] <- geneInfo %>% filter(str_detect(protein_names, TE.genes)) %>% nrow() / geneInfo %>% filter(!is.na(SPID)) %>% nrow()
te.perc[1,3] <- geneInfo %>% filter(str_detect(protein_names, TE.genes)) %>% nrow()
te.perc[1,4] <- geneInfo %>% filter(!is.na(SPID)) %>% nrow() #number of genes total 
te.perc[1,5] <- "Co-expression module"

for (i in 1:length(modules)) {
  te.perc[i+1,1] <- modules[[i]]
  te.perc[i+1,2] <- geneInfo %>% filter(moduleColor == modules[[i]], str_detect(protein_names, TE.genes)) %>% nrow() / 
                        geneInfo %>% filter(moduleColor == modules[[i]], !is.na(SPID)) %>% nrow()
  te.perc[i+1,3] <- geneInfo %>% filter(moduleColor == modules[[i]], str_detect(protein_names, TE.genes)) %>% nrow()
  te.perc[i+1,4] <- geneInfo %>% filter(moduleColor == modules[[i]], !is.na(SPID)) %>% nrow()
  te.perc[i+1,5] <- "Co-expression module"
}
modules
# Assign factor to each module indicating if it contains upreg/downreg/up-then-down genes 
te.perc <- te.perc %>% mutate(change = 
                     as.factor(case_when(
                       module=="background" ~ "Background",
                       module %in% c("magenta","lightcyan","firebrick4","darkviolet","royalblue","coral1","plum1","pink") ~ "Downregulated", 
                       module %in% c("blue2", "lightsteelblue1", "ivory", "lightgreen", "purple","green") ~ "Upregulated")))

# Add differential expression results 
cont <- c("z_AM", "z_AL", "z_ML")
diffex <- list(diffex.AM.annot, diffex.AL.annot, diffex.ML.annot)
colors <- c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c")

for (i in c(1,2,3)) {
# Upregulated
te.perc[i+15,1] <- cont[i]
te.perc[i+15,2] <- diffex[[i]] %>% filter(log2FoldChange<0, str_detect(protein_names, TE.genes)) %>% nrow() / 
                                    diffex[[i]] %>% filter(log2FoldChange<0, !is.na(spid)) %>% nrow()
te.perc[i+15,3] <- diffex[[i]] %>% filter(log2FoldChange<0, str_detect(protein_names, TE.genes)) %>% nrow()
te.perc[i+15,4] <- diffex[[i]] %>% filter(log2FoldChange<0, !is.na(spid)) %>% nrow()
te.perc[i+15,5] <- "DEGs"
te.perc[i+15,6] <- "Upregulated"
}

for (i in c(1,2,3)) {
# Downregulated
te.perc[i+18,1] <- cont[i]
te.perc[i+18,2] <- diffex[[i]] %>% filter(log2FoldChange>0, str_detect(protein_names, TE.genes)) %>% nrow() / 
                                    diffex[[i]] %>% filter(log2FoldChange>0, !is.na(spid)) %>% nrow()
te.perc[i+18,3] <- diffex[[i]] %>% filter(log2FoldChange>0, str_detect(protein_names, TE.genes)) %>% nrow()
te.perc[i+18,4] <- diffex[[i]] %>% filter(log2FoldChange>0, !is.na(spid)) %>% nrow()
te.perc[i+18,5] <- "DEGs"
te.perc[i+18,6] <- "Downregulated"
}

te.perc %>% 
  filter(module %in% modules, change == "Upregulated") %>% droplevels() %>%
  summarize(mean=mean(te.perc))

te.perc %>% 
  filter(module %in% modules, change == "Downregulated") %>% droplevels() %>%
  summarize(mean=mean(te.perc))


# Boxplot of the percent of each gene set that are TEs 
#pdf(file = "../../manuscript/Figures/Figure7.pdf", width = 3.5, height = 6.25)
tiff(file = "../../manuscript/PLOS-Climate-submission/Revision/Final/Figures-correct-dims/Fig8.tiff", width = 2.35, height = 4.2, units = "in", res = 600)
te.perc %>%
#  filter(module %in% modules) %>%
  filter(module != "background") %>%
  filter(module != "z_ML") %>%
  mutate(module=str_replace(module, "ivory", "khaki")) %>%
  mutate(gene_set = case_when(
    (module == "z_AM" & change == "Upregulated") ~ "Moderate",
    (module == "z_AL" & change == "Upregulated") ~ "Severe",
    (module == "z_ML" & change == "Upregulated") ~ "Severe",
    (module == "z_AM" & change == "Downregulated") ~ "Ambient",
    (module == "z_AL" & change == "Downregulated") ~ "Ambient",
    (module == "z_ML" & change == "Downregulated") ~ "Moderate",
    TRUE ~ "modules")) %>%
  mutate(gene_set=factor(gene_set, levels=c("modules", "Severe", "Moderate", "Ambient"))) %>%
  ggplot(aes(x=change, y=100*te.perc)) + 
  geom_boxplot(outlier.size = NA) + 
  #geom_jitter(position=position_jitter(seed=10), aes(fill=type, size=n, shape=gene_set)) + 
  geom_jitter(position=position_jitter(seed=4, width = 0.18), aes(fill=type, shape=type), size=3) +  #simplified figure
  # geom_text(position=position_jitter(width = .2,seed = 10),
  #           aes(color=module, label=paste(100*round(te.perc, digits = 2), "%", sep=""))) +
  theme_minimal() + 
  scale_fill_manual(name = "Gene set", values = c("gray65", "gray30") , guide="none") +     
  # scale_shape_manual(name = "Gene set", values = c(21, 22, 23, 24), 
  #                    labels=c("Co-expression module", "Upregulated in Severe OA", "Upregulated in Moderate OA", "Downregulated in OA")) +
  scale_shape_manual(name = NULL, values = c(21, 22), 
                     labels=c("Co-expressed gene modules", "Differentially expressed genes")) +
  # guides(shape = guide_legend(order=1, override.aes = list(size=4.25, fill=c("gray65", "gray30", "gray30", "gray30"))),
  #        size = guide_legend(order=0, override.aes = list(fill="white", shape=21))) +
  guides(shape = guide_legend(nrow=2,byrow=TRUE, order=1, override.aes = list(size=3.75, fill=c("gray65", "gray30"))),
         size = guide_legend(order=0, override.aes = list(fill="white", shape=21))) +
#  scale_size(range = c(3,9), name = "Number of genes in set") +
  theme(axis.title=element_text(size=8.5, colour = "gray10"), axis.text=element_text(size=8.5, colour = "gray10"), 
        legend.position="bottom", legend.justification = c(0,1), 
        legend.text=element_text(size=7.5, colour = "gray35"), legend.spacing.y = unit(0.05, 'lines'),
        legend.margin = margin(t = 5, r = 0, b = 0, l = -15, unit = "pt")) +
  ylab("Percent Transposable Elements") + xlab(NULL) + ggtitle(NULL) + #ggtitle("Percent of genes that are TEs\nby gene set") +
  geom_hline(yintercept = 100*te.perc[te.perc$module=="background",]$te.perc, linetype="dashed", color="gray20") 
dev.off()

```
Statistically test TE composition - chi square test maybe? 

```{r}
te.perc %>% filter(change !="Background") %>%
  ggplot() + geom_boxplot(aes(x=change, y=te.perc))
te.perc %>% filter(change !="Background") %>%
  ggplot() + geom_boxplot(aes(x=change, y=te.n))

(te.perc %>% filter(change !="Background"))$te.perc %>% hist()
summary(aov(te.perc ~ change, data=te.perc %>% filter(change !="Background")))

# Use beta-regression, more appropriate for 
install.packages("betareg")
install.packages("rcompanion")
install.packages("emmeans")
require(betareg)
require(car)
library(rcompanion)
require(emmeans)

model.beta <- betareg(te.perc ~ change, data=te.perc %>% filter(change !="Background"))
summary(model.beta)
Anova(model.beta)
joint_tests(model.beta)

test<-te.perc %>% filter(change !="Background")
lrtest(betareg(te.perc ~ change, data=test))
```
Chi-square test 

```{r}

 %>%
  pivot_wider(names_from=change, values_from=te.n)
```


## Save list of unique TEs identified in data for supplemental tables 

```{r}
geneInfo %>% dplyr::select(SPID, protein_names) %>% unique() %>% filter(str_detect(protein_names, TE.genes)) %>% rownames_to_column("geneID") %>%
  left_join(P.camt.blast %>% select(geneID, SeqID, Start, End, species, evalue, gene.Uni)) %>%
  select(geneID, SeqID, Start, End, evalue, SPID, species, gene.Uni, protein_names) %>%
  mutate(evalue=as.numeric(formatC(evalue, format = "e", digits = 2))) %>%
  write_csv(file = "../Supplemental-Table-6_Transposons.csv", col_names = TRUE)
```

## Visualize GO Enrichment Results using Bubble Plots 

```{bash}
## Remove files which list GO_MWU results filenames if they already exist  
rm BP_filenames.txt
rm MF_filenames.txt
rm CC_filenames.txt

for file in MWU_BP*
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 4  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> BP_filenames.txt
done

# Molecular functions results 
for file in MWU_MF*
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 4  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> MF_filenames.txt
done

# Cellular components results 
for file in MWU_CC*
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 4  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> CC_filenames.txt
done
```

```{r}
modules
```

## Read in enriched Biological Processes 

```{r}
filenames <- read_delim(file="BP_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = " "))}

(gene_sets.BP <- bind_rows(file_list, .id = "gene_set") %>% filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(category=as.factor("Biological Processes")) %>%  #add column designating go category 
    mutate(gene_set=factor(gene_set, levels=c("amb-mod", "amb-low", "mod-low", modules, "lightsteelblue", "mediumpurple2", "coral2"))) %>% #Convert gene_set column to factor, define all gene_sets 
    tidyr::complete(gene_set, category) %>% # add rows for gene_sets that are missing from dataframe

    # Create new column that identifies which treatment is upregulated in each DEG contrast (doesn't apply for WGCNA modules)
    mutate(gene_set_up = case_when((gene_set == "amb-mod" & delta.rank < 0) ~ "Ambient",
                                      (gene_set == "amb-mod" & delta.rank > 0) ~ "Moderate",
                                        (gene_set == "amb-low" & delta.rank < 0) ~ "Ambient",
                                        (gene_set == "amb-low" & delta.rank > 0) ~ "Low",
                                        (gene_set == "mod-low" & delta.rank < 0) ~ "Moderate",
                                        (gene_set == "mod-low" & delta.rank > 0) ~ "Moderate",
                                        gene_set == modules[[1]] ~ modules[[1]],
                                        gene_set == modules[[2]] ~ modules[[2]],
                                        gene_set == modules[[3]] ~ modules[[3]],
                                        gene_set == modules[[4]] ~ modules[[4]],
                                        gene_set == modules[[5]] ~ modules[[5]],
                                        gene_set == modules[[6]] ~ modules[[6]],
                                        gene_set == modules[[7]] ~ modules[[7]],
                                        gene_set == modules[[8]] ~ modules[[8]],
                                        gene_set == modules[[9]] ~ modules[[9]],
                                        gene_set == modules[[10]] ~ modules[[10]],
                                        gene_set == modules[[11]] ~ modules[[11]],
                                        gene_set == modules[[12]] ~ modules[[12]],
                                        gene_set == modules[[13]] ~ modules[[13]],
                                        gene_set == modules[[14]] ~ modules[[14]],
                                        gene_set == modules.interest[[1]] ~ modules.interest[[1]],
                                        gene_set == modules.interest[[2]] ~ modules.interest[[2]],
                                        gene_set == modules.interest[[3]] ~ modules.interest[[3]])))  
```

## Read in enriched Molecular Functions

```{r}
filenames <- read_delim(file="MF_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = " "))}

(gene_sets.MF <- bind_rows(file_list, .id = "gene_set") %>% filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(category=as.factor("Molecular Functions")) %>%  #add column designating go category 
    mutate(gene_set=factor(gene_set, levels=c("amb-mod", "amb-low", "mod-low", modules, "lightsteelblue", "mediumpurple2", "coral2"))) %>% #Convert gene_set column to factor, define all gene_sets 
    tidyr::complete(gene_set, category) %>% # add rows for gene_sets that are missing from dataframe

    # Create new column that identifies which treatment is upregulated in each DEG contrast (doesn't apply for WGCNA modules)
    mutate(gene_set_up = case_when((gene_set == "amb-mod" & delta.rank < 0) ~ "Ambient",
                                      (gene_set == "amb-mod" & delta.rank > 0) ~ "Moderate",
                                        (gene_set == "amb-low" & delta.rank < 0) ~ "Ambient",
                                        (gene_set == "amb-low" & delta.rank > 0) ~ "Low",
                                        (gene_set == "mod-low" & delta.rank < 0) ~ "Moderate",
                                        (gene_set == "mod-low" & delta.rank > 0) ~ "Moderate",
                                        gene_set == modules[[1]] ~ modules[[1]],
                                        gene_set == modules[[2]] ~ modules[[2]],
                                        gene_set == modules[[3]] ~ modules[[3]],
                                        gene_set == modules[[4]] ~ modules[[4]],
                                        gene_set == modules[[5]] ~ modules[[5]],
                                        gene_set == modules[[6]] ~ modules[[6]],
                                        gene_set == modules[[7]] ~ modules[[7]],
                                        gene_set == modules[[8]] ~ modules[[8]],
                                        gene_set == modules[[9]] ~ modules[[9]],
                                        gene_set == modules[[10]] ~ modules[[10]],
                                        gene_set == modules[[11]] ~ modules[[11]],
                                        gene_set == modules[[12]] ~ modules[[12]],
                                        gene_set == modules[[13]] ~ modules[[13]],
                                        gene_set == modules[[14]] ~ modules[[14]],
                                        gene_set == modules.interest[[1]] ~ modules.interest[[1]],
                                        gene_set == modules.interest[[2]] ~ modules.interest[[2]],
                                        gene_set == modules.interest[[3]] ~ modules.interest[[3]])))  
```

## Read in enriched Cellular Componenets  

```{r}
filenames <- read_delim(file="CC_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = " "))}

(gene_sets.CC <- bind_rows(file_list, .id = "gene_set") %>% filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(category=as.factor("Cellular Components")) %>%  #add column designating go category 
    mutate(gene_set=factor(gene_set, levels=c("amb-mod", "amb-low", "mod-low", modules, modules.interest))) %>% #Convert gene_set column to factor, define all gene_sets 
    tidyr::complete(gene_set, category) %>% # add rows for gene_sets that are missing from dataframe

    # Create new column that identifies which treatment is upregulated in each DEG contrast (doesn't apply for WGCNA modules)
    mutate(gene_set_up = case_when((gene_set == "amb-mod" & delta.rank < 0) ~ "Ambient",
                                      (gene_set == "amb-mod" & delta.rank > 0) ~ "Moderate",
                                        (gene_set == "amb-low" & delta.rank < 0) ~ "Ambient",
                                        (gene_set == "amb-low" & delta.rank > 0) ~ "Low",
                                        (gene_set == "mod-low" & delta.rank < 0) ~ "Moderate",
                                        (gene_set == "mod-low" & delta.rank > 0) ~ "Moderate",
                                        gene_set == modules[[1]] ~ modules[[1]],
                                        gene_set == modules[[2]] ~ modules[[2]],
                                        gene_set == modules[[3]] ~ modules[[3]],
                                        gene_set == modules[[4]] ~ modules[[4]],
                                        gene_set == modules[[5]] ~ modules[[5]],
                                        gene_set == modules[[6]] ~ modules[[6]],
                                        gene_set == modules[[7]] ~ modules[[7]],
                                        gene_set == modules[[8]] ~ modules[[8]],
                                        gene_set == modules[[9]] ~ modules[[9]],
                                        gene_set == modules[[10]] ~ modules[[10]],
                                        gene_set == modules[[11]] ~ modules[[11]],
                                        gene_set == modules[[12]] ~ modules[[12]],
                                        gene_set == modules[[13]] ~ modules[[13]],
                                        gene_set == modules[[14]] ~ modules[[14]],
                                        gene_set == modules.interest[[1]] ~ modules.interest[[1]],
                                        gene_set == modules.interest[[2]] ~ modules.interest[[2]],
                                        gene_set == modules.interest[[3]] ~ modules.interest[[3]])))  
```

```{r}
paste("No. of BP enriched in Amb vs Mod = ", gene_sets.BP %>% filter(gene_set=="amb-mod") %>% nrow())
paste("No. of BP enriched in Amb vs Low = ", gene_sets.BP %>% filter(gene_set=="amb-low") %>% nrow())
paste("No. of BP enriched in Mod vs Low = ", gene_sets.BP %>% filter(gene_set=="mod-low") %>% nrow())

paste("No. of MF enriched in Amb vs Mod = ", gene_sets.MF %>% filter(gene_set=="amb-mod") %>% nrow())
paste("No. of MF enriched in Amb vs Low = ", gene_sets.MF %>% filter(gene_set=="amb-low") %>% nrow())
paste("No. of MF enriched in Mod vs Low = ", gene_sets.MF %>% filter(gene_set=="mod-low") %>% nrow())

paste("No. of CC enriched in Amb vs Mod = ", gene_sets.CC %>% filter(gene_set=="amb-mod") %>% nrow())
paste("No. of CC enriched in Amb vs Low = ", gene_sets.CC %>% filter(gene_set=="amb-low") %>% nrow())
paste("No. of CC enriched in Mod vs Low = ", gene_sets.CC %>% filter(gene_set=="mod-low") %>% nrow())
```

```{r}
gene_sets.BP %>% filter(gene_set=="amb-mod") %>% dplyr::select(name) %>% write_clip()
gene_sets.BP %>% filter(gene_set=="amb-low") %>% dplyr::select(name) %>% write_clip()
gene_sets.BP %>% filter(gene_set=="mod-low") %>% dplyr::select(name) %>% write_clip()
```


## Bubble plot- enriched Biological Processes by module 

The column "gene_set_up" denotes the up/down for the DEG contrasts (it is zero for WGCNA modules)

```{r}
# Plot enriched Biological Processes 
pdf(file="Enriched_BP.pdf", height = 7, width = 7.5)
ggplot(gene_sets.BP, aes(y=str_wrap(name),x=gene_set, col=gene_set_up)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Upregulated", breaks=c("Ambient", "Moderate", "Low"), 
                     values=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c",
                                                  cyan="cyan", pink="pink", brown="brown", 
                                                  lightcyan1="lightcyan1", darkgreen="darkgreen", green="green"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Enriched Biological Processes in gene sets") +
  geom_vline(xintercept = c(3.5), colour="gray50", linetype="dashed") #+
  # geom_vline(xintercept = c(1.5,2.5,3.5,4.5,5.5,7.5), colour="gray50", linetype="dashed") 
  # annotate("text", x = 1.4, y = -.75, label = "Dabob Bay\nvs. Fidalgo Bay", size=2.8, col="gray15") +  
  # annotate("text", x = 3.5, y = -.75, label = "Dabob Bay\nvs. Oyster Bay", size=2.8, col="gray15") +  
  # annotate("text", x = 5.6, y = -.75, label = "Fidalgo Bay\nvs. Oyster Bay", size=2.8, col="gray15") +  
  # coord_cartesian(clip = "off", ylim = c(-1.75, 63.5))
dev.off()
```

```{r}
# Plot enriched Molecular Functions
pdf(file="Enriched_MF.pdf", height = 14, width = 7.5)
ggplot(gene_sets.MF, aes(y=str_wrap(name),x=gene_set, col=gene_set_up)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Upregulated", breaks=c("Ambient", "Moderate", "Low"), 
                     values=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c",
                                                  cyan="cyan", pink="pink", brown="brown", 
                                                  lightcyan1="lightcyan1", darkgreen="darkgreen", green="green"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Enriched Molecular Functions in gene sets") +
  geom_vline(xintercept = c(3.5), colour="gray50", linetype="dashed") #+
  # geom_vline(xintercept = c(1.5,2.5,3.5,4.5,5.5,7.5), colour="gray50", linetype="dashed") 
  # annotate("text", x = 1.4, y = -.75, label = "Dabob Bay\nvs. Fidalgo Bay", size=2.8, col="gray15") +  
  # annotate("text", x = 3.5, y = -.75, label = "Dabob Bay\nvs. Oyster Bay", size=2.8, col="gray15") +  
  # annotate("text", x = 5.6, y = -.75, label = "Fidalgo Bay\nvs. Oyster Bay", size=2.8, col="gray15") +  
  # coord_cartesian(clip = "off", ylim = c(-1.75, 63.5))
dev.off()
```

```{r}
# Plot enriched Cellular Components
pdf(file="Enriched_CC.pdf", height = 4.5, width = 5.5)

ggplot(gene_sets.CC, aes(y=str_wrap(name),x=gene_set, col=gene_set_up)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Upregulated", breaks=c("Ambient", "Moderate", "Low"), 
                     values=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c",
                                                  cyan="cyan", pink="pink", brown="brown", 
                                                  lightcyan1="lightcyan1", darkgreen="darkgreen", green="green"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Enriched Cellular Componenets in gene sets") +
  geom_vline(xintercept = c(3.5), colour="gray50", linetype="dashed") #+
  # geom_vline(xintercept = c(1.5,2.5,3.5,4.5,5.5,7.5), colour="gray50", linetype="dashed") 
  # annotate("text", x = 1.4, y = -.75, label = "Dabob Bay\nvs. Fidalgo Bay", size=2.8, col="gray15") +  
  # annotate("text", x = 3.5, y = -.75, label = "Dabob Bay\nvs. Oyster Bay", size=2.8, col="gray15") +  
  # annotate("text", x = 5.6, y = -.75, label = "Fidalgo Bay\nvs. Oyster Bay", size=2.8, col="gray15") +  
  # coord_cartesian(clip = "off", ylim = c(-1.75, 63.5))
dev.off()
```
```{r}
gene_sets.BP %>% filter(gene_set %in% modules) %>% droplevels() %>% 
  mutate(module=as.numeric(gene_set)) %>%
                  mutate(name=as.factor(name)) %>% 
  mutate_at("module", function(x) {
    case_when(
      x == 1 ~ 2,
      x == 6|7 ~ 8)}) %>%
  mutate(name=fct_reorder2(name, module, pval, .desc = TRUE))

?fct_reorder()
```


## ONLY WGCNA bubble plots 

```{r}
# Biological Processes
ggplot(gene_sets.BP %>% filter(gene_set %in% modules) %>% droplevels() %>% 
         filter(p.adj < 0.1),
       aes(x=gene_set, col=gene_set, y=str_wrap(name))) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Enriched Biological Processes in gene sets")

# Molecular Functions 
ggplot(gene_sets.MF %>% filter(gene_set %in% modules) %>% droplevels() %>% 
       filter(pval<0.1), 
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Enriched Molecular Functions in gene sets")

# Cellular components
ggplot(gene_sets.CC %>% filter(gene_set %in% modules) %>% droplevels() %>% 
       filter(pval<0.1), 
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Enriched Cellular Componenets in gene sets")
```


```{r}
gene_sets.BP %>% filter(gene_set %!in% modules) %>% droplevels() %>% 
         filter(name %in% unique(c(best_GO$`amb-mod_BP`$name, best_GO$`amb-low_BP`$name, best_GO$`mod-low_BP`$name))) %>% 
         mutate(name=as.factor(name)) %>% mutate(name=fct_reorder(name, delta.rank, .desc = TRUE))


```


## ONLY DEG bubble plots 

```{r}
# Create function that is the oppositve of %in% 
`%!in%` = Negate(`%in%`)

# Biological Processes
#pdf(file="Enriched_BP_DEGs.pdf", height = 13.5, width = 7.5)
ggplot(gene_sets.BP %>% filter(gene_set %!in% modules) %>% droplevels() %>% 
         #filter(p.adj<0.01) %>%
         filter(name %in% unique(c(best_GO$`amb-mod_BP`$name, best_GO$`amb-low_BP`$name, best_GO$`mod-low_BP`$name))) %>%  # use only the best GO terms (determined within GO_MWU ananlysis)
         mutate(name=as.factor(name)) %>% mutate(name=fct_reorder(name, delta.rank, .desc = TRUE)), 
       #aes(y=str_wrap(name),x=gene_set, col=gene_set_up)) + 
       aes(y=name,x=gene_set, col=gene_set_up)) + 
  geom_point(aes(alpha=p.adj, size=nseqs)) + #size=6,  
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_blank(), #element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.ticks.x = element_blank(), 
        legend.position = "right", 
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.25, label = "Ambient\nvs. Moderate", size=2.25, col="gray15") +  
  annotate("text", x = 2, y = -.25, label = "Ambient\nvs. Low", size=2.25, col="gray15") +  
  annotate("text", x = 3, y = -.25, label = "Moderate\nvs. Low", size=2.25, col="gray15") +  
  #coord_cartesian(clip = "off", ylim = c(-0.6, 80))
  coord_cartesian(clip = "off", ylim = c(-0.6, 20))
#dev.off()

# Molecular Functions 
#pdf(file="Enriched_MF_DEGs.pdf", height = 25, width = 7.5)
ggplot(gene_sets.MF %>% filter(gene_set %!in% modules) %>% droplevels() %>% 
          #filter(p.adj<0.01) %>%
           filter(name %in% unique(c(best_GO$`amb-mod_MF`$name, best_GO$`amb-low_MF`$name, best_GO$`mod-low_MF`$name))) %>%  # use only the best GO terms 
         mutate(name=as.factor(name)) %>% mutate(name=fct_reorder(name, delta.rank, .desc = TRUE)),  
#       aes(y=str_wrap(name),x=gene_set, col=gene_set_up)) + 
        aes(y=name,x=gene_set, col=gene_set_up)) + 
  geom_point(aes(alpha=p.adj, size=nseqs)) + #size=6,  
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_blank(), #element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.ticks.x = element_blank(), 
        legend.position = "right", 
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle(NULL) + 
  annotate("text", x = 1, y = -.25, label = "Ambient\nvs. Moderate", size=2.25, col="gray15") +  
  annotate("text", x = 2, y = -.25, label = "Ambient\nvs. Low", size=2.25, col="gray15") +  
  annotate("text", x = 3, y = -.25, label = "Moderate\nvs. Low", size=2.25, col="gray15") +  
  coord_cartesian(clip = "off", ylim = c(-.6, 27.25))
#dev.off()

# Cellular components
#pdf(file="Enriched_CC_DEGs.pdf", height = 5.5, width = 6)
ggplot(gene_sets.CC %>% filter(gene_set %!in% modules) %>% droplevels() %>% 
          filter(name %in% unique(c(best_GO$`amb-mod_CC`$name, best_GO$`amb-low_CC`$name, best_GO$`mod-low_CC`$name))) %>%  # use only the best GO terms 
          #filter(p.adj<0.01) %>%
         mutate(name=as.factor(name)) %>% mutate(name=fct_reorder(name, delta.rank, .desc = TRUE)),  
#       aes(y=str_wrap(name),x=gene_set, col=gene_set_up)) + 
        aes(y=name,x=gene_set, col=gene_set_up)) + 
  geom_point(aes(alpha=p.adj, size=nseqs)) + #size=6,  
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_color_manual(name="Upregulated", values=c(Ambient="#2c7bb6", Moderate="#fdae61", Low="#d7191c"),
             guide = guide_legend(override.aes = list(size=4))) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(3,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_blank(),  #element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.ticks.x = element_blank(), 
        legend.position = "right", 
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle(NULL) +
  annotate("text", x = 1, y = -.05, label = "Ambient\nvs. Moderate", size=2.25, col="gray15") +  
  annotate("text", x = 2, y = -.05, label = "Ambient\nvs. Low", size=2.25, col="gray15") +  
  annotate("text", x = 3, y = -.05, label = "Moderate\nvs. Low", size=2.25, col="gray15") +  
  coord_cartesian(clip = "off", ylim = c(-0.1, 4))
#dev.off()
```

```{r}
gene_sets.MF %>% filter(gene_set %!in% modules) %>% arrange(name)
```



```{r}
module.lineplots[[1]]
```


# Standard GO_MWU analysis (Fischer Exact Test) to look at enrichment of low Coeff. Variation upregulated DEGs, perhaps critical to function in OA?

This function performs standard GO enrichment analysis based on Fisher's exact test, using binary measure (1 or 0, i.e., either sgnificant or not). 


```{r}
go.mwu.standard <- function(file) {
  
  input = file
  goAnnotations = "DESeq-genes_for-GOMWU.tab"

for (j in 1:length(goDivisions)) {
  tryCatch({
  
# ------------- Calculating stats
# It might take a few minutes for MF and BP. Do not rerun it if you just want to replot the data with different cutoffs, go straight to gomwuPlot. If you change any of the numeric values below, delete the files that were generated in previos runs first.

gomwuStats(input, goDatabase, goAnnotations, goDivisions[j],
	perlPath="C:/Strawberry/perl/bin/perl.exe", # replace with full path to perl executable if it is not in your system's PATH already
	#largest=0.1,  # a GO category will not be considered if it contains more than this fraction of the total number of genes
	#smallest=5,   # a GO category should contain at least this many genes to be considered
	clusterCutHeight=0.30, # threshold for merging similar (gene-sharing) terms. See README for details.
#	Alternative="g" # by default the MWU test is two-tailed; specify "g" or "l" of you want to test for "greater" or "less" instead. 
	#Module=TRUE,Alternative="g" # un-remark this if you are analyzing a SIGNED WGCNA module (values: 0 for not in module genes, kME for in-module genes). In the call to gomwuPlot below, specify absValue=0.001 (count number of "good genes" that fall into the module)
#	Module=TRUE # un-remark this if you are analyzing an UNSIGNED WGCNA module 
)
# do not continue if the printout shows that no GO terms pass 10% FDR.


# ----------- Plotting results

#quartz()
results=gomwuPlot(input,goAnnotations,goDivisions[j],
 	#absValue=-log(0.05,10),   # genes with the measure value exceeding this will be counted as "good genes". This setting is for signed log-pvalues. Specify absValue=0.001 if you are doing Fisher's exact test for standard GO enrichment or analyzing a WGCNA module (all non-zero genes = "good genes").
 absValue=0.001,
 	#absValue=.5, # un-remark this if you are using log2-fold changes
 level1=0.05,level2=0.01,level3=0.001,	#stricter settings for many GO terms 
 #level1=0.1, # FDR threshold for plotting. 
 	#level2=.05, # FDR cutoff to print in regular (not italic) font.
 	#level3=.01, # FDR cutoff to print in large bold font.
  #level1=1, level2=0.2, level3=0.05, # Specify level1=1 to plot all GO categories containing genes exceeding the absValue.

 	txtsize=1,    # decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
 	treeHeight=0.5, # height of the hierarchical clustering tree
 	#colors=c(color, color, color, color, color)
 colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") # these are default colors, un-remar and change if needed
 )
mtext(paste(contrast, " - ", goDivisions[j], sep=""), side = 3, line = 3, adj = -.05, col="black")
 # manually rescale the plot so the tree matches the text 
 # if there are too many categories displayed, try make it more stringent with level1=0.05,level2=0.01,level3=0.001.  
 
 # text representation of results, with actual adjusted p-values
 print(results[[1]])


# ------- extracting representative GOs

# this module chooses GO terms that best represent *independent* groups of significant GO terms

pcut=1e-2 # adjusted pvalue cutoff for representative GO
hcut=0.9 # height at which cut the GO terms tree to get "independent groups". 

# plotting the GO tree with the cut level (un-remark the next two lines to plot)
# plot(results[[2]],cex=0.6)
# abline(h=hcut,col="red")

# cutting
ct=cutree(results[[2]],h=hcut)
annots=c();ci=1
for (ci in unique(ct)) {
  message(ci)
	rn=names(ct)[ct==ci]
	obs=grep("obsolete",rn)
	if(length(obs)>0) { rn=rn[-obs] }
	if (length(rn)==0) {next}
	rr=results[[1]][rn,]
	bestrr=rr[which(rr$pval==min(rr$pval)),]
	best=1
	if(nrow(bestrr)>1) {
		nns=sub(" .+","",row.names(bestrr))
		fr=c()
		for (i in 1:length(nns)) { fr=c(fr,eval(parse(text=nns[i]))) }
		best=which(fr==max(fr))
	}
	if (bestrr$pval[best]<=pcut) { annots=c(annots,sub("\\d+\\/\\d+ ","",row.names(bestrr)[best]))}
}

mwus=read.table(paste("MWU",goDivisions[j],input,sep="_"),header=T)
bestGOs=mwus[mwus$name %in% annots,]
print(bestGOs)
},
error=function(e){cat("ERROR:", conditionMessage(e), "\n")})
}
}
```


# Standard GO enrichment using DEG lists

```{r}
go.mwu.standard(file="DEGs_amb-mod_STANDARD.csv")
```

```{r}
go.mwu.standard(file="DEGs_amb-low_STANDARD.csv")
```

```{r}
go.mwu.standard(file="DEGs_mod-low_STANDARD.csv")
```


# Standard GO enrichment using the list of genes that are upregulated at very consistent levels 

```{r}
go.mwu.standard(file = "DEGs-upreg-lowCV_amb-mod.csv")
```

```{r}
go.mwu.standard(file = "DEGs-upreg-lowCV_amb-low.csv")
```

```{r}
go.mwu.standard(file = "DEGs-upreg-lowCV_amb.csv")
```

```{r}
go.mwu.standard(file = "WGCNA-module_purple.csv")
```




```{r}
(gene_sets.BP %>% filter(gene_set %in% modules.up) %>% droplevels() %>% 
                  mutate(module=as.numeric(gene_set)) %>% 
  arrange(gene_set,pval) %>%
  mutate(name=factor(name, name)))$name %>% levels()
```

# Let's to generate separate bubble plots for modules that have upregulated, downregulated, and "roof" (up then down)

# UPREGULATED IN OA 
```{r}
# Biological Processes
print(ggplot(gene_sets.BP %>% filter(gene_set %in% modules.up) %>% droplevels() %>% 
                  mutate(module=as.numeric(gene_set)) %>% 
  arrange(gene_set,pval) %>%
  mutate(name=factor(name, levels = name, ordered = T)),
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules.up,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Upregulated Genes"))

# Molecular Functions 
print(ggplot(gene_sets.MF %>% filter(gene_set %in% modules.up) %>% droplevels() %>% 
                  mutate(module=as.numeric(gene_set)) %>% 
  arrange(gene_set,pval) %>%
  mutate(name=factor(name, ordered = T)),
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules.up,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Upregulated Genes"))

# Cellular components
print(ggplot(gene_sets.CC %>% filter(gene_set %in% modules.up) %>% droplevels() %>% 
                  mutate(module=as.numeric(gene_set)) %>% 
  arrange(gene_set,pval) %>%
  mutate(name=factor(name, levels = name, ordered = T)),
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules.up,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Upregulated Genes"))

```
# DOWNREGULATED IN OA 
```{r}
# Biological Processes
print(ggplot(gene_sets.BP %>% filter(gene_set %in% modules.down) %>% droplevels() %>% 
                  mutate(module=as.numeric(gene_set)) %>% arrange(gene_set,pval) %>%
  mutate(name=factor(name, ordered = T)) %>% filter(pval<0.0001),
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules.down,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("downregulated Genes"))

# Molecular Functions 
print(ggplot(gene_sets.MF %>% filter(gene_set %in% modules.down) %>% droplevels() %>% filter(pval<0.01), 
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules.down,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("downregulated Genes"))

# Cellular components
print(ggplot(gene_sets.CC %>% filter(gene_set %in% modules.down) %>% droplevels() %>% filter(pval<0.05), 
       aes(y=str_wrap(name),x=gene_set, col=gene_set)) + 
  geom_point(aes(alpha=p.adj,size=nseqs))+
 facet_wrap(~category,scales="free", nrow = 2) +
  scale_alpha("FDR-adjusted\nP-value", range = c(1,0.35), #breaks = c(0, .25, .5, .75, 1), 
              guide = guide_legend(override.aes = list(size=4))) + 
  scale_size("Gene count", range = c(2,10), #breaks = c(2, 5, 10), 
             guide = guide_legend(override.aes = list(col="gray50"))) + 
  scale_color_manual(name="Gene Set", values=modules.down,
             guide = guide_legend(override.aes = list(size=4))) +
  scale_x_discrete(drop=FALSE) +
  theme_cleveland() + 
  theme(axis.title.x=element_blank(), 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        #axis.ticks.x = element_blank(), 
        legend.position = "right", axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=10)) +
  ylab("Enriched Gene Ontology (GO) Term") + 
  ggtitle("Downregulated Genes"))

```

## Prep for metacoder

metacoder creates really pretty network plots, BUT I can only get it to work if I use an old version of metacoder, and an old version of R. Therefore, I use RStudio Cloud. 

Metacoder needs a file with GO IDs, GO name, and other things like number of genes & pvalue. BUT my GO_MWU results lists multiple GO IDs, so ... 


### Identify GO ID that was selected to represent each group of GO terms. 

From the GO_MWU README file: "highly similar categories are merged according to complete linkage clustering based on the fraction of shared genes. The distance measure for clustering, introduced in Kosiol et al 2008, is the number of genes shared among the two GO categories within the analyzed dataset divided by the size of the smaller of the two categories. The resulting hierarchical tree is then “cut” at the adjustable “height” ('cutTreeHeight' parameter in the call to gomwuStats) to merge clustered categories. The default for cutTreeHeight is 0.25, implying that a group of categories will be merged if the most dissimilar two of them share >75% of genes included in the smaller of the two. **The merged categories inherit the name of the largest one.**" 

So, the GO_MWU results use terms that "represent" similar terms (and not GO_SLIMs). GO_MWU doesn't specify which GO ID was used (it just uses the name), so first I need to match the IDs to the names. 

## Using results from GO_MWU 

```{r}
#Here I import the go.obo object and convert to dataframe. 
install.packages("ontologyIndex")
require(ontologyIndex)
ontology <- get_ontology("go.obo") %>% as.data.frame()
goslims <- get_ontology("goslim_plant.obo") %>% as.data.frame()

# Identify the representative GO ID selected for each group of merged IDs, column "id", select columns, save to object
(BP.metacoder <- gene_sets.BP %>% left_join(ontology, by="name") %>% 
  dplyr::select(gene_set, id, name, nseqs, pval, p.adj, 
                delta.rank, gene_set_up) %>%
  rename(ID=id, term=name))

(MF.metacoder <- gene_sets.MF %>% left_join(ontology, by="name") %>% 
  dplyr::select(gene_set, id, name, nseqs, pval, p.adj, 
                delta.rank, gene_set_up) %>%
  rename(ID=id, term=name))

(CC.metacoder <- gene_sets.CC %>% left_join(ontology, by="name") %>% 
  dplyr::select(gene_set, id, name, nseqs, pval, p.adj, 
                delta.rank, gene_set_up) %>%
  rename(ID=id, term=name))

save(BP.metacoder, file = "BP.metacoder")
save(MF.metacoder, file = "MF.metacoder")
save(CC.metacoder, file = "CC.metacoder")
```

## USING RESULTS FROM DAVID 

```{bash}
## Remove files which list GO_MWU results filenames if they already exist  
rm DAVID_BP_filenames.txt
rm DAVID_MF_filenames.txt
rm DAVID_CC_filenames.txt

for file in DAVID_BP*
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 3  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> DAVID_BP_filenames.txt
done

# Molecular functions results 
for file in DAVID_MF*
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 3  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> DAVID_MF_filenames.txt
done

# Cellular components results 
for file in DAVID_CC*
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 3  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> DAVID_CC_filenames.txt
done
```

```{r}
# Create objects from DAVID files 

# BIOLOGICAL PROCESSES
filenames <- read_delim(file="DAVID_BP_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)
for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}
(DAVID.BP <- bind_rows(file_list, .id = "gene_set") %>% 
    mutate(category=as.factor("Biological Processes")) %>%  #add column designating go category 
    mutate(gene_set=factor(gene_set)) %>% #Convert gene_set column to factor, define all gene_sets 
    tidyr::complete(gene_set, category) %>% # add rows for gene_sets that are missing from dataframe
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename(c("nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR")))

# MOLECULAR PROCESSES 
filenames <- read_delim(file="DAVID_MF_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)
for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}
(DAVID.MF <- bind_rows(file_list, .id = "gene_set") %>% 
    mutate(category=as.factor("Biological Processes")) %>%  #add column designating go category 
    mutate(gene_set=factor(gene_set)) %>% #Convert gene_set column to factor, define all gene_sets 
    tidyr::complete(gene_set, category) %>% # add rows for gene_sets that are missing from dataframe
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename(c("nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR")))

# CELLULAR COMPONENTS 
filenames <- read_delim(file="DAVID_CC_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)
for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}
(DAVID.CC <- bind_rows(file_list, .id = "gene_set") %>% 
    mutate(category=as.factor("Biological Processes")) %>%  #add column designating go category 
    mutate(gene_set=factor(gene_set)) %>% #Convert gene_set column to factor, define all gene_sets 
    tidyr::complete(gene_set, category) %>% # add rows for gene_sets that are missing from dataframe
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename(c("nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR")))

save(DAVID.BP, file = "DAVID.BP")
save(DAVID.MF, file = "DAVID.MF")
save(DAVID.CC, file = "DAVID.CC")

DAVID.BP %>% View()
```








## Map GO terms to GO Slims 

```{r}
# Biological Processes
myIds <- (gene_sets.BP %>% left_join(ontology, by="name") %>% 
            filter(!is.na(term)))$term %>% unique()

myCollection <- GOCollection(myIds)
fl <- system.file("extdata", "goslim_plant.obo", package="GSEABase")
slim <- getOBOCollection(fl)

# Sets ontology category to "Molecular Function". Use "BP" for "Biological Process"
slimdf <- goSlim(myCollection, slim, ontology = "BP")

# This should match the ontology used above.
# E.g. GOBPOFFSPRING or GOCCOFFSPRING
GO.db::GOBPOFFSPRING

mappedIds <-
  function(df, collection, OFFSPRING)
  {
    map <- as.list(OFFSPRING[rownames(df)])
    mapped <- lapply(map, intersect, ids(collection))
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L))
    df
  }

slimdf <- mappedIds(slimdf, myCollection, GOBPOFFSPRING)
slimdf %>% filter(Count>0) %>% View()

write.csv(slimdf %>% filter(Count>0), file = "slim_bp.csv", quote = FALSE)

getOBOCollection(goDatabase) %>% str()
```

```{r}
## Molecular Functions 

myIds <- (gene_sets.MF %>% filter(gene_set %in% "amb-low") %>% 
            droplevels() %>% separate_rows(term, sep = ";") %>% 
            filter(!is.na(term)))$term %>% unique()
myCollection <- GOCollection(myIds)
fl <- system.file("extdata", "goslim_plant.obo", package="GSEABase")
slim <- getOBOCollection(fl)

# Sets ontology category to "Molecular Function". Use "BP" for "Biological Process"
slimdf <- goSlim(myCollection, slim, ontology = "MF")

# This should match the ontology used above.
# E.g. GOBPOFFSPRING or GOCCOFFSPRING
GO.db::GOMFOFFSPRING

mappedIds <-
  function(df, collection, OFFSPRING)
  {
    map <- as.list(OFFSPRING[rownames(df)])
    mapped <- lapply(map, intersect, ids(collection))
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L))
    df
  }

slimdf.AL <- mappedIds(slimdf, myCollection, GOMFOFFSPRING) %>% filter(Count>0)

left_join(
  slimdf.AL %>% separate_rows(go_terms, sep = ";") %>% dplyr::select(Term, go_terms),
  
  gene_sets.MF %>% filter(gene_set %in% "amb-mod") %>% 
            droplevels() %>% separate_rows(term, sep = ";") %>%
    dplyr::select(gene_set, delta.rank, p.adj, term, name, gene_set_up),
  by=c("go_terms"="term")) %>% filter(Term != "molecular_function") %>% View()



#slimdf %>% filter(Count>0) %>% View()
#write.csv(slimdf.AM, file = "slim_bp.csv", quote = FALSE)
```

# READ IN DAVID ENRICHMENT RESULTS! 

First create list of file names and module / DEG set 

```{bash}
## Remove files which list GO_MWU results filenames if they already exist  
rm DAVID_mods-degs_filenames.txt
rm DAVID_DEGs-FC0.5_filenames.txt
rm DAVID_mods-degs-GS_filenames.txt

# Modules (using only genes that are also DEGs)
for file in DAVID_*_DEGs.txt
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 2  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> DAVID_mods-degs_filenames.txt
done


# DEGs that are up/down in pairwise comparisons
for file in DAVID_*_FC0.5.txt
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 2  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> DAVID_DEGs-FC0.5_filenames.txt
done


# DEGs that are up/down in pairwise comparisons
for file in DAVID_*_DEGs-or-GS.txt
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "_" -f 2  | cut -d "." -f 1)"
printf "%s\t%s\n" "$filename" "$sample" >> DAVID_mods-degs-GS_filenames.txt
done

```

## Read in enriched Biological Processes 

IMPORTANT:  The files contain all results from DAVID's Functional Annotation Chart, which includes the GO_BP, GO_MF, and GO_CC, but also other enriched "things", like Uniprot Keywords (UP_KW_BIOLOGICAL_PROCESS, UP_KW_MOLECULAR_FUNCTION, UP_KW_CELLULAR_COMPONENT), and KEGG Pathways (KEGG_PATHWAY). Very cool!  It provides more insight into the broad functions of my gene sets.  

```{r}
filenames <- read_delim(file="DAVID_mods-degs_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}

(DAVID_mods_degs <- bind_rows(file_list, .id = "gene_set") %>% #filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(Category=as.factor(Category)) %>% 
    mutate(gene_set=factor(gene_set, levels=modules)) %>%  #Convert gene_set column to factor, define all gene_sets 
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename("category"="Category", "nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR") %>%

    # Create new column that identifies which treatment is upregulated in each DEG contrast (doesn't apply for WGCNA modules)
    mutate(OA.response = as.factor(case_when(gene_set %in% modules.up ~ "Upregulated",
                                   gene_set %in% modules.down ~ "Downregulated",
                                   gene_set %in% modules.roof ~ "Up-then-Down"))))
save(DAVID_mods_degs, file = "DAVID_mods_degs")
```
## Exploring results of DAVID 

```{r}
DAVID_mods_degs %>% filter(OA.response=="Upregulated") %>% View()
```


```{r}
# What are enriched GO terms? 

# Create objects for upregualted, downregualted, and up-then-down regulated 
up.go <- DAVID_mods_degs %>% filter(category %in% c("GOTERM_BP_DIRECT","GOTERM_CC_DIRECT", "GOTERM_MF_DIRECT")) %>% filter(OA.response=="Upregulated")
updown.go <- DAVID_mods_degs %>% filter(category %in% c("GOTERM_BP_DIRECT", "GOTERM_MF_DIRECT")) %>% filter(OA.response=="Up-then-Down")
down.go <- DAVID_mods_degs %>% filter(category %in% c("GOTERM_BP_DIRECT", "GOTERM_MF_DIRECT")) %>% filter(OA.response=="Downregulated")

# Check out BP and CC in upregulated, downreulgated, and up-then-down regulated genes. While doing so, remove terms that are based on the same exact genes, selecting the KW that has lowest p-value 

# Upregulated Biological processes
up.go %>% filter(category %in% "GOTERM_BP_DIRECT") %>% 
  group_by(gene_set, category, nseqs, Genes) %>%
  summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(up.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>%  #re-add data
  filter(pval<0.05) %>% View()

# Upregulated cellular componenets
up.go %>% filter(category %in% "GOTERM_CC_DIRECT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(up.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data

# Upregulated cellular componenets
up.go %>% filter(category %in% "GOTERM_MF_DIRECT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(up.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data

View(up.go)



# Up-then-down Biological processes
updown.go %>% filter(category %in% "GOTERM_BP_DIRECT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(updown.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data

# Up-then-down cellular componenets
updown.go %>% filter(category %in% "GOTERM_CC_DIRECT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(updown.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data



# Downregulated Biological processes
down.go %>% filter(category %in% "GOTERM_BP_DIRECT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(down.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>%
    filter(pval<0.05, nseqs>=3) %>% arrange(gene_set, pval) %>% 
  select(gene_set, ID, term, pval, nseqs, Genes) %>% View()
  filter(gene_set == "magenta") %>% nrow()
  head(n=10) %>% 
  write_clip()


# Downregulated Cellular Components
down.go %>% filter(category %in% "GOTERM_MF_DIRECT") %>% filter(pval<0.05, nseqs>=3) %>% arrange(gene_set, pval) %>% View()
  group_by(gene_set, category, nseqs, Genes) %>%
  summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(down.go %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>%
    filter(pval<0.05, nseqs>=3) %>% arrange(gene_set, pval) %>% 
  select(gene_set, ID, term, pval, nseqs, Genes) %>% View()



```

```{r}
# What are enriched Uniprot Keywords? 

# Create objects for upregualted, downregualted, and up-then-down regulated 
up.kw <- DAVID_mods_degs %>% filter(category == c("UP_KW_BIOLOGICAL_PROCESS", "UP_KW_CELLULAR_COMPONENT")) %>% filter(OA.response=="Upregulated")
updown.kw <- DAVID_mods_degs %>%  filter(category == c("UP_KW_BIOLOGICAL_PROCESS", "UP_KW_CELLULAR_COMPONENT")) %>% filter(OA.response=="Up-then-Down")
down.kw <- DAVID_mods_degs %>%filter(category == c("UP_KW_BIOLOGICAL_PROCESS", "UP_KW_CELLULAR_COMPONENT")) %>% filter(OA.response=="Downregulated")

# Check out BP and CC in upregulated, downreulgated, and up-then-down regulated genes. While doing so, remove terms that are based on the same exact genes, selecting the KW that has lowest p-value 

# Upregulated Biological processes
up.kw %>% filter(category %in% "UP_KW_BIOLOGICAL_PROCESS") %>% group_by(gene_set, category, nseqs, Genes) %>% View()
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(up.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data

# Upregulated cellular componenets
up.kw %>% filter(category %in% "UP_KW_CELLULAR_COMPONENT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(up.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data



# Up-then-down Biological processes
updown.kw %>% filter(category %in% "UP_KW_BIOLOGICAL_PROCESS") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(updown.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data

# Up-then-down cellular componenets
updown.kw %>% filter(category %in% "UP_KW_CELLULAR_COMPONENT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(updown.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data



# Downregulated Biological processes
down.kw %>% filter(category %in% "UP_KW_BIOLOGICAL_PROCESS") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(down.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>%
  filter(pval<0.05) %>% View() #re-add data

# Downregulated Biological processes
down.kw %>% filter(category %in% "UP_KW_CELLULAR_COMPONENT") %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(down.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) %>% View() #re-add data

```

## DAVID results using DEGs with abs(L2FC) > 0.5. Very simlar to the WGCNA results, so I'm going to use the above results. 

```{r}
filenames <- read_delim(file="DAVID_DEGs-FC0.5_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}

(DAVID_degs_fc0.5 <- bind_rows(file_list, .id = "gene_set") %>% #filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(Category=as.factor(Category)) %>% 
    separate(gene_set, into=c("gene_set", "OA.response")) %>%
    mutate(gene_set=factor(gene_set), OA.response=as.factor(OA.response)) %>%  #Convert gene_set column to factor, define all gene_sets 
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename("category"="Category", "nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR"))
save(DAVID_degs_fc0.5, file = "DAVID_degs_fc0.5")

# What are enriched Uniprot Keywords? 
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "UP_KW_BIOLOGICAL_PROCESS") %>% filter(OA.response=="UP") #in genes upregulated in OA  
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "UP_KW_BIOLOGICAL_PROCESS") %>% filter(OA.response=="DOWN") #in genes downreulgated in OA 

# enriched KEGG pathways?
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "KEGG_PATHWAY") %>% filter(OA.response=="UP")
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "KEGG_PATHWAY") %>% filter(OA.response=="DOWN")
```
```{r}
# What are enriched Uniprot Keywords? 

# Create objects for upregualted, downregualted, and up-then-down regulated 
up.kw <- DAVID_degs_fc0.5 %>% filter(category == "UP_KW_BIOLOGICAL_PROCESS") %>% filter(OA.response=="UP")
down.kw <- DAVID_degs_fc0.5 %>% filter(category == "UP_KW_BIOLOGICAL_PROCESS") %>% filter(OA.response=="DOWN")

# Remove Uniprot KWs that are based on the same exact genes, selecting the KW that has lowest p-value 
up.kw %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(up.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) #re-add data

down.kw %>% group_by(gene_set, category, nseqs, Genes) %>%
summarise(pval=min(pval)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
left_join(down.kw %>% select(gene_set, term, ID, Genes, pval, p.adj, percent, Fold.Enrichment, OA.response)) #re-add data

```

### DAVID enrich results from DEGs with FC>0.5 - how many enriched GO terms with p-val < 0.05? 
```{r}
DAVID_degs_fc0.5 %>% filter(category %in% c("GOTERM_BP_DIRECT", "GOTERM_MF_DIRECT", "GOTERM_CC_DIRECT")) %>%
  filter(pval<0.05) %>%
  group_by(gene_set, category, OA.response) %>% tally()
```

## DAVID results using gene modules, but filtering to retain those that are also DEGS OR have Gene Significance p-values < 0.05. 

```{r}
filenames <- read_delim(file="DAVID_mods-degs-GS_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}

(DAVID_mods_degs_GS <- bind_rows(file_list, .id = "gene_set") %>% #filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(Category=as.factor(Category)) %>% 
    mutate(gene_set=factor(gene_set, levels=modules)) %>%  #Convert gene_set column to factor, define all gene_sets 
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename("category"="Category", "nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR") %>%

    # Create new column that identifies which treatment is upregulated in each DEG contrast (doesn't apply for WGCNA modules)
    mutate(OA.response = as.factor(case_when(gene_set %in% modules.up ~ "Upregulated",
                                   gene_set %in% modules.down ~ "Downregulated",
                                   gene_set %in% modules.roof ~ "Up-then-Down"))))
save(DAVID_mods_degs_GS, file = "DAVID_mods_degs_GS")

load(file = "DAVID_mods_degs_GS")
```
What Uniprot Keywords were enriched in gene modules? 

```{r}
DAVID_mods_degs_GS %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
  filter(pval<0.05) %>%
  arrange(OA.response, gene_set, pval) %>%
  select(gene_set, ID, term, Fold.Enrichment, pval, p.adj) %>%
  mutate(across(c("Fold.Enrichment", "pval", "p.adj"), signif, 2))
```





# DAVID enrichment analysis - using DEGs that were also:
-- assigned to a pCO2-correlated module AND
-- correlated with pCO2 on an individual basis (p.GS < 0.05)

```{r}
# These are genes that are assigned to a pCO2-correlated module, AND correlate with pCO2 on an individual basis
sig.genes.wgcna <- (geneInfo %>%
  filter(moduleColor %in% modules))$id #%>% 
#  filter(p.GS.pco2<0.05))$id
```


```{r}
# BACKGROUND - all genes submitted to DESeq2 analysis 
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>%
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()


# -------
# DEGs between Ambient & Moderate - DOWNREGULATED IN MODERATE OA VS. AMBIENT
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  filter(gene %in% sig.genes.wgcna) %>%
  filter(padj<0.05) %>% 
  filter(log2FoldChange>0.5) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between Ambient & Moderate - UPREGULATED IN MODERATE OA VS. AMBIENT
res.pco2.AM %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  filter(gene %in% sig.genes.wgcna) %>%
  filter(padj<0.05) %>% 
  filter(log2FoldChange<(-0.5)) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# -------
# DEGs between Ambient & Low - DOWNREGULATED IN SEVERE OA VS. AMBIENT
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  filter(gene %in% sig.genes.wgcna) %>%
  filter(padj<0.05) %>% 
  filter(log2FoldChange>0.5) %>%
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between Ambient & Low - UPREGUALTED IN SEVERE OA VS. AMBIENT
res.pco2.AL %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  filter(gene %in% sig.genes.wgcna) %>%
  filter(padj<0.05) %>% 
  filter(log2FoldChange<(-0.5)) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# -------
# DEGs between Moderate & Low - DOWNREGULATED IN SEVERE OA VS MODERATE OA 
res.pco2.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  filter(gene %in% sig.genes.wgcna) %>%
  filter(padj<0.05) %>% 
  filter(log2FoldChange>0.5) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()

# DEGs between Moderate & Low - UPREGUALTED IN SEVERE OA VS MODERATE OA
res.pco2.ML %>% as.data.frame() %>% rownames_to_column("gene") %>% 
  filter(gene %in% sig.genes.wgcna) %>%
  filter(padj<0.05) %>% 
  filter(log2FoldChange<(-0.5)) %>% 
  left_join(P.camt.blast.GO %>% dplyr::select(gene_id, gene_ontology_i_ds, spid), by = c("gene"="gene_id")) %>% 
  dplyr::select(spid) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```


## DAVID results using DEGs that are ALSO assigned to a pCO2-correlated module AND correlated with pCO2 on an individual basis (p.GS < 0.05)

```{r}
filenames <- read_delim(file="DAVID_DEGs-FC0.5_filenames.txt", col_names = c("filename", "gene_set"))
files <- file.path(filenames$filename) #extract vector of filenames 
all(file.exists(files)) #easy code to check that all files exist! 

file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=files[i], delim = "\t"))}

(DAVID_degs_fc0.5 <- bind_rows(file_list, .id = "gene_set") %>% #filter(p.adj<0.1) %>%  # Bind dataframe for each gene_set together
    mutate(Category=as.factor(Category)) %>% 
    separate(gene_set, into=c("gene_set", "OA.response")) %>%
    mutate(gene_set=factor(gene_set), OA.response=as.factor(OA.response)) %>%  #Convert gene_set column to factor, define all gene_sets 
    separate(Term, into = c("ID", "term"), sep = "~") %>%
    rename("category"="Category", "nseqs"="Count", "percent"="X.", "pval"="PValue", "p.adj"="FDR"))
save(DAVID_degs_fc0.5, file = "DAVID_degs_fc0.5")

# What are enriched Uniprot Keywords? 
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "UP_KW_BIOLOGICAL_PROCESS") %>% filter(OA.response=="UP") #in genes upregulated in OA  
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "UP_KW_BIOLOGICAL_PROCESS") %>% filter(OA.response=="DOWN") #in genes downreulgated in OA 

# enriched KEGG pathways?
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "KEGG_PATHWAY") %>% filter(OA.response=="UP")
DAVID_degs_fc0.5 %>% filter(gene_set != "ML") %>% filter(category == "KEGG_PATHWAY") %>% filter(OA.response=="DOWN")
```


